// @portinari/portinari-templates/esm2015/lib/components/po-page-dynamic-search/po-advanced-filter/po-advanced-filter-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { PoModalComponent } from '@portinari/portinari-ui';
import { poLocaleDefault } from '../../../utils/util';
/** @type {?} */
export const poAdvancedFiltersLiteralsDefault = {
    en: (/** @type {?} */ ({
        title: 'Advanced search',
        cancelLabel: 'Cancel',
        confirmLabel: 'Apply filters'
    })),
    es: (/** @type {?} */ ({
        title: 'Búsqueda avanzada',
        cancelLabel: 'Cancelar',
        confirmLabel: 'Aplicar filtros'
    })),
    pt: (/** @type {?} */ ({
        title: 'Busca avançada',
        cancelLabel: 'Cancelar',
        confirmLabel: 'Aplicar filtros'
    })),
    ru: (/** @type {?} */ ({
        title: 'Расширенный поиск',
        cancelLabel: 'отменить',
        confirmLabel: 'Применить фильтры'
    }))
};
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Filtro de busca avançada criado a partir de um formulário dinâmico.
 * Componente de uso interno.
 */
export class PoAdvancedFilterBaseComponent {
    /**
     * @param {?} languageService
     */
    constructor(languageService) {
        this._filters = [];
        this.filter = {};
        this.language = poLocaleDefault;
        this.primaryAction = {
            action: (/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const models = this.getValuesFromForm();
                this.searchEvent.emit(models);
                this.poModal.close();
            }),
            label: this.literals.confirmLabel
        };
        this.secondaryAction = {
            action: (/**
             * @return {?}
             */
            () => {
                this.poModal.close();
            }),
            label: this.literals.cancelLabel
        };
        /**
         * Função que será disparada e receberá os valores do formulário ao ser clicado no botão buscar.
         */
        this.searchEvent = new EventEmitter();
        this.language = languageService.getShortLanguage();
    }
    /**
     * Coleção de objetos que implementam a interface PoDynamicFormField, para definição dos campos que serão criados
     * dinamicamente.
     * @param {?} filters
     * @return {?}
     */
    set filters(filters) {
        this._filters = Array.isArray(filters) ? [...filters] : [];
    }
    /**
     * @return {?}
     */
    get filters() {
        return this._filters;
    }
    /**
     * Objeto com as literais usadas no `po-advanced-filter`.
     * @param {?} value
     * @return {?}
     */
    set literals(value) {
        if (value instanceof Object && !(value instanceof Array)) {
            this._literals = Object.assign({}, poAdvancedFiltersLiteralsDefault[poLocaleDefault], poAdvancedFiltersLiteralsDefault[this.language], value);
        }
        else {
            this._literals = poAdvancedFiltersLiteralsDefault[this.language];
        }
        this.primaryAction.label = this.literals.confirmLabel;
        this.secondaryAction.label = this.literals.cancelLabel;
    }
    /**
     * @return {?}
     */
    get literals() {
        return this._literals || poAdvancedFiltersLiteralsDefault[this.language];
    }
    // Retorna os models dos campos preenchidos
    /**
     * @private
     * @return {?}
     */
    getValuesFromForm() {
        Object.keys(this.filter).forEach((/**
         * @param {?} property
         * @return {?}
         */
        property => {
            if (this.filter[property] === undefined || this.filter[property] === '') {
                delete this.filter[property];
            }
        }));
        return this.filter;
    }
}
PoAdvancedFilterBaseComponent.propDecorators = {
    poModal: [{ type: ViewChild, args: [PoModalComponent, { static: true },] }],
    filters: [{ type: Input, args: ['p-filters',] }],
    literals: [{ type: Input, args: ['p-literals',] }],
    searchEvent: [{ type: Output, args: ['p-search-event',] }]
};
if (false) {
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.poModal;
    /**
     * @type {?}
     * @private
     */
    PoAdvancedFilterBaseComponent.prototype._filters;
    /**
     * @type {?}
     * @private
     */
    PoAdvancedFilterBaseComponent.prototype._literals;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.filter;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.language;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.primaryAction;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.secondaryAction;
    /**
     * Função que será disparada e receberá os valores do formulário ao ser clicado no botão buscar.
     * @type {?}
     */
    PoAdvancedFilterBaseComponent.prototype.searchEvent;
}
