// @portinari/portinari-templates/esm2015/lib/components/po-page-dynamic-search/po-advanced-filter/po-advanced-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { PoDynamicFormComponent, PoLanguageService } from '@portinari/portinari-ui';
import { PoAdvancedFilterBaseComponent } from './po-advanced-filter-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends PoAdvancedFilterBaseComponent
 *
 * \@examplePrivate
 *
 * <example-private name="po-advanced-filter" title="Portinari Busca Avançada">
 *   <file name="sample-po-advanced-filter.component.html"> </file>
 *   <file name="sample-po-advanced-filter.component.ts"> </file>
 * </example-private>
 */
export class PoAdvancedFilterComponent extends PoAdvancedFilterBaseComponent {
    /**
     * @param {?} languageService
     */
    constructor(languageService) {
        super(languageService);
    }
    /**
     * @return {?}
     */
    open() {
        this.filter = {};
        this.poModal.open();
    }
}
PoAdvancedFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-advanced-filter',
                template: "<po-modal\r\n  p-hide-close\r\n  [p-primary-action]=\"primaryAction\"\r\n  [p-secondary-action]=\"secondaryAction\"\r\n  [p-title]=\"literals.title\">\r\n\r\n  <po-dynamic-form\r\n    [p-fields]=\"filters\"\r\n    [p-value]=\"filter\">\r\n  </po-dynamic-form>\r\n\r\n</po-modal>\r\n"
            }] }
];
/** @nocollapse */
PoAdvancedFilterComponent.ctorParameters = () => [
    { type: PoLanguageService }
];
PoAdvancedFilterComponent.propDecorators = {
    poDynamicForm: [{ type: ViewChild, args: [PoDynamicFormComponent, { static: true },] }]
};
if (false) {
    /** @type {?} */
    PoAdvancedFilterComponent.prototype.poDynamicForm;
}
