// @portinari/portinari-templates/esm5/lib/components/po-page-job-scheduler/po-page-job-scheduler-execution/po-page-job-scheduler-execution.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
import { isTypeof } from '../../../utils/util';
import { PoPageJobSchedulerLookupService } from '../po-page-job-scheduler-lookup.service';
import { PoPageJobSchedulerService } from '../po-page-job-scheduler.service';
var PoPageJobSchedulerExecutionComponent = /** @class */ (function () {
    function PoPageJobSchedulerExecutionComponent(poPageJobSchedulerService, poPageJobSchedulerLookup) {
        this.poPageJobSchedulerService = poPageJobSchedulerService;
        this.poPageJobSchedulerLookup = poPageJobSchedulerLookup;
        this._value = {};
        this.dayPattern = '^(3[0-1]|[0-2][0-9]|[1-9]|0[1-9])$';
        this.existProcessAPI = true;
        this.minDateFirstExecution = new Date();
        this.periodicityOptions = [];
        this.timePattern = '^(2[0-3]|[01][0-9]):?([0-5][0-9])$';
        this.weekDays = [];
        this.isEdit = false;
        this.literals = (/** @type {?} */ ({}));
        this.changeProcess = new EventEmitter();
    }
    Object.defineProperty(PoPageJobSchedulerExecutionComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._value = value && isTypeof(value, 'object') ? value : {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageJobSchedulerExecutionComponent.prototype, "startDateFirstExecution", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isEdit ? undefined : this.minDateFirstExecution;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoPageJobSchedulerExecutionComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.subscribeProcessIdValueChanges();
        }));
    };
    /**
     * @return {?}
     */
    PoPageJobSchedulerExecutionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.periodicityTemplates = {
            daily: this.dailyTemplate,
            monthly: this.monthlyTemplate,
            weekly: this.weeklyTempalte
        };
        this.checkExistsProcessesAPI();
        this.periodicityOptions = this.getPeriodicityOptions();
        this.weekDays = this.getWeekDays();
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerExecutionComponent.prototype.checkExistsProcessesAPI = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.poPageJobSchedulerService.getHeadProcesses().subscribe(undefined, (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.existProcessAPI = false;
        }));
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerExecutionComponent.prototype.getPeriodicityOptions = /**
     * @private
     * @return {?}
     */
    function () {
        return [
            { label: this.literals.single, value: 'single' },
            { label: this.literals.daily, value: 'daily' },
            { label: this.literals.weekly, value: 'weekly' },
            { label: this.literals.monthly, value: 'monthly' }
        ];
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerExecutionComponent.prototype.getWeekDays = /**
     * @private
     * @return {?}
     */
    function () {
        return [
            { label: this.literals.sunday, value: 'Sunday' },
            { label: this.literals.monday, value: 'Monday' },
            { label: this.literals.tuesday, value: 'Tuesday' },
            { label: this.literals.wednesday, value: 'Wednesday' },
            { label: this.literals.thursday, value: 'Thursday' },
            { label: this.literals.friday, value: 'Friday' },
            { label: this.literals.saturday, value: 'Saturday' }
        ];
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerExecutionComponent.prototype.subscribeProcessIdValueChanges = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.form.controls['processID'].valueChanges.subscribe((/**
         * @param {?} processId
         * @return {?}
         */
        function (processId) {
            _this.changeProcess.emit({ processId: processId, existAPI: _this.existProcessAPI });
        }));
    };
    PoPageJobSchedulerExecutionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-job-scheduler-execution',
                    template: "<form #formExecution=\"ngForm\">\r\n  <div class=\"po-row\">\r\n    <ng-container *ngIf=\"existProcessAPI; then lookupProcessesID; else inputProcessesID\"></ng-container>\r\n  </div>\r\n\r\n  <div class=\"po-row\">\r\n    <po-divider class=\"po-md-12\" [p-label]=\"literals.firstExecution\"></po-divider>\r\n\r\n    <po-datepicker\r\n      class=\"po-md-4\"\r\n      name=\"firstExecution\"\r\n      [(ngModel)]=\"value.firstExecution\"\r\n      p-placeholder=\"dd/mm/aaaa\"\r\n      p-required\r\n      [p-disabled]=\"isEdit\"\r\n      [p-label]=\"literals.day\"\r\n      [p-min-date]=\"startDateFirstExecution\">\r\n    </po-datepicker>\r\n\r\n    <po-input\r\n      class=\"po-md-2\"\r\n      name=\"firstExecutionHour\"\r\n      [(ngModel)]=\"value.firstExecutionHour\"\r\n      p-mask=\"99:99\"\r\n      p-mask-format-model\r\n      p-placeholder=\"HH:mm\"\r\n      p-required\r\n      [p-disabled]=\"isEdit\"\r\n      [p-label]=\"literals.hour\"\r\n      [p-pattern]=\"timePattern\">\r\n    </po-input>\r\n\r\n  </div>\r\n\r\n  <hr>\r\n\r\n  <div class=\"po-row\">\r\n    <po-radio-group\r\n      class=\"po-sm-12\"\r\n      name=\"periodicity\"\r\n      [(ngModel)]=\"value.periodicity\"\r\n      p-columns=\"4\"\r\n      p-required\r\n      [p-label]=\"literals.periodicity\"\r\n      [p-options]=\"periodicityOptions\">\r\n    </po-radio-group>\r\n\r\n    <ng-container *ngIf=\"value.periodicity !== 'single'\">\r\n      <po-divider class=\"po-md-12\" [p-label]=\"literals.periodicityData\">\r\n      </po-divider>\r\n\r\n      <div class=\"po-row\">\r\n        <ng-container *ngTemplateOutlet=\"periodicityTemplates[value.periodicity]\">\r\n        </ng-container>\r\n      </div>\r\n\r\n      <div class=\"po-row\">\r\n        <po-switch\r\n          class=\"po-md-3\"\r\n          name=\"recurrent\"\r\n          [(ngModel)]=\"value.recurrent\"\r\n          [p-label]=\"literals.recurrent\"\r\n          [p-label-off]=\"literals.no\"\r\n          [p-label-on]=\"literals.yes\">\r\n        </po-switch>\r\n      </div>\r\n    </ng-container>\r\n  </div>\r\n\r\n  <ng-template #lookupProcessesID>\r\n    <po-lookup\r\n      class=\"po-md-12\"\r\n      name=\"processID\"\r\n      [(ngModel)]=\"value.processID\"\r\n      p-field-label=\"description\"\r\n      p-field-value=\"processID\"\r\n      p-required\r\n      [p-disabled]=\"isEdit\"\r\n      [p-filter-service]=\"poPageJobSchedulerLookup\"\r\n      [p-label]=\"literals.process\"\r\n      [p-placeholder]=\"literals.enterProcess\">\r\n    </po-lookup>\r\n  </ng-template>\r\n\r\n  <ng-template #inputProcessesID>\r\n    <po-input\r\n      class=\"po-md-12\"\r\n      name=\"processID\"\r\n      [(ngModel)]=\"value.processID\"\r\n      p-required\r\n      [p-disabled]=\"isEdit\"\r\n      [p-label]=\"literals.process\"\r\n      [p-placeholder]=\"literals.enterProcess\">\r\n    </po-input>\r\n  </ng-template>\r\n\r\n  <ng-template #dailyTemplate>\r\n    <ng-container *ngTemplateOutlet=\"inputHourTemplate\">\r\n    </ng-container>\r\n  </ng-template>\r\n\r\n  <ng-template #weeklyTemplate>\r\n    <div class=\"po-row\">\r\n      <ng-container *ngTemplateOutlet=\"inputHourTemplate\">\r\n      </ng-container>\r\n    </div>\r\n\r\n    <po-checkbox-group\r\n      class=\"po-md-12\"\r\n      name=\"daysOfWeek\"\r\n      [(ngModel)]=\"value.daysOfWeek\"\r\n      p-columns=\"4\"\r\n      p-required\r\n      [p-label]=\"literals.weekDays\"\r\n      [p-options]=\"weekDays\">\r\n    </po-checkbox-group>\r\n  </ng-template>\r\n\r\n  <ng-template #monthlyTemplate>\r\n    <po-number\r\n      class=\"po-md-3\"\r\n      name=\"dayOfMonth\"\r\n      [(ngModel)]=\"value.dayOfMonth\"\r\n      p-required\r\n      [p-error-pattern]=\"'Dia inv\u00E1lido'\"\r\n      [p-label]=\"literals.day\"\r\n      [p-max]=\"31\"\r\n      [p-pattern]=\"dayPattern\">\r\n    </po-number>\r\n\r\n    <ng-container *ngTemplateOutlet=\"inputHourTemplate\">\r\n    </ng-container>\r\n  </ng-template>\r\n\r\n  <ng-template #inputHourTemplate>\r\n    <po-input\r\n      class=\"po-md-2\"\r\n      name=\"hour\"\r\n      [(ngModel)]=\"value.hour\"\r\n      p-mask=\"99:99\"\r\n      p-mask-format-model\r\n      p-placeholder=\"HH:mm\"\r\n      p-required\r\n      [p-label]=\"literals.hour\"\r\n      [p-pattern]=\"timePattern\">\r\n    </po-input>\r\n  </ng-template>\r\n</form>\r\n"
                }] }
    ];
    /** @nocollapse */
    PoPageJobSchedulerExecutionComponent.ctorParameters = function () { return [
        { type: PoPageJobSchedulerService },
        { type: PoPageJobSchedulerLookupService }
    ]; };
    PoPageJobSchedulerExecutionComponent.propDecorators = {
        form: [{ type: ViewChild, args: ['formExecution', { static: true },] }],
        dailyTemplate: [{ type: ViewChild, args: ['dailyTemplate', { static: true },] }],
        monthlyTemplate: [{ type: ViewChild, args: ['monthlyTemplate', { static: true },] }],
        weeklyTempalte: [{ type: ViewChild, args: ['weeklyTemplate', { static: true },] }],
        isEdit: [{ type: Input, args: ['p-is-edit',] }],
        literals: [{ type: Input, args: ['p-literals',] }],
        value: [{ type: Input, args: ['p-value',] }],
        changeProcess: [{ type: Output, args: ['p-change-process',] }]
    };
    return PoPageJobSchedulerExecutionComponent;
}());
export { PoPageJobSchedulerExecutionComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerExecutionComponent.prototype._value;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.dayPattern;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.existProcessAPI;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.minDateFirstExecution;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.periodicityOptions;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.periodicityTemplates;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.timePattern;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.weekDays;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.form;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.dailyTemplate;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.monthlyTemplate;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.weeklyTempalte;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.isEdit;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.literals;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.changeProcess;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerExecutionComponent.prototype.poPageJobSchedulerService;
    /** @type {?} */
    PoPageJobSchedulerExecutionComponent.prototype.poPageJobSchedulerLookup;
}
