// @portinari/portinari-templates/esm5/lib/components/po-page-blocked-user/po-page-blocked-user-contacts/po-page-blocked-user-contacts.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, Input, ViewChild } from '@angular/core';
/** @type {?} */
var poPageBlockedUserContactItemMargin = 16;
var PoPageBlockedUserContactsComponent = /** @class */ (function () {
    function PoPageBlockedUserContactsComponent(changeDetector) {
        this.changeDetector = changeDetector;
        this.overflowItem = false;
    }
    /**
     * @return {?}
     */
    PoPageBlockedUserContactsComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.checkContactItemWidth();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    PoPageBlockedUserContactsComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.email || changes.phone) {
            this.checkContactItemWidth();
        }
    };
    /**
     * @private
     * @return {?}
     */
    PoPageBlockedUserContactsComponent.prototype.checkContactItemWidth = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.email || !this.phone) {
            this.overflowItem = true;
            return;
        }
        else {
            setTimeout((/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var phoneWidth = _this.phoneItem.nativeElement.offsetWidth;
                /** @type {?} */
                var mailWidth = _this.mailItem.nativeElement.offsetWidth;
                /** @type {?} */
                var contactGroupHalfWidth = _this.contactGroup.nativeElement.offsetWidth / 2;
                _this.overflowItem = phoneWidth > contactGroupHalfWidth || mailWidth > contactGroupHalfWidth - poPageBlockedUserContactItemMargin;
            }));
        }
        this.changeDetector.detectChanges();
    };
    PoPageBlockedUserContactsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-blocked-user-contacts',
                    template: "<div #contactGroup class=\"po-page-blocked-user-contact-group po-row po-mb-3\">\n  <div *ngIf=\"phone\"\n    class=\"po-page-blocked-user-contact-item\"\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\">\n    <a\n      class=\"po-page-blocked-user-link po-clickable\"\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"tel:{{ phone }}\"\n      target=\"_self\">\n      <div class=\"po-page-blocked-user-contact-group-item\">\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-telephone po-pr-1\"></span>\n        <span #phoneItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ phone }}</span>\n      </div>\n    </a>\n  </div>\n  <div *ngIf=\"email\"\n    class=\"po-page-blocked-user-contact-item po-page-blocked-user-contact-mail\"\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\">\n    <a\n      class=\"po-page-blocked-user-link po-clickable\"\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"mailto:{{ email }}\"\n      target=\"_self\">\n      <div class=\"po-page-blocked-user-contact-group-item\">\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-mail po-pr-1 \"></span>\n        <span #mailItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ email }}</span>\n      </div>\n    </a>\n  </div>\n</div>"
                }] }
    ];
    /** @nocollapse */
    PoPageBlockedUserContactsComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    PoPageBlockedUserContactsComponent.propDecorators = {
        email: [{ type: Input, args: ['p-email',] }],
        phone: [{ type: Input, args: ['p-phone',] }],
        contactGroup: [{ type: ViewChild, args: ['contactGroup', { static: true },] }],
        mailItem: [{ type: ViewChild, args: ['mailItem', { static: true },] }],
        phoneItem: [{ type: ViewChild, args: ['phoneItem', { static: true },] }]
    };
    return PoPageBlockedUserContactsComponent;
}());
export { PoPageBlockedUserContactsComponent };
if (false) {
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.literals;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.overflowItem;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserContactsComponent.prototype.mailText;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserContactsComponent.prototype.phoneText;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.email;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.phone;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.contactGroup;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.mailItem;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.phoneItem;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserContactsComponent.prototype.changeDetector;
}
