// @portinari/portinari-templates/esm5/lib/components/po-page-blocked-user/po-page-blocked-user.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ActivatedRoute, Router } from '@angular/router';
import { Component } from '@angular/core';
import { browserLanguage, isExternalLink, poLocaleDefault } from '../../utils/util';
import { PoPageBlockedUserBaseComponent } from './po-page-blocked-user-base.component';
/** @type {?} */
export var poPageBlockedUserButtonLiterals = {
    en: (/** @type {?} */ ({
        primaryButton: 'Back to home screen',
    })),
    es: (/** @type {?} */ ({
        primaryButton: 'Volver al inicio',
    })),
    pt: (/** @type {?} */ ({
        primaryButton: 'Voltar para o início',
    })),
    ru: (/** @type {?} */ ({
        primaryButton: 'Перейти к началу страницы',
    }))
};
/**
 * \@docsExtends PoPageBlockedUserBaseComponent
 *
 * \@example
 *
 * <example name="po-page-blocked-user-basic" title="Portinari Page Blocked User Basic">
 *  <file name="sample-po-page-blocked-user-basic/sample-po-page-blocked-user-basic.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-basic/sample-po-page-blocked-user-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-blocked-user-labs" title="Portinari Page Blocked User Labs">
 *  <file name="sample-po-page-blocked-user-labs/sample-po-page-blocked-user-labs.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-labs/sample-po-page-blocked-user-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-blocked-user-exceeded-attempts" title="Portinari Page Blocked User Exceeded Attempts">
 *  <file name="sample-po-page-blocked-user-exceeded-attempts/sample-po-page-blocked-user-exceeded-attempts.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-exceeded-attempts/sample-po-page-blocked-user-exceeded-attempts.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-blocked-user-expired-password" title="Portinari Page Blocked User Expired Password">
 *  <file name="sample-po-page-blocked-user-expired-password/sample-po-page-blocked-user-expired-password.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-expired-password/sample-po-page-blocked-user-expired-password.component.ts"> </file>
 * </example>
 */
var PoPageBlockedUserComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoPageBlockedUserComponent, _super);
    function PoPageBlockedUserComponent(activatedRoute, router) {
        var _this = _super.call(this) || this;
        _this.activatedRoute = activatedRoute;
        _this.router = router;
        _this.literals = tslib_1.__assign({}, poPageBlockedUserButtonLiterals[poLocaleDefault], poPageBlockedUserButtonLiterals[browserLanguage()]);
        return _this;
    }
    /**
     * @return {?}
     */
    PoPageBlockedUserComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    PoPageBlockedUserComponent.prototype.navigateTo = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        isExternalLink(url) ? window.open(url) : this.router.navigate([url || '/']);
    };
    /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    PoPageBlockedUserComponent.prototype.checkingForMetadataProperty = /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    function (object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    };
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    PoPageBlockedUserComponent.prototype.checkingForRouteMetadata = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (Object.keys(data).length !== 0) {
            this.contactEmail = this.checkingForMetadataProperty(data, 'contactEmail') || this.contactEmail;
            this.contactPhone = this.checkingForMetadataProperty(data, 'contactPhone') || this.contactPhone;
            this.reason = this.checkingForMetadataProperty(data, 'reason') || this.reason;
            this.urlBack = this.checkingForMetadataProperty(data, 'urlBack') || this.urlBack;
        }
    };
    PoPageBlockedUserComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-blocked-user',
                    template: "<div class=\"po-page-blocked-user-wrapper\">\n  <div class=\"po-page-background-wrap\">\n    <po-page-background\n      [p-hide-logo]=\"true\"\n      [p-logo]=\"logo\"\n      [p-secondary-logo]=\"secondaryLogo\">\n\n      <po-page-blocked-user-reason\n        [p-params]=\"params\"\n        [p-reason]=\"reason\">\n      </po-page-blocked-user-reason>\n\n      <div class=\"po-page-blocked-user-mobile-bottom-alignment\">\n        <po-page-blocked-user-contacts\n          [p-email]=\"contactEmail\"\n          [p-phone]=\"contactPhone\">\n        </po-page-blocked-user-contacts>\n\n        <div class=\"po-row\">\n          <po-button\n            class=\"po-lg-12 po-sm-12 po-page-blocked-user-button\"\n            p-type=\"primary\"\n            [p-label]=\"literals.primaryButton\"\n            (p-click)=\"navigateTo(urlBack)\">\n          </po-button>\n        </div>\n      </div>\n    </po-page-background>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    PoPageBlockedUserComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: Router }
    ]; };
    return PoPageBlockedUserComponent;
}(PoPageBlockedUserBaseComponent));
export { PoPageBlockedUserComponent };
if (false) {
    /** @type {?} */
    PoPageBlockedUserComponent.prototype.literals;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserComponent.prototype.router;
}
