// @portinari/portinari-templates/esm5/lib/services/po-page-customization/po-page-customization.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { from } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
var PoPageCustomizationService = /** @class */ (function () {
    function PoPageCustomizationService(http) {
        this.http = http;
    }
    /**
     * @template T
     * @param {?} origin
     * @param {?} originalOption
     * @param {?} optionSchema
     * @return {?}
     */
    PoPageCustomizationService.prototype.getCustomOptions = /**
     * @template T
     * @param {?} origin
     * @param {?} originalOption
     * @param {?} optionSchema
     * @return {?}
     */
    function (origin, originalOption, optionSchema) {
        var _this = this;
        return this.createObservable(origin).pipe(map((/**
         * @param {?} newPageOptions
         * @return {?}
         */
        function (newPageOptions) { return _this.mergePageOptions(originalOption, newPageOptions, optionSchema); })));
    };
    /**
     * @template T, K
     * @param {?} objectToChange
     * @param {?} newOptions
     * @return {?}
     */
    PoPageCustomizationService.prototype.changeOriginalOptionsToNewOptions = /**
     * @template T, K
     * @param {?} objectToChange
     * @param {?} newOptions
     * @return {?}
     */
    function (objectToChange, newOptions) {
        Object.keys(newOptions).forEach((/**
         * @param {?} key
         * @return {?}
         */
        function (key) {
            /** @type {?} */
            var value = newOptions[key];
            if (objectToChange[key]) {
                if (Array.isArray(value)) {
                    objectToChange[key] = tslib_1.__spread(value);
                    return;
                }
                if ((typeof (value) === 'number' || typeof (value) === 'string')) {
                    objectToChange[key] = value;
                    return;
                }
                if (typeof (value) === 'object') {
                    objectToChange[key] = tslib_1.__assign({}, value);
                }
            }
        }));
    };
    /**
     * @private
     * @template T
     * @param {?} origin
     * @return {?}
     */
    PoPageCustomizationService.prototype.createObservable = /**
     * @private
     * @template T
     * @param {?} origin
     * @return {?}
     */
    function (origin) {
        if (typeof origin === 'string') {
            return this.http.post(origin, {});
        }
        return from(Promise.resolve(origin()));
    };
    /**
     * @private
     * @template T
     * @param {?} originalOption
     * @param {?} newPageOptions
     * @param {?} optionSchema
     * @return {?}
     */
    PoPageCustomizationService.prototype.mergePageOptions = /**
     * @private
     * @template T
     * @param {?} originalOption
     * @param {?} newPageOptions
     * @param {?} optionSchema
     * @return {?}
     */
    function (originalOption, newPageOptions, optionSchema) {
        var _this = this;
        /** @type {?} */
        var mergePageOptions = optionSchema.schema.reduce((/**
         * @param {?} objWithNewProp
         * @param {?} prop
         * @return {?}
         */
        function (objWithNewProp, prop) {
            var _a;
            return (tslib_1.__assign({}, objWithNewProp, (_a = {}, _a[prop.nameProp] = _this.createNewProp(prop, originalOption, newPageOptions), _a)));
        }), (/** @type {?} */ ({})));
        Object.keys(mergePageOptions).forEach((/**
         * @param {?} key
         * @return {?}
         */
        function (key) { return !mergePageOptions[key] && delete mergePageOptions[key]; }));
        return mergePageOptions;
    };
    /**
     * @private
     * @template T
     * @param {?} prop
     * @param {?} originalOption
     * @param {?} newPageOptions
     * @return {?}
     */
    PoPageCustomizationService.prototype.createNewProp = /**
     * @private
     * @template T
     * @param {?} prop
     * @param {?} originalOption
     * @param {?} newPageOptions
     * @return {?}
     */
    function (prop, originalOption, newPageOptions) {
        if (prop.merge) {
            return this.mergeOptions(originalOption[prop.nameProp], newPageOptions[prop.nameProp], prop.keyForMerge);
        }
        else {
            return newPageOptions[prop.nameProp] || originalOption[prop.nameProp];
        }
    };
    /**
     * @private
     * @template T
     * @param {?} originalOptions
     * @param {?} newOptions
     * @param {?=} filterProp
     * @return {?}
     */
    PoPageCustomizationService.prototype.mergeOptions = /**
     * @private
     * @template T
     * @param {?} originalOptions
     * @param {?} newOptions
     * @param {?=} filterProp
     * @return {?}
     */
    function (originalOptions, newOptions, filterProp) {
        if (!originalOptions && !newOptions) {
            return;
        }
        if (!newOptions) {
            return originalOptions;
        }
        if (!originalOptions) {
            return newOptions;
        }
        if (originalOptions instanceof Array && newOptions instanceof Array) {
            return this.mergeOptionsArray(originalOptions, newOptions, filterProp);
        }
        return tslib_1.__assign({}, originalOptions, newOptions);
    };
    /**
     * @private
     * @template T
     * @param {?} originalOptions
     * @param {?} newOptions
     * @param {?} filterProp
     * @return {?}
     */
    PoPageCustomizationService.prototype.mergeOptionsArray = /**
     * @private
     * @template T
     * @param {?} originalOptions
     * @param {?} newOptions
     * @param {?} filterProp
     * @return {?}
     */
    function (originalOptions, newOptions, filterProp) {
        /** @type {?} */
        var deduplicateNewOptions = newOptions.filter((/**
         * @param {?} newItem
         * @return {?}
         */
        function (newItem) { return !originalOptions.find((/**
         * @param {?} originalItem
         * @return {?}
         */
        function (originalItem) { return originalItem[filterProp] === newItem[filterProp]; })); }));
        /** @type {?} */
        var mergedOriginalOptions = originalOptions.map((/**
         * @param {?} originalItem
         * @return {?}
         */
        function (originalItem) {
            /** @type {?} */
            var newItem = newOptions.find((/**
             * @param {?} newOptionsItem
             * @return {?}
             */
            function (newOptionsItem) { return originalItem[filterProp] === newOptionsItem[filterProp]; })) || originalItem;
            return tslib_1.__assign({}, originalItem, newItem);
        }));
        return tslib_1.__spread(mergedOriginalOptions, deduplicateNewOptions);
    };
    PoPageCustomizationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    PoPageCustomizationService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    /** @nocollapse */ PoPageCustomizationService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function PoPageCustomizationService_Factory() { return new PoPageCustomizationService(i0.ɵɵinject(i1.HttpClient)); }, token: PoPageCustomizationService, providedIn: "root" });
    return PoPageCustomizationService;
}());
export { PoPageCustomizationService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageCustomizationService.prototype.http;
}
