// @portinari/portinari-ui/esm2015/lib/components/po-toolbar/po-toolbar.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, ViewContainerRef } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { PoToolbarBaseComponent } from './po-toolbar-base.component';
/**
 * @docsExtends PoToolbarBaseComponent
 *
 * @example
 *
 * <example name="po-toolbar-basic" title="Portinari Toolbar Basic">
 *  <file name="sample-po-toolbar-basic/sample-po-toolbar-basic.component.html"> </file>
 *  <file name="sample-po-toolbar-basic/sample-po-toolbar-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-toolbar-labs" title="Portinari Toolbar Labs">
 *  <file name="sample-po-toolbar-labs/sample-po-toolbar-labs.component.html"> </file>
 *  <file name="sample-po-toolbar-labs/sample-po-toolbar-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-toolbar-logged" title="Portinari Toolbar - Logged">
 *  <file name="sample-po-toolbar-logged/sample-po-toolbar-logged.component.html"> </file>
 *  <file name="sample-po-toolbar-logged/sample-po-toolbar-logged.component.ts"> </file>
 * </example>
 */
let PoToolbarComponent = class PoToolbarComponent extends PoToolbarBaseComponent {
    constructor(titleService, viewRef) {
        super();
        this.titleService = titleService;
        this.parentRef = viewRef['_view']['component'];
    }
    ngOnInit() {
        this.setTitle(this.title);
    }
    setTitle(newTitle) {
        this.titleService.setTitle(newTitle);
    }
};
PoToolbarComponent.ctorParameters = () => [
    { type: Title },
    { type: ViewContainerRef }
];
PoToolbarComponent = tslib_1.__decorate([
    Component({
        selector: 'po-toolbar',
        template: "<div class=\"po-toolbar\">\n  <div class=\"po-toolbar-title\"> {{title}} </div>\n\n  <div class=\"po-toolbar-group-icon\">\n    <po-toolbar-actions *ngIf=\"actions && actions.length\"\n      [p-actions]=\"actions\"\n      [p-actions-icon]=\"actionsIcon\">\n    </po-toolbar-actions>\n\n    <po-toolbar-notification *ngIf=\"showNotification && notificationActions\"\n      [p-notification-actions]=\"notificationActions\"\n      [p-notification-number]=\"notificationNumber\">\n    </po-toolbar-notification>\n\n    <po-toolbar-profile *ngIf=\"isShowProfile\"\n      [p-profile]=\"profile\"\n      [p-profile-actions]=\"profileActions\">\n    </po-toolbar-profile>\n  </div>\n</div>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [Title, ViewContainerRef])
], PoToolbarComponent);
export { PoToolbarComponent };
