// @portinari/portinari-ui/esm2015/lib/components/po-accordion/po-accordion-item-header/po-accordion-item-header.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
let PoAccordionItemHeaderComponent = class PoAccordionItemHeaderComponent {
    constructor() {
        this.expanded = false;
        this.toggle = new EventEmitter();
    }
    onClick() {
        this.expanded = !this.expanded;
        this.toggle.emit(this.expanded);
    }
};
tslib_1.__decorate([
    Input('p-expanded'),
    tslib_1.__metadata("design:type", Boolean)
], PoAccordionItemHeaderComponent.prototype, "expanded", void 0);
tslib_1.__decorate([
    Input('p-label'),
    tslib_1.__metadata("design:type", String)
], PoAccordionItemHeaderComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Output('p-toggle'),
    tslib_1.__metadata("design:type", Object)
], PoAccordionItemHeaderComponent.prototype, "toggle", void 0);
PoAccordionItemHeaderComponent = tslib_1.__decorate([
    Component({
        selector: 'po-accordion-item-header',
        template: "<header class=\"po-accordion-item-header\">\n  <button class=\"po-accordion-item-header-button po-clickable\" (click)=\"onClick()\">\n    <div class=\"po-text-ellipsis po-accordion-item-header-title\">{{ label }}</div>\n    <span #icon class=\"po-icon po-accordion-item-header-icon po-icon-arrow-down\">\n    </span>\n  </button>\n</header>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], PoAccordionItemHeaderComponent);
export { PoAccordionItemHeaderComponent };
