// @portinari/portinari-templates/esm5/lib/components/po-page-dynamic-detail/po-page-dynamic-detail.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { Router, ActivatedRoute } from '@angular/router';
import * as util from '../../utils/util';
import { PoDialogService, PoNotificationService } from '@portinari/portinari-ui';
import { PoPageDynamicService } from '../../services/po-page-dynamic/po-page-dynamic.service';
/** @type {?} */
export var poPageDynamicDetailLiteralsDefault = {
    en: {
        pageActionEdit: 'Edit',
        pageActionRemove: 'Delete',
        pageActionBack: 'Back',
        confirmRemoveTitle: 'Confirm delete',
        confirmRemoveMessage: 'Are you sure you want to delete this record? You can not undo this action.',
        removeNotificationSuccess: 'Item deleted successfully.',
        registerNotFound: 'Register not found.'
    },
    es: {
        pageActionEdit: 'Editar',
        pageActionRemove: 'Borrar',
        pageActionBack: 'Regreso',
        confirmRemoveTitle: 'Confirmar la exclusión',
        confirmRemoveMessage: '¿Está seguro de que desea eliminar este registro? No puede deshacer esta acción.',
        removeNotificationSuccess: 'Elemento eliminado con éxito.',
        registerNotFound: 'Registro no encontrado.'
    },
    pt: {
        pageActionEdit: 'Editar',
        pageActionRemove: 'Excluir',
        pageActionBack: 'Voltar',
        confirmRemoveTitle: 'Confirmar exclusão',
        confirmRemoveMessage: 'Tem certeza de que deseja excluir esse registro? Você não poderá desfazer essa ação.',
        removeNotificationSuccess: 'Item excluído com sucesso.',
        registerNotFound: 'Registro não encontrado.'
    }
};
/**
 * \@description
 *
 * O `po-page-dynamic-detail` é uma página que serve para exibir registros em detalhes,
 * o mesmo também suporta metadados conforme especificado na documentação.
 *
 * \@example
 *
 * <example name="po-page-dynamic-detail-basic" title="Portinari Page Dynamic Detail Basic">
 *  <file name="sample-po-page-dynamic-detail-basic/sample-po-page-dynamic-detail-basic.component.html"> </file>
 *  <file name="sample-po-page-dynamic-detail-basic/sample-po-page-dynamic-detail-basic.component.ts"> </file>
 * </example>
 */
var PoPageDynamicDetailComponent = /** @class */ (function () {
    function PoPageDynamicDetailComponent(router, activatedRoute, poNotification, poDialogService, poPageDynamicService) {
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.poNotification = poNotification;
        this.poDialogService = poDialogService;
        this.poPageDynamicService = poPageDynamicService;
        this._actions = {};
        this._autoRouter = false;
        this._duplicates = [];
        this._fields = [];
        this._keys = [];
        this._pageActions = [];
        this.literals = tslib_1.__assign({}, poPageDynamicDetailLiteralsDefault[util.poLocaleDefault], poPageDynamicDetailLiteralsDefault[util.browserLanguage()]);
        this.model = {};
        /**
         * Objeto com propriedades do breadcrumb.
         */
        this.breadcrumb = { items: [] };
    }
    Object.defineProperty(PoPageDynamicDetailComponent.prototype, "actions", {
        /**
         * @optional
         *
         * @description
         *
         * Define as ações da página de acordo com a interface `PoPageDynamicDetailActions`.
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Define as ações da página de acordo com a interface `PoPageDynamicDetailActions`.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._actions = this.isObject(value) ? value : {};
            this._pageActions = this.getPageActions(this._actions);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicDetailComponent.prototype, "autoRouter", {
        get: /**
         * @return {?}
         */
        function () {
            return this._autoRouter;
        },
        /**
         * @todo Validar rotas na mão pois se existir uma rota '**' o catch do navigation não funciona.
         *
         * @optional
         *
         * @description
         *
         * Cria automaticamente as rotas de edição (novo/duplicate) e detalhes caso as ações
         * estejam definidas nas ações.
         *
         * > Para o correto funcionamento não pode haver nenhum rota coringa (`**`) especificada.
         *
         * @default false
         */
        set: /**
         * \@todo Validar rotas na mão pois se existir uma rota '**' o catch do navigation não funciona.
         *
         * \@optional
         *
         * \@description
         *
         * Cria automaticamente as rotas de edição (novo/duplicate) e detalhes caso as ações
         * estejam definidas nas ações.
         *
         * > Para o correto funcionamento não pode haver nenhum rota coringa (`**`) especificada.
         *
         * \@default false
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._autoRouter = util.convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicDetailComponent.prototype, "fields", {
        get: /**
         * @return {?}
         */
        function () {
            return this._fields;
        },
        /** Lista dos campos exibidos na página. */
        set: /**
         * Lista dos campos exibidos na página.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._fields = Array.isArray(value) ? tslib_1.__spread(value) : [];
            this._keys = this.getKeysByFields(this.fields);
            this._duplicates = this.getDuplicatesByFields(this.fields);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var paramId = this.activatedRoute.snapshot.params['id'];
        if (this.activatedRoute.snapshot.data.serviceApi) {
            this.serviceApi = this.activatedRoute.snapshot.data.serviceApi;
            this.poPageDynamicService.configServiceApi({ endpoint: this.serviceApi });
            this.loadMetadata(paramId);
        }
        else {
            this.poPageDynamicService.configServiceApi({ endpoint: this.serviceApi });
            this.loadData(paramId);
        }
    };
    Object.defineProperty(PoPageDynamicDetailComponent.prototype, "duplicates", {
        get: /**
         * @return {?}
         */
        function () {
            return tslib_1.__spread(this._duplicates);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicDetailComponent.prototype, "keys", {
        get: /**
         * @return {?}
         */
        function () {
            return tslib_1.__spread(this._keys);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicDetailComponent.prototype, "pageActions", {
        get: /**
         * @return {?}
         */
        function () {
            return tslib_1.__spread(this._pageActions);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.confirmRemove = /**
     * @private
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var confirmOptions = {
            title: this.literals.confirmRemoveTitle,
            message: this.literals.confirmRemoveMessage,
            confirm: this.remove.bind(this, path)
        };
        this.poDialogService.confirm(confirmOptions);
    };
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.formatUniqueKey = /**
     * @private
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var keys = util.mapObjectByProperties(item, this.keys);
        return util.valuesFromObject(keys).join('|');
    };
    /**
     * @private
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.goBack = /**
     * @private
     * @return {?}
     */
    function ( /*path*/) {
        window.history.back();
        // if (path) {
        //   this.navigateTo({ path, component: PoPageDynamicEditComponent });
        // } else {
        //   window.history.back();
        // }
    };
    /**
     * @private
     * @param {?} id
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.loadData = /**
     * @private
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        this.poPageDynamicService.getResource(id).toPromise().then((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            _this.model = response;
        })).catch((/**
         * @return {?}
         */
        function () {
            _this.model = undefined;
            _this.actions = undefined;
        }));
    };
    /**
     * @private
     * @param {?} id
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.loadMetadata = /**
     * @private
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        this.poPageDynamicService.getMetadata('detail').toPromise().then((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            _this.autoRouter = response.autoRouter;
            _this.actions = response.actions || {};
            _this.breadcrumb = response.breadcrumb || { items: [] };
            _this.fields = response.fields || [];
            _this.title = response.title;
            _this.loadData(id);
        }));
    };
    // @todo Validar rotas na mão pois se existir uma rota '**' o catch do navigation não funciona.
    // @todo Validar rotas na mão pois se existir uma rota '**' o catch do navigation não funciona.
    /**
     * @private
     * @param {?} route
     * @param {?=} forceStopAutoRouter
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.navigateTo = 
    // @todo Validar rotas na mão pois se existir uma rota '**' o catch do navigation não funciona.
    /**
     * @private
     * @param {?} route
     * @param {?=} forceStopAutoRouter
     * @return {?}
     */
    function (route, forceStopAutoRouter) {
        var _this = this;
        if (forceStopAutoRouter === void 0) { forceStopAutoRouter = false; }
        this.router.navigate([route.url || route.path], { queryParams: route.params })
            .catch((/**
         * @return {?}
         */
        function () {
            if (forceStopAutoRouter || !_this.autoRouter) {
                return;
            }
            _this.router.config.unshift((/** @type {?} */ ({
                path: route.path, component: route.component, data: { serviceApi: _this.serviceApi, autoRouter: true }
            })));
            _this.navigateTo(route, true);
        }));
    };
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.openEdit = /**
     * @private
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.resolveUrl(this.model, path);
        // this.navigateTo({ path, url, component: PoPageDynamicEditComponent });
        this.navigateTo({ path: path, url: url });
    };
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.remove = /**
     * @private
     * @param {?} path
     * @return {?}
     */
    function (path) {
        var _this = this;
        /** @type {?} */
        var uniqueKey = this.formatUniqueKey(this.model);
        this.poPageDynamicService.deleteResource(uniqueKey).toPromise().then((/**
         * @return {?}
         */
        function () {
            _this.poNotification.success(_this.literals.removeNotificationSuccess);
            _this.navigateTo({ path: path });
            // this.navigateTo({ path: path, component: PoPageDynamicTableComponent });
        }));
    };
    /**
     * @private
     * @param {?} item
     * @param {?} path
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.resolveUrl = /**
     * @private
     * @param {?} item
     * @param {?} path
     * @return {?}
     */
    function (item, path) {
        /** @type {?} */
        var uniqueKey = this.formatUniqueKey(item);
        return path.replace(/:id/g, uniqueKey);
    };
    /**
     * @private
     * @param {?=} actions
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.getPageActions = /**
     * @private
     * @param {?=} actions
     * @return {?}
     */
    function (actions) {
        if (actions === void 0) { actions = {}; }
        /** @type {?} */
        var pageActions = [];
        if (actions.edit) {
            pageActions.push({ label: this.literals.pageActionEdit, action: this.openEdit.bind(this, actions.edit) });
        }
        if (actions.remove) {
            pageActions.push({ label: this.literals.pageActionRemove, action: this.confirmRemove.bind(this, actions.remove) });
        }
        if (actions.back === undefined || actions.back) {
            pageActions.push({ label: this.literals.pageActionBack, action: this.goBack.bind(this, actions.back) });
        }
        return pageActions;
    };
    /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.getKeysByFields = /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    function (fields) {
        if (fields === void 0) { fields = []; }
        return fields.filter((/**
         * @param {?} field
         * @return {?}
         */
        function (field) { return field.key === true; })).map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) { return field.property; }));
    };
    /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.getDuplicatesByFields = /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    function (fields) {
        if (fields === void 0) { fields = []; }
        return fields.filter((/**
         * @param {?} field
         * @return {?}
         */
        function (field) { return field.duplicate === true; })).map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) { return field.property; }));
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    PoPageDynamicDetailComponent.prototype.isObject = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return !!value && typeof value === 'object' && !Array.isArray(value);
    };
    PoPageDynamicDetailComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-dynamic-detail',
                    template: "<po-page-default\r\n  [p-actions]=\"pageActions\"\r\n  [p-breadcrumb]=\"breadcrumb\"\r\n  [p-title]=\"title\">\r\n\r\n  <ng-container *ngIf=\"model; then viewFieldsTemplate; else registerNotFoundTemplate\">\r\n  </ng-container>\r\n\r\n</po-page-default>\r\n\r\n<ng-template #registerNotFoundTemplate>\r\n\r\n  <po-widget>\r\n    <span class=\"po-icon po-icon-info\"></span>\r\n    <span class=\"po-font-text-large\">\r\n      {{ literals.registerNotFound }}\r\n    </span>\r\n  </po-widget>\r\n\r\n</ng-template>\r\n\r\n<ng-template #viewFieldsTemplate>\r\n\r\n  <po-dynamic-view [p-fields]=\"fields\" [p-value]=\"model\">\r\n  </po-dynamic-view>\r\n\r\n</ng-template>\r\n",
                    providers: [PoPageDynamicService]
                }] }
    ];
    /** @nocollapse */
    PoPageDynamicDetailComponent.ctorParameters = function () { return [
        { type: Router },
        { type: ActivatedRoute },
        { type: PoNotificationService },
        { type: PoDialogService },
        { type: PoPageDynamicService }
    ]; };
    PoPageDynamicDetailComponent.propDecorators = {
        actions: [{ type: Input, args: ['p-actions',] }],
        autoRouter: [{ type: Input, args: ['p-auto-router',] }],
        breadcrumb: [{ type: Input, args: ['p-breadcrumb',] }],
        fields: [{ type: Input, args: ['p-fields',] }],
        title: [{ type: Input, args: ['p-title',] }],
        serviceApi: [{ type: Input, args: ['p-service-api',] }]
    };
    return PoPageDynamicDetailComponent;
}());
export { PoPageDynamicDetailComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype._actions;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype._autoRouter;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype._duplicates;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype._fields;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype._keys;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype._pageActions;
    /** @type {?} */
    PoPageDynamicDetailComponent.prototype.literals;
    /** @type {?} */
    PoPageDynamicDetailComponent.prototype.model;
    /**
     * Objeto com propriedades do breadcrumb.
     * @type {?}
     */
    PoPageDynamicDetailComponent.prototype.breadcrumb;
    /**
     * Título da página.
     * @type {?}
     */
    PoPageDynamicDetailComponent.prototype.title;
    /**
     * \@description
     *
     * Endpoint usado pelo template para requisição do recurso que serão exibido.
     *
     * Caso a ação `remove` estiver configurada, será feito uma requisição de exclusão nesse mesmo endpoint passando os campos
     * setados como `key: true`.
     *
     * > `DELETE {end-point}/{keys}`
     *
     * ```
     *  <po-page-dynamic-detail
     *    [p-actions]="{ remove: '/' }"
     *    [p-fields]="[ { property: 'id', key: true } ]"
     *    p-service="/api/po-samples/v1/people"
     *    ...>
     *  </po-page-dynamic-detail>
     * ```
     *
     * Resquisição disparada, onde a propriedade `id` é igual a 2:
     *
     * ```
     *  DELETE /api/po-samples/v1/people/2 HTTP/1.1
     *  Host: localhost:4000
     *  Connection: keep-alive
     *  Accept: application/json, text/plain
     *  ...
     * ```
     *
     * > Caso esteja usando metadados com o template, será disparado uma requisição na inicialização do template para buscar
     * > os metadados da página passando o tipo do metadado esperado e a versão cacheada pelo browser.
     * >
     * > `GET {end-point}/metadata?type=detail&version={version}`
     * @type {?}
     */
    PoPageDynamicDetailComponent.prototype.serviceApi;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype.poNotification;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype.poDialogService;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicDetailComponent.prototype.poPageDynamicService;
}
