// @portinari/portinari-templates/esm5/lib/components/po-page-dynamic-edit/po-page-dynamic.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { map } from 'rxjs/operators';
var PoPageDynamicService = /** @class */ (function () {
    function PoPageDynamicService(http) {
        this.http = http;
        this.endpoint = '/';
        this.headers = new HttpHeaders({
            'X-PORTINARI-SCREEN-LOCK': 'true'
        });
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    PoPageDynamicService.prototype.configServiceApi = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        this.endpoint = config.endpoint;
        this.metadata = config.metadata || this.metadata;
    };
    /**
     * @param {?=} type
     * @return {?}
     */
    PoPageDynamicService.prototype.getMetadata = /**
     * @param {?=} type
     * @return {?}
     */
    function (type) {
        if (type === void 0) { type = 'list'; }
        /** @type {?} */
        var key = this.endpoint + "-" + type + "-metadata";
        /** @type {?} */
        var cache = JSON.parse(localStorage.getItem(key)) || {};
        /** @type {?} */
        var metadataUrlBase = this.metadata || this.endpoint + "/metadata";
        /** @type {?} */
        var url = metadataUrlBase + "?type=" + type + "&version=" + (cache.version || '');
        return this.http.get(url).pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            if (response.version === cache.version) {
                return cache;
            }
            localStorage.setItem(key, JSON.stringify(response));
            return tslib_1.__assign({}, cache, response);
        })));
    };
    // Deleta um único recurso
    // Deleta um único recurso
    /**
     * @param {?} id
     * @return {?}
     */
    PoPageDynamicService.prototype.deleteResource = 
    // Deleta um único recurso
    /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.delete(this.endpoint + "/" + id, { headers: this.headers });
    };
    // Deleta recursos em lote
    // Deleta recursos em lote
    /**
     * @param {?} ids
     * @return {?}
     */
    PoPageDynamicService.prototype.deleteResources = 
    // Deleta recursos em lote
    /**
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        return this.http.request('delete', "" + this.endpoint, { headers: this.headers, body: ids });
    };
    // Busca uma lista de recursos
    // Busca uma lista de recursos
    /**
     * @param {?=} params
     * @return {?}
     */
    PoPageDynamicService.prototype.getResources = 
    // Busca uma lista de recursos
    /**
     * @param {?=} params
     * @return {?}
     */
    function (params) {
        if (params === void 0) { params = {}; }
        return this.http.get(this.endpoint, { headers: this.headers, params: params });
    };
    // Busca um único recurso
    // Busca um único recurso
    /**
     * @param {?} id
     * @return {?}
     */
    PoPageDynamicService.prototype.getResource = 
    // Busca um único recurso
    /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.get(this.endpoint + "/" + id, { headers: this.headers });
    };
    // Cria um recurso
    // Cria um recurso
    /**
     * @param {?} resource
     * @return {?}
     */
    PoPageDynamicService.prototype.createResource = 
    // Cria um recurso
    /**
     * @param {?} resource
     * @return {?}
     */
    function (resource) {
        return this.http.post("" + this.endpoint, resource, { headers: this.headers });
    };
    // Atualiza um recurso
    // Atualiza um recurso
    /**
     * @param {?} id
     * @param {?} resource
     * @return {?}
     */
    PoPageDynamicService.prototype.updateResource = 
    // Atualiza um recurso
    /**
     * @param {?} id
     * @param {?} resource
     * @return {?}
     */
    function (id, resource) {
        return this.http.put(this.endpoint + "/" + id, resource, { headers: this.headers });
    };
    PoPageDynamicService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PoPageDynamicService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return PoPageDynamicService;
}());
export { PoPageDynamicService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicService.prototype.endpoint;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicService.prototype.metadata;
    /** @type {?} */
    PoPageDynamicService.prototype.headers;
    /**
     * @type {?}
     * @private
     */
    PoPageDynamicService.prototype.http;
}
