// @portinari/portinari-ui/esm5/lib/components/po-field/po-lookup/po-lookup-modal/po-lookup-modal-base.component.js
import * as tslib_1 from "tslib";
import { EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { browserLanguage, isTypeof, poLocaleDefault } from '../../../../utils/util';
import { PoModalComponent } from '../../../../components/po-modal/po-modal.component';
import { PoTableColumnSortType } from '../../../po-table';
import { poTableLiteralsDefault } from '../../../po-table/po-table-base.component';
export var poLookupLiteralsDefault = {
    en: {
        modalPrimaryActionLabel: 'Select',
        modalSecondaryActionLabel: 'Cancel',
        modalPlaceholder: 'Search',
        modalTitle: 'Select a record',
        modalTableNoColumns: poTableLiteralsDefault.en.noColumns,
        modalTableNoData: poTableLiteralsDefault.en.noData,
        modalTableLoadingData: poTableLiteralsDefault.en.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.en.loadMoreData
    },
    es: {
        modalPrimaryActionLabel: 'Seleccionar',
        modalSecondaryActionLabel: 'Cancelar',
        modalPlaceholder: 'Buscar',
        modalTitle: 'Seleccione un registro',
        modalTableNoColumns: poTableLiteralsDefault.es.noColumns,
        modalTableNoData: poTableLiteralsDefault.es.noData,
        modalTableLoadingData: poTableLiteralsDefault.es.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.es.loadMoreData
    },
    pt: {
        modalPrimaryActionLabel: 'Selecionar',
        modalSecondaryActionLabel: 'Cancelar',
        modalPlaceholder: 'Pesquisar',
        modalTitle: 'Selecione um registro',
        modalTableNoColumns: poTableLiteralsDefault.pt.noColumns,
        modalTableNoData: poTableLiteralsDefault.pt.noData,
        modalTableLoadingData: poTableLiteralsDefault.pt.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.pt.loadMoreData
    },
    ru: {
        modalPrimaryActionLabel: 'выбирать',
        modalSecondaryActionLabel: 'отменить',
        modalPlaceholder: 'поиск',
        modalTitle: 'Выберите запись',
        modalTableNoColumns: poTableLiteralsDefault.ru.noColumns,
        modalTableNoData: poTableLiteralsDefault.ru.noData,
        modalTableLoadingData: poTableLiteralsDefault.ru.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.ru.loadMoreData
    }
};
/**
 * @docsPrivate
 *
 * Classe base do componente Po Lookup Modal.
 */
var PoLookupModalBaseComponent = /** @class */ (function () {
    function PoLookupModalBaseComponent() {
        var _this = this;
        this.hasNext = true;
        this.isLoading = false;
        this.page = 1;
        this.pageSize = 10;
        this.primaryAction = {
            action: function () {
                _this.items.forEach(function (element) {
                    if (element['$selected']) {
                        _this.model.emit(element);
                        _this.poModal.close();
                    }
                });
            },
            label: this.literals.modalPrimaryActionLabel
        };
        this.searchValue = '';
        this.secondaryAction = {
            action: function () {
                _this.model.emit(null);
                _this.poModal.close();
            },
            label: this.literals.modalSecondaryActionLabel
        };
        /** Evento utilizado ao selecionar um registro da tabela. */
        this.model = new EventEmitter();
    }
    Object.defineProperty(PoLookupModalBaseComponent.prototype, "literals", {
        get: function () {
            return this._literals || poLookupLiteralsDefault[browserLanguage()];
        },
        /** Objeto com as literais usadas no `po-lookup-modal`. */
        set: function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, poLookupLiteralsDefault[poLocaleDefault], poLookupLiteralsDefault[browserLanguage()], value);
                if (value.modalTitle) {
                    this.title = this.literals.modalTitle;
                }
            }
            else {
                this._literals = poLookupLiteralsDefault[browserLanguage()];
            }
            this.primaryAction.label = this.literals.modalPrimaryActionLabel;
            this.secondaryAction.label = this.literals.modalSecondaryActionLabel;
            this.setTableLiterals();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoLookupModalBaseComponent.prototype, "title", {
        get: function () {
            return this._title;
        },
        /** Título da modal. */
        set: function (value) {
            this._title = isTypeof(value, 'string') ? value : this.literals.modalTitle;
        },
        enumerable: true,
        configurable: true
    });
    PoLookupModalBaseComponent.prototype.ngOnDestroy = function () {
        if (this.filterSubscription) {
            this.filterSubscription.unsubscribe();
        }
        if (this.searchSubscription) {
            this.searchSubscription.unsubscribe();
        }
        if (this.showMoreSubscription) {
            this.showMoreSubscription.unsubscribe();
        }
    };
    PoLookupModalBaseComponent.prototype.ngOnInit = function () {
        this.initializeData();
        this.setTableLiterals();
    };
    PoLookupModalBaseComponent.prototype.search = function () {
        var _this = this;
        this.page = 1;
        if (this.searchValue) {
            this.isLoading = true;
            this.searchSubscription = this.getFilteredItems(this.searchValue).subscribe(function (data) {
                _this.items = data.items;
                _this.hasNext = data.hasNext;
                _this.isLoading = false;
            });
        }
        else {
            this.initializeData();
        }
    };
    PoLookupModalBaseComponent.prototype.showMoreEvent = function () {
        var _this = this;
        this.page++;
        this.isLoading = true;
        this.showMoreSubscription = this.getFilteredItems(this.searchValue).subscribe(function (data) {
            data.items.forEach(function (item) {
                _this.items.push(item);
            });
            _this.hasNext = data.hasNext;
            _this.isLoading = false;
        });
    };
    PoLookupModalBaseComponent.prototype.getFilteredItems = function (filter) {
        var _a = this, page = _a.page, pageSize = _a.pageSize, filterParams = _a.filterParams;
        if (this.filterService.getFilteredItems) {
            var filteredParams = this.getFilteredParams(filter);
            return this.filterService.getFilteredItems(filteredParams);
        }
        return this.filterService.getFilteredData(filter, page, pageSize, filterParams);
    };
    PoLookupModalBaseComponent.prototype.getFilteredParams = function (filter) {
        var _a = this, page = _a.page, pageSize = _a.pageSize, filterParams = _a.filterParams, sort = _a.sort;
        var filteredParams = {};
        var order = this.getOrderParam(sort);
        var params = { filter: filter, page: page, pageSize: pageSize, order: order, filterParams: filterParams };
        for (var key in params) {
            if (params.hasOwnProperty(key) && params[key]) {
                filteredParams[key] = params[key];
            }
        }
        return filteredParams;
    };
    PoLookupModalBaseComponent.prototype.getOrderParam = function (sort) {
        if (sort === void 0) { sort = { type: undefined }; }
        var column = sort.column, type = sort.type;
        if (!column) {
            return;
        }
        if (type === PoTableColumnSortType.Descending) {
            return "-" + column.property;
        }
        return "" + column.property;
    };
    PoLookupModalBaseComponent.prototype.initializeData = function () {
        var _this = this;
        this.isLoading = true;
        this.filterSubscription = this.getFilteredItems('').subscribe(function (data) {
            _this.items = data.items;
            _this.hasNext = data.hasNext;
            _this.isLoading = false;
        });
    };
    PoLookupModalBaseComponent.prototype.setTableLiterals = function () {
        this.tableLiterals = {
            'noColumns': this.literals.modalTableNoColumns,
            'noData': this.literals.modalTableNoData,
            'loadingData': this.literals.modalTableLoadingData,
            'loadMoreData': this.literals.modalTableLoadMoreData,
        };
    };
    tslib_1.__decorate([
        ViewChild(PoModalComponent, { static: true }),
        tslib_1.__metadata("design:type", PoModalComponent)
    ], PoLookupModalBaseComponent.prototype, "poModal", void 0);
    tslib_1.__decorate([
        Input('p-columns'),
        tslib_1.__metadata("design:type", Array)
    ], PoLookupModalBaseComponent.prototype, "columns", void 0);
    tslib_1.__decorate([
        Input('p-items'),
        tslib_1.__metadata("design:type", Array)
    ], PoLookupModalBaseComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoLookupModalBaseComponent.prototype, "literals", null);
    tslib_1.__decorate([
        Input('p-title'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoLookupModalBaseComponent.prototype, "title", null);
    tslib_1.__decorate([
        Input('p-filter-service'),
        tslib_1.__metadata("design:type", Object)
    ], PoLookupModalBaseComponent.prototype, "filterService", void 0);
    tslib_1.__decorate([
        Input('p-filter-params'),
        tslib_1.__metadata("design:type", Object)
    ], PoLookupModalBaseComponent.prototype, "filterParams", void 0);
    tslib_1.__decorate([
        Output('p-change-model'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoLookupModalBaseComponent.prototype, "model", void 0);
    return PoLookupModalBaseComponent;
}());
export { PoLookupModalBaseComponent };
