// @portinari/portinari-ui/esm5/lib/components/po-page/po-page-edit/po-page-edit-base.component.js
import * as tslib_1 from "tslib";
import { Input, ViewChild } from '@angular/core';
import { browserLanguage, poLocaleDefault } from './../../../utils/util';
import { PoPageContentComponent } from '../po-page-content/po-page-content.component';
export var poPageEditLiteralsDefault = {
    en: {
        cancel: 'Cancel',
        save: 'Save',
        saveNew: 'Save and New'
    },
    es: {
        cancel: 'Cancelar',
        save: 'Guardar',
        saveNew: 'Guardar y Nuevo'
    },
    pt: {
        cancel: 'Cancelar',
        save: 'Salvar',
        saveNew: 'Salvar e Novo'
    },
    ru: {
        cancel: 'отменить',
        save: 'экономить',
        saveNew: 'Сохранить и новый'
    }
};
/**
 * @description
 *
 * O componente **po-page-edit** é utilizado como container principal para tela de edição ou adição de um
 * registro.
 * Por padrão possui 3 ações, cada ação na tela executa uma função no componente que está utilizando o po-page-edit,
 * são elas:
 *  - Cancelar (função: cancel);
 *  - Salvar e Novo (função: saveNew);
 *  - Salvar (função: save);
 *
 * Caso não estiver implementado alguma função, listado anteriormente, o mesmo não será apresentado.
 *
 * Os botões "Salvar" e "Salvar e Novo" podem ser habilitados/desabilitados utilizando a propriedade p-disable-submit.
 * Esta propriedade pode ser utilizada para desabilitar os botões caso exista um formulário inválido na página ou alguma
 * regra de negócio não tenha sido atendida.
 */
var PoPageEditBaseComponent = /** @class */ (function () {
    function PoPageEditBaseComponent() {
    }
    Object.defineProperty(PoPageEditBaseComponent.prototype, "literals", {
        get: function () {
            return this._literals || poPageEditLiteralsDefault[browserLanguage()];
        },
        /**
         * @optional
         *
         * @description
         *
         * Objeto com as literais usadas no `po-page-edit`.
         *
         * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
         *
         * ```
         *  const customLiterals: PoPageEditLiterals = {
         *    cancel: 'Voltar',
         *    save: 'Confirmar',
         *    saveNew: 'Confirmar e criar um novo'
         *  };
         * ```
         *
         * Ou passando apenas as literais que deseja customizar:
         *
         * ```
         *  const customLiterals: PoPageEditLiterals = {
         *    cancel: 'Cancelar processo'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <po-page-edit
         *   [p-literals]="customLiterals">
         * </po-page-edit>
         * ```
         *
         *  > O objeto padrão de literais será traduzido de acordo com o idioma do browser (pt, en, es).
         */
        set: function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, poPageEditLiteralsDefault[poLocaleDefault], poPageEditLiteralsDefault[browserLanguage()], value);
            }
            else {
                this._literals = poPageEditLiteralsDefault[browserLanguage()];
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageEditBaseComponent.prototype, "title", {
        get: function () {
            return this._title;
        },
        /** Título da página. */
        set: function (title) {
            this._title = title;
            this.poPageContent.recalculateHeaderSize();
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        ViewChild(PoPageContentComponent, { static: true }),
        tslib_1.__metadata("design:type", PoPageContentComponent)
    ], PoPageEditBaseComponent.prototype, "poPageContent", void 0);
    tslib_1.__decorate([
        Input('p-breadcrumb'),
        tslib_1.__metadata("design:type", Object)
    ], PoPageEditBaseComponent.prototype, "breadcrumb", void 0);
    tslib_1.__decorate([
        Input('p-disable-submit'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoPageEditBaseComponent.prototype, "disableSubmit", void 0);
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoPageEditBaseComponent.prototype, "literals", null);
    tslib_1.__decorate([
        Input('p-title'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoPageEditBaseComponent.prototype, "title", null);
    return PoPageEditBaseComponent;
}());
export { PoPageEditBaseComponent };
