// @portinari/portinari-ui/esm5/lib/components/po-tabs/po-tab-button/po-tab-button.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, OnChanges, Output, SimpleChanges } from '@angular/core';
import { convertToBoolean } from '../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente responsável por manipular os botões de aba.
 */
var PoTabButtonComponent = /** @class */ (function () {
    function PoTabButtonComponent(elementRef) {
        this.elementRef = elementRef;
        // Função sera emitida quando a tab ficar ativada
        this.activated = new EventEmitter();
        // Função sera emitida quando a tab ficar desabilitada ou escondida
        this.changeState = new EventEmitter();
        // Método recebido do usuário para ser disparado quando clicar na aba
        this.click = new EventEmitter();
    }
    Object.defineProperty(PoTabButtonComponent.prototype, "active", {
        get: function () {
            return this._active;
        },
        // Ativa o botão
        set: function (value) {
            this._active = value;
            this.emitActivated();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTabButtonComponent.prototype, "hide", {
        get: function () {
            return this._hide;
        },
        // Oculta o botão
        set: function (value) {
            this._hide = convertToBoolean(value);
            this.setDisplayOnHide();
        },
        enumerable: true,
        configurable: true
    });
    PoTabButtonComponent.prototype.ngOnChanges = function (changes) {
        if ((changes.hide && changes.hide.currentValue) || (changes.disabled && changes.disabled.currentValue)) {
            this.changeState.emit(this);
        }
    };
    PoTabButtonComponent.prototype.onClick = function () {
        if (!this.disabled) {
            this.click.emit(this.id);
        }
    };
    PoTabButtonComponent.prototype.emitActivated = function () {
        if (this.active) {
            this.activated.emit(this);
        }
    };
    PoTabButtonComponent.prototype.setDisplayOnHide = function () {
        this.elementRef.nativeElement.style.display = this.hide ? 'none' : '';
    };
    PoTabButtonComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    tslib_1.__decorate([
        Input('p-active'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoTabButtonComponent.prototype, "active", null);
    tslib_1.__decorate([
        Input('p-disabled'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoTabButtonComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        Input('p-hide'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoTabButtonComponent.prototype, "hide", null);
    tslib_1.__decorate([
        Input('p-id'),
        tslib_1.__metadata("design:type", String)
    ], PoTabButtonComponent.prototype, "id", void 0);
    tslib_1.__decorate([
        Input('p-label'),
        tslib_1.__metadata("design:type", String)
    ], PoTabButtonComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input('p-small'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoTabButtonComponent.prototype, "small", void 0);
    tslib_1.__decorate([
        Output('p-activated'),
        tslib_1.__metadata("design:type", Object)
    ], PoTabButtonComponent.prototype, "activated", void 0);
    tslib_1.__decorate([
        Output('p-change-state'),
        tslib_1.__metadata("design:type", Object)
    ], PoTabButtonComponent.prototype, "changeState", void 0);
    tslib_1.__decorate([
        Output('p-click'),
        tslib_1.__metadata("design:type", Object)
    ], PoTabButtonComponent.prototype, "click", void 0);
    PoTabButtonComponent = tslib_1.__decorate([
        Component({
            selector: 'po-tab-button',
            template: "<div\n  class=\"po-tab-button-md\"\n  [attr.tabindex]=\"disabled ? null : 0\"\n  [class.po-tab-button-active]=\"active\"\n  [class.po-tab-button-disabled]=\"disabled\"\n  [class.po-tab-button-sm]=\"small\"\n  (click)=\"onClick()\"\n  (keyup.enter)=\"onClick()\">\n\n  <span class=\"po-tab-button-label po-text-ellipsis\">\n    {{label}}\n  </span>\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef])
    ], PoTabButtonComponent);
    return PoTabButtonComponent;
}());
export { PoTabButtonComponent };
