// @portinari/portinari-ui/esm5/lib/components/po-dropdown/po-dropdown.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, Renderer2, ViewChild } from '@angular/core';
import { isKeyCodeEnter } from './../../utils/util';
import { PoDropdownBaseComponent } from './po-dropdown-base.component';
/**
 * @docsExtends PoDropdownBaseComponent
 *
 * @example
 *
 * <example name="po-dropdown-basic" title="Portinari Dropdown Basic" >
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.html"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.ts"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-dropdown-labs" title="Portinari Dropdown Labs" >
 *  <file name="sample-po-dropdown-labs/sample-po-dropdown-labs.component.html"> </file>
 *  <file name="sample-po-dropdown-labs/sample-po-dropdown-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-dropdown-social-network" title="Portinari Dropdown - Social Network" >
 *  <file name="sample-po-dropdown-social-network/sample-po-dropdown-social-network.component.html"> </file>
 *  <file name="sample-po-dropdown-social-network/sample-po-dropdown-social-network.component.ts"> </file>
 * </example>
 */
var PoDropdownComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoDropdownComponent, _super);
    function PoDropdownComponent(renderer) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.onScroll = function () {
            if (_this.open) {
                _this.hideDropdown();
            }
        };
        return _this;
    }
    PoDropdownComponent.prototype.onKeyDown = function (event) {
        if (isKeyCodeEnter(event)) {
            this.toggleDropdown();
        }
    };
    PoDropdownComponent.prototype.toggleDropdown = function () {
        this.dropdownRef && !this.open && !this.disabled ? this.showDropdown() : this.hideDropdown();
    };
    PoDropdownComponent.prototype.checkClickArea = function (event) {
        return this.dropdownRef && this.dropdownRef.nativeElement.contains(event.target);
    };
    PoDropdownComponent.prototype.hideDropdown = function () {
        this.icon = 'po-icon-arrow-down';
        this.removeListeners();
        this.popupRef.close();
        this.open = false;
    };
    PoDropdownComponent.prototype.initializeListeners = function () {
        var _this = this;
        this.clickoutListener = this.renderer.listen('document', 'click', function (event) {
            _this.wasClickedOnDropdown(event);
        });
        this.resizeListener = this.renderer.listen('window', 'resize', function () {
            _this.hideDropdown();
        });
        window.addEventListener('scroll', this.onScroll, true);
    };
    PoDropdownComponent.prototype.removeListeners = function () {
        if (this.clickoutListener) {
            this.clickoutListener();
        }
        if (this.resizeListener) {
            this.resizeListener();
        }
        window.removeEventListener('scroll', this.onScroll, true);
    };
    PoDropdownComponent.prototype.showDropdown = function () {
        this.icon = 'po-icon-arrow-up';
        this.initializeListeners();
        this.popupRef.open();
        this.open = true;
    };
    PoDropdownComponent.prototype.wasClickedOnDropdown = function (event) {
        var clickedOnDropdown = this.checkClickArea(event);
        if (!clickedOnDropdown) {
            this.hideDropdown();
        }
    };
    PoDropdownComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    tslib_1.__decorate([
        ViewChild('dropdownRef', { read: ElementRef, static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoDropdownComponent.prototype, "dropdownRef", void 0);
    tslib_1.__decorate([
        ViewChild('popupRef', { static: false }),
        tslib_1.__metadata("design:type", Object)
    ], PoDropdownComponent.prototype, "popupRef", void 0);
    PoDropdownComponent = tslib_1.__decorate([
        Component({
            selector: 'po-dropdown',
            template: "<div #dropdownRef\n  class=\"po-dropdown\"\n  tabindex=\"{{ disabled ? -1 : 0 }}\"\n  (click)=\"toggleDropdown()\"\n  (keydown)=\"onKeyDown($event)\">\n\n  <div class=\"po-dropdown-button\" [ngClass]=\"{'po-dropdown-button-disabled' : disabled}\">\n    {{ label }}\n    <div class=\"po-dropdown-icon po-icon {{icon}}\"></div>\n  </div>\n\n  <po-popup #popupRef\n    p-hide-arrow\n    p-is-corner-align\n    p-position=\"bottom-left\"\n    [p-actions]=\"actions\"\n    [p-custom-positions]=\"['bottom-left', 'top-left']\"\n    [p-target]=\"dropdownRef\">\n  </po-popup>\n\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2])
    ], PoDropdownComponent);
    return PoDropdownComponent;
}(PoDropdownBaseComponent));
export { PoDropdownComponent };
