// @portinari/portinari-ui/esm5/lib/services/services.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { PoColorPaletteModule } from './po-color-palette/po-color-palette.module';
import { PoComponentInjectorModule } from './po-component-injector/po-component-injector.module';
import { PoControlPositionModule } from './po-control-position/po-control-position.module';
import { PoDateTimeModule } from './po-date/po-date.module';
import { PoDialogModule } from './po-dialog/po-dialog.module';
import { PoI18nPipe } from './po-i18n/po-i18n.pipe';
import { PoLanguageModule } from './po-language/po-language.module';
import { PoNotificationModule } from './po-notification/po-notification.module';
var PoServicesModule = /** @class */ (function () {
    function PoServicesModule() {
    }
    PoServicesModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                PoI18nPipe
            ],
            imports: [
                PoColorPaletteModule,
                PoComponentInjectorModule,
                PoControlPositionModule,
                PoDateTimeModule,
                PoDialogModule,
                PoLanguageModule,
                PoNotificationModule
            ],
            exports: [
                PoColorPaletteModule,
                PoComponentInjectorModule,
                PoControlPositionModule,
                PoDateTimeModule,
                PoDialogModule,
                PoI18nPipe,
                PoNotificationModule
            ],
            providers: [],
            bootstrap: []
        })
    ], PoServicesModule);
    return PoServicesModule;
}());
export { PoServicesModule };
