// @portinari/portinari-ui/esm5/lib/components/po-page/po-page-content/po-page-content.component.js
import * as tslib_1 from "tslib";
import { AfterViewInit, Component, OnDestroy, Renderer2 } from '@angular/core';
import { PoPageContentBaseComponent } from './po-page-content-base.component';
/**
 * @docsPrivate
 *
 * @docsExtends PoPageContentBaseComponent
 */
var PoPageContentComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoPageContentComponent, _super);
    function PoPageContentComponent(renderer) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.contentOpacity = 0;
        _this.height = '90%';
        _this.overflowY = 'none';
        _this.initializeListeners();
        return _this;
    }
    PoPageContentComponent.prototype.ngAfterViewInit = function () {
        this.recalculateHeaderSize();
    };
    PoPageContentComponent.prototype.ngOnDestroy = function () {
        this.removeListeners();
    };
    PoPageContentComponent.prototype.recalculateHeaderSize = function () {
        var _this = this;
        setTimeout(function () {
            var pageHeaderElement = document.querySelector('div.po-page-header');
            _this.setHeightContent(pageHeaderElement);
            _this.contentOpacity = 1;
        });
    };
    PoPageContentComponent.prototype.setHeightContent = function (poPageHeader) {
        var bodyHeight = document.body.clientHeight;
        var pageHeaderHeight = poPageHeader ? poPageHeader.offsetTop + poPageHeader.offsetHeight : 0;
        var newHeight = bodyHeight - pageHeaderHeight;
        this.height = newHeight + "px";
    };
    PoPageContentComponent.prototype.initializeListeners = function () {
        var _this = this;
        this.resizeListener = this.renderer.listen('window', 'resize', function () {
            _this.recalculateHeaderSize();
        });
    };
    PoPageContentComponent.prototype.removeListeners = function () {
        this.resizeListener();
    };
    PoPageContentComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    PoPageContentComponent = tslib_1.__decorate([
        Component({
            selector: 'po-page-content',
            template: "<div class=\"po-page-content\"\n  [style.height]=\"height\"\n  [style.opacity]=\"contentOpacity\"\n  [style.overflow-y]=\"overflowY\">\n  <ng-content></ng-content>\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2])
    ], PoPageContentComponent);
    return PoPageContentComponent;
}(PoPageContentBaseComponent));
export { PoPageContentComponent };
