// @portinari/portinari-ui/esm5/lib/components/po-tabs/po-tab-dropdown/po-tab-dropdown.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { PoPopoverComponent } from '../../po-popover/po-popover.component';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para agrupamento de componentes `po-tab-button` que os rendereiza em uma um caixa de diálogo
 * no formato de lista.
 */
var PoTabDropdownComponent = /** @class */ (function () {
    function PoTabDropdownComponent() {
        // Lista de abas
        this.tabs = [];
        // Evento que será emitido ao ativar uma aba
        this.activated = new EventEmitter();
        // Evento que será emitido a aba for desabilitada ou ocultada
        this.changeState = new EventEmitter();
        // Evento de click
        this.click = new EventEmitter();
    }
    PoTabDropdownComponent.prototype.getContainerClass = function () {
        return this.small ? 'po-tab-dropdown-container-sm' : 'po-tab-dropdown-container';
    };
    PoTabDropdownComponent.prototype.togglePopover = function () {
        this.popover.isHidden ? this.popover.open() : this.popover.close();
    };
    tslib_1.__decorate([
        ViewChild('popover', { static: true }),
        tslib_1.__metadata("design:type", PoPopoverComponent)
    ], PoTabDropdownComponent.prototype, "popover", void 0);
    tslib_1.__decorate([
        ViewChild('tabDrop', { read: ElementRef, static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoTabDropdownComponent.prototype, "tabDropdown", void 0);
    tslib_1.__decorate([
        Input('p-label'),
        tslib_1.__metadata("design:type", String)
    ], PoTabDropdownComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input('p-small'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoTabDropdownComponent.prototype, "small", void 0);
    tslib_1.__decorate([
        Input('p-tabs'),
        tslib_1.__metadata("design:type", Array)
    ], PoTabDropdownComponent.prototype, "tabs", void 0);
    tslib_1.__decorate([
        Output('p-activated'),
        tslib_1.__metadata("design:type", Object)
    ], PoTabDropdownComponent.prototype, "activated", void 0);
    tslib_1.__decorate([
        Output('p-change-state'),
        tslib_1.__metadata("design:type", Object)
    ], PoTabDropdownComponent.prototype, "changeState", void 0);
    tslib_1.__decorate([
        Output('p-click'),
        tslib_1.__metadata("design:type", Object)
    ], PoTabDropdownComponent.prototype, "click", void 0);
    PoTabDropdownComponent = tslib_1.__decorate([
        Component({
            selector: 'po-tab-dropdown',
            template: "<div #tabDrop\n  class=\"po-tab-dropdown-content po-tab-button-md\"\n  [attr.tabindex]=\"0\"\n  [class.po-tab-button-sm]=\"small\"\n  (keyup.enter)=\"togglePopover()\">\n  <div class=\"po-tab-button-label\">\n    {{ label }}\n    <span class=\"po-icon\" [ngClass]=\"(popover.isHidden ? 'po-icon-arrow-down' : 'po-icon-arrow-up')\">\n    </span>\n  </div>\n</div>\n\n<po-popover #popover\n  p-hide-arrow\n  p-position=\"bottom\"\n  [p-target]=\"tabDropdown\">\n\n  <div [ngClass]=\"getContainerClass()\">\n    <po-tab-button *ngFor=\"let tab of tabs\"\n      class=\"po-tab-dropdown-button\"\n      [class.po-tab-dropdown-button-active]=\"tab.active\"\n      [p-active]=\"tab.active\"\n      [p-disabled]=\"tab.disabled\"\n      [p-hide]=\"tab.hide\"\n      [p-id]=\"tab.id\"\n      [p-label]=\"tab.label\"\n      [p-small]=\"small\"\n      (p-activated)=\"activated.emit(tab)\"\n      (p-change-state)=\"changeState.emit(tab)\"\n      (p-click)=\"popover.close();click.emit(tab);\">\n    </po-tab-button>\n  </div>\n\n</po-popover>\n"
        })
    ], PoTabDropdownComponent);
    return PoTabDropdownComponent;
}());
export { PoTabDropdownComponent };
