// @portinari/portinari-ui/esm5/lib/components/po-table/po-table-detail/po-table-detail.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { DecimalPipe } from '@angular/common';
import { capitalizeFirstLetter, isTypeof } from '../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente responsável por apresentar o detalhe de cada linha da tabela.
 */
var PoTableDetailComponent = /** @class */ (function () {
    function PoTableDetailComponent(decimalPipe) {
        this.decimalPipe = decimalPipe;
        /**
         * Define se a tabela possui a opção de `selectable` habilitada.
         */
        this.isSelectable = false;
        /**
         * Ação executada ao selecionar ou desmarcar a seleção de uma linha de detalhe do `po-table`.
         */
        this.selectRow = new EventEmitter();
    }
    Object.defineProperty(PoTableDetailComponent.prototype, "detail", {
        get: function () {
            return this._detail;
        },
        /**
         * Configuração da linha de detalhes.
         */
        set: function (value) {
            this._detail = this.returnPoTableDetailObject(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableDetailComponent.prototype, "detailColumns", {
        get: function () {
            return this.detail && this.detail.columns || [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableDetailComponent.prototype, "typeHeaderInline", {
        get: function () {
            return this.detail && !this.detail['typeHeader'] || this.detail['typeHeader'] === 'inline';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableDetailComponent.prototype, "typeHeaderTop", {
        get: function () {
            return this.detail && this.detail['typeHeader'] === 'top';
        },
        enumerable: true,
        configurable: true
    });
    PoTableDetailComponent.prototype.formatNumberDetail = function (value, format) {
        if (!format) {
            return value;
        }
        return this.decimalPipe.transform(value, format);
    };
    PoTableDetailComponent.prototype.getColumnTitleLabel = function (detail) {
        return detail.label || capitalizeFirstLetter(detail.property);
    };
    PoTableDetailComponent.prototype.onSelectRow = function (item) {
        item.$selected = !item.$selected;
        this.selectRow.emit(item);
    };
    PoTableDetailComponent.prototype.returnPoTableDetailObject = function (value) {
        if (value && isTypeof(value, 'object')) {
            if (value.columns) {
                value.columns.forEach(function (column) { return column.property = column.property || column.column; });
            }
            if (Array.isArray(value)) {
                return { columns: value };
            }
            if (value.columns) {
                return value;
            }
        }
    };
    PoTableDetailComponent.ctorParameters = function () { return [
        { type: DecimalPipe }
    ]; };
    tslib_1.__decorate([
        Input('p-detail'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoTableDetailComponent.prototype, "detail", null);
    tslib_1.__decorate([
        Input('p-items'),
        tslib_1.__metadata("design:type", Array)
    ], PoTableDetailComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        Input('p-selectable'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoTableDetailComponent.prototype, "isSelectable", void 0);
    tslib_1.__decorate([
        Output('p-select-row'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoTableDetailComponent.prototype, "selectRow", void 0);
    PoTableDetailComponent = tslib_1.__decorate([
        Component({
            selector: 'po-table-detail',
            template: "<table class=\"po-table-master-detail\">\n  <thead *ngIf=\"typeHeaderTop\">\n    <tr>\n      <th class=\"po-table-header po-table-column-selectable\" *ngIf=\"isSelectable\"></th>\n      <th class=\"po-table-header po-table-header-column po-table-header-master-detail\"></th>\n      <th class=\"po-table-header po-table-header-ellipsis\" *ngFor=\"let detail of detail.columns\">\n        {{ getColumnTitleLabel(detail) }}\n      </th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr class=\"po-table-detail-row\"\n        [class.po-table-row-active]=\"item.$selected && isSelectable\"\n        *ngFor=\"let item of items\">\n\n      <ng-container *ngIf=\"isSelectable; else masterDetailSpace\">\n\n        <td class=\"po-table-column-master-detail-space-checkbox\"></td>\n        <td class=\"po-table-column po-table-column-selectable\">\n          <input\n            class=\"po-table-checkbox\"\n            type=\"checkbox\"\n            [class.po-table-checkbox-checked]=\"item.$selected\">\n          <label class=\"po-table-checkbox-label po-clickable\" (click)=\"onSelectRow(item)\">\n          </label>\n        </td>\n\n      </ng-container>\n\n      <ng-template #masterDetailSpace>\n        <td class=\"po-table-column-master-detail-space\"></td>\n      </ng-template>\n\n      <td class=\"po-table-column-master-detail po-table-master-detail-label\"\n        (click)=\"isSelectable ? onSelectRow(item) : 'javascript:;'\"\n        *ngFor=\"let detail of detailColumns\">\n        <strong *ngIf=\"typeHeaderInline\"> {{ getColumnTitleLabel(detail) }}: </strong>\n\n        <ng-container [ngSwitch]=\"detail.type\">\n          <span *ngSwitchCase=\"'currency'\">{{ item[detail.property] | currency: detail.format:'symbol':'1.2-2' }}</span>\n          <span *ngSwitchCase=\"'date'\">{{ item[detail.property] | date: detail.format || 'dd/MM/yyyy' }}</span>\n          <span *ngSwitchCase=\"'time'\">{{ item[detail.property] | po_time: detail.format || 'HH:mm:ss.ffffff' }}</span>\n          <span *ngSwitchCase=\"'dateTime'\">{{ item[detail.property] | date: detail.format || 'dd/MM/yyyy HH:mm:ss' }}</span>\n          <span *ngSwitchCase=\"'number'\">{{ formatNumberDetail(item[detail.property], detail.format) }}</span>\n          <span *ngSwitchDefault>{{ item[detail.property] }}</span>\n        </ng-container>\n      </td>\n    </tr>\n  </tbody>\n</table>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [DecimalPipe])
    ], PoTableDetailComponent);
    return PoTableDetailComponent;
}());
export { PoTableDetailComponent };
