// @portinari/portinari-ui/esm5/lib/components/po-stepper/po-step/po-step.component.js
import * as tslib_1 from "tslib";
import { AfterContentInit, Component, ElementRef, Input } from '@angular/core';
import { uuid } from '../../../utils/util';
import { PoStepperStatus } from '../enums/po-stepper-status.enum';
/**
 * @description
 *
 * O componente `po-step` é utilizado para envolver e renderizar o conteúdo de um passo (*step*) do `po-stepper`, por exemplo:
 *
 * ```
 * <po-stepper>
 *
 *    <po-step p-label="Endereço">
 *
 *        <!-- Conteúdo referente ao endereço -->
 *
 *    </po-step>
 *
 *    <po-step p-label="Pagamento">
 *
 *        <!-- Conteúdo referente ao pagamento -->
 *
 *    </po-step>
 *
 *  </po-stepper>
 * ```
 *
 * A renderização do conteúdo envolvido na tela e o controle dos status são feitos automaticamente. No qual, o primeiro
 * `po-step` encontrado será colocado como ativo, o próximo fica com o status *default* e os demais ficam
 * desabilitados (*disabled*).
 *
 * Ao clicar no `po-step` que está com o status *default*, o que está ativo ficará com o
 * status de concluído (*done*) e o próximo que estava desabilitado ficará com o status *default* e o restante permanecerá
 * com o status desabilitado.
 *
 * > Ao utilizar o `po-step`, o componente `po-stepper` funcionará de forma sequencial, ou seja, não será possível
 * pular para outro `po-step` que esteja com o status igual a desabilitado (*disabled*).
 *
 * Acesse a [documentação do `po-stepper`](/documentation/po-stepper) para ter mais informações sobre o seu funcionamento
 * e exemplos de uso.
 */
var PoStepComponent = /** @class */ (function () {
    function PoStepComponent(elementRef) {
        this.elementRef = elementRef;
        this._status = PoStepperStatus.Disabled;
        // ID do step
        this.id = uuid();
    }
    Object.defineProperty(PoStepComponent.prototype, "status", {
        get: function () {
            return this._status;
        },
        // Estado de exibição do `po-step`.
        set: function (status) {
            this._status = status;
            this.setDisplayOnActiveOrError();
        },
        enumerable: true,
        configurable: true
    });
    PoStepComponent.prototype.ngAfterContentInit = function () {
        this.setDisplayOnActiveOrError();
    };
    PoStepComponent.prototype.setDisplayOnActiveOrError = function () {
        this.elementRef.nativeElement.style.display =
            (this.status === PoStepperStatus.Active || this.status === PoStepperStatus.Error) ? '' : 'none';
    };
    PoStepComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    tslib_1.__decorate([
        Input('p-can-active-next-step'),
        tslib_1.__metadata("design:type", Function)
    ], PoStepComponent.prototype, "canActiveNextStep", void 0);
    tslib_1.__decorate([
        Input('p-label'),
        tslib_1.__metadata("design:type", String)
    ], PoStepComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input('p-status'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoStepComponent.prototype, "status", null);
    PoStepComponent = tslib_1.__decorate([
        Component({
            selector: 'po-step',
            template: "<ng-content></ng-content>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef])
    ], PoStepComponent);
    return PoStepComponent;
}());
export { PoStepComponent };
