// @portinari/portinari-ui/esm2015/lib/components/po-field/po-lookup/po-lookup-modal/po-lookup-modal.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, ViewChild } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime, filter } from 'rxjs/operators';
import { PoLookupModalBaseComponent } from '../po-lookup-modal/po-lookup-modal-base.component';
/**
 * @docsPrivate
 *
 * @docsExtends PoLookupModalBaseComponent
 */
let PoLookupModalComponent = class PoLookupModalComponent extends PoLookupModalBaseComponent {
    /**
     * @docsPrivate
     *
     * @docsExtends PoLookupModalBaseComponent
     */
    constructor() {
        super(...arguments);
        this.keyUpObservable = null;
        this.containerHeight = 375;
        this.tableHeight = 370;
    }
    ngOnInit() {
        super.ngOnInit();
        this.initializeEventInput();
        this.setTableHeight();
    }
    initializeEventInput() {
        this.keyUpObservable = fromEvent(this.inputSearchEl.nativeElement, 'keyup')
            .pipe(filter((e) => this.validateEnterPressed(e)), debounceTime(400));
        this.keyUpObservable.subscribe(() => {
            this.search();
        });
    }
    openModal() {
        this.poModal.open();
    }
    sortBy(sort) {
        this.sort = sort;
    }
    setTableHeight() {
        if (window.innerHeight < 615) {
            this.tableHeight -= 50;
            this.containerHeight -= 50;
        }
    }
    validateEnterPressed(e) {
        return e.keyCode === 13;
    }
};
tslib_1.__decorate([
    ViewChild('inpsearch', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], PoLookupModalComponent.prototype, "inputSearchEl", void 0);
PoLookupModalComponent = tslib_1.__decorate([
    Component({
        selector: 'po-lookup-modal',
        template: "<po-modal\n  p-click-out=\"false\"\n  p-hide-close=\"false\"\n  p-size=\"lg\"\n  [p-primary-action]=\"primaryAction\"\n  [p-secondary-action]=\"secondaryAction\"\n  [p-title]=\"title\">\n\n  <po-field-container\n    class=\"po-lookup-header po-md-6 po-pull-right\"\n    [p-optional]=\"false\">\n\n    <div class=\"po-field-container-content\">\n      <input #inpsearch\n        class=\"po-input po-input-icon-right\"\n        name=\"contentSearch\"\n        [(ngModel)]=\"searchValue\"\n        [placeholder]=\"literals.modalPlaceholder\"\n        type=\"text\">\n\n      <div class=\"po-field-icon-container-right\">\n        <span #iconLookup\n          class=\"po-icon po-field-icon po-icon-search\"\n          (click)=\"search()\">\n        </span>\n      </div>\n    </div>\n\n  </po-field-container>\n\n  <div class=\"po-row po-lookup-container-table\" [style.height.px]=\"containerHeight\">\n\n    <po-table\n      class=\"po-md-12\"\n      p-checkbox=\"true\"\n      p-hide-detail=\"true\"\n      p-single-select=\"true\"\n      p-sort=\"true\"\n      [p-columns]=\"columns\"\n      [p-height]=\"tableHeight\"\n      [p-items]=\"items\"\n      [p-literals]=\"tableLiterals\"\n      [p-loading]=\"isLoading\"\n      [p-show-more-disabled]=\"!hasNext\"\n      (p-show-more)=\"showMoreEvent()\"\n      (p-sort-by)=\"sortBy($event)\">\n    </po-table>\n\n  </div>\n\n</po-modal>\n"
    })
], PoLookupModalComponent);
export { PoLookupModalComponent };
