// @portinari/portinari-ui/esm5/lib/components/po-tree-view/po-tree-view-item/po-tree-view-item.component.js
import * as tslib_1 from "tslib";
import { animate, state, style, transition, trigger } from '@angular/animations';
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { PoTreeViewService } from '../services/po-tree-view.service';
var PoTreeViewItemComponent = /** @class */ (function () {
    function PoTreeViewItemComponent(treeViewService) {
        this.treeViewService = treeViewService;
    }
    Object.defineProperty(PoTreeViewItemComponent.prototype, "hasSubItems", {
        get: function () {
            return !!(this.item.subItems && this.item.subItems.length);
        },
        enumerable: true,
        configurable: true
    });
    PoTreeViewItemComponent.prototype.onClick = function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.item.expanded = !this.item.expanded;
        this.treeViewService.emitExpandedEvent(tslib_1.__assign({}, this.item));
    };
    PoTreeViewItemComponent.prototype.onSelect = function (selectedItem) {
        this.treeViewService.emitSelectedEvent(tslib_1.__assign({}, selectedItem));
    };
    PoTreeViewItemComponent.prototype.trackByFunction = function (index) {
        return index;
    };
    PoTreeViewItemComponent.ctorParameters = function () { return [
        { type: PoTreeViewService }
    ]; };
    tslib_1.__decorate([
        Input('p-item'),
        tslib_1.__metadata("design:type", Object)
    ], PoTreeViewItemComponent.prototype, "item", void 0);
    tslib_1.__decorate([
        Input('p-selectable'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoTreeViewItemComponent.prototype, "selectable", void 0);
    PoTreeViewItemComponent = tslib_1.__decorate([
        Component({
            selector: 'po-tree-view-item',
            template: "<li class=\"po-tree-view-item\">\n\n  <po-tree-view-item-header\n    [p-item]=\"item\"\n    [p-selectable]=\"selectable\"\n    (p-expanded)=\"onClick($event)\"\n    (p-selected)=\"onSelect(item)\">\n  </po-tree-view-item-header>\n\n  <ul *ngIf=\"hasSubItems\"\n    class=\"po-tree-view-item-group\"\n    [@toggleBody]=\"item.expanded ? 'expanded' : 'collapsed'\">\n\n    <po-tree-view-item *ngFor=\"let subItem of item.subItems; trackBy: trackByFunction\"\n      [p-item]=\"subItem\"\n      [p-selectable]=\"selectable\">\n    </po-tree-view-item>\n  </ul>\n\n</li>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [
                trigger('toggleBody', [
                    state('collapsed', style({
                        'overflow-y': 'hidden',
                        visibility: 'hidden',
                        opacity: 0,
                        height: '0'
                    })),
                    transition('expanded => collapsed', [
                        style({ height: '*' }),
                        animate(100, style({ opacity: 0 })),
                        animate(200, style({ height: 0 }))
                    ]),
                    transition('collapsed => expanded', [
                        style({ height: '0' }),
                        animate(100, style({ opacity: 1 })),
                        animate(200, style({ height: '*' }))
                    ])
                ])
            ]
        }),
        tslib_1.__metadata("design:paramtypes", [PoTreeViewService])
    ], PoTreeViewItemComponent);
    return PoTreeViewItemComponent;
}());
export { PoTreeViewItemComponent };
