// @portinari/portinari-templates/esm5/lib/components/po-page-dynamic-search/po-advanced-filter/po-advanced-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { PoDynamicFormComponent, PoLanguageService } from '@portinari/portinari-ui';
import { PoAdvancedFilterBaseComponent } from './po-advanced-filter-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends PoAdvancedFilterBaseComponent
 *
 * \@examplePrivate
 *
 * <example-private name="po-advanced-filter" title="Portinari Busca Avançada">
 *   <file name="sample-po-advanced-filter.component.html"> </file>
 *   <file name="sample-po-advanced-filter.component.ts"> </file>
 * </example-private>
 */
var PoAdvancedFilterComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoAdvancedFilterComponent, _super);
    function PoAdvancedFilterComponent(languageService) {
        return _super.call(this, languageService) || this;
    }
    /**
     * @return {?}
     */
    PoAdvancedFilterComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.filter = {};
        this.poModal.open();
    };
    PoAdvancedFilterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-advanced-filter',
                    template: "<po-modal\r\n  p-hide-close\r\n  [p-primary-action]=\"primaryAction\"\r\n  [p-secondary-action]=\"secondaryAction\"\r\n  [p-title]=\"literals.title\">\r\n\r\n  <po-dynamic-form\r\n    [p-fields]=\"filters\"\r\n    [p-value]=\"filter\">\r\n  </po-dynamic-form>\r\n\r\n</po-modal>\r\n"
                }] }
    ];
    /** @nocollapse */
    PoAdvancedFilterComponent.ctorParameters = function () { return [
        { type: PoLanguageService }
    ]; };
    PoAdvancedFilterComponent.propDecorators = {
        poDynamicForm: [{ type: ViewChild, args: [PoDynamicFormComponent, { static: true },] }]
    };
    return PoAdvancedFilterComponent;
}(PoAdvancedFilterBaseComponent));
export { PoAdvancedFilterComponent };
if (false) {
    /** @type {?} */
    PoAdvancedFilterComponent.prototype.poDynamicForm;
}
