// @portinari/portinari-templates/esm5/lib/components/po-page-dynamic-search/po-advanced-filter/po-advanced-filter-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { PoModalComponent } from '@portinari/portinari-ui';
import { poLocaleDefault } from '../../../utils/util';
/** @type {?} */
export var poAdvancedFiltersLiteralsDefault = {
    en: (/** @type {?} */ ({
        title: 'Advanced search',
        cancelLabel: 'Cancel',
        confirmLabel: 'Apply filters'
    })),
    es: (/** @type {?} */ ({
        title: 'Búsqueda avanzada',
        cancelLabel: 'Cancelar',
        confirmLabel: 'Aplicar filtros'
    })),
    pt: (/** @type {?} */ ({
        title: 'Busca avançada',
        cancelLabel: 'Cancelar',
        confirmLabel: 'Aplicar filtros'
    })),
    ru: (/** @type {?} */ ({
        title: 'Расширенный поиск',
        cancelLabel: 'отменить',
        confirmLabel: 'Применить фильтры'
    }))
};
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Filtro de busca avançada criado a partir de um formulário dinâmico.
 * Componente de uso interno.
 */
var PoAdvancedFilterBaseComponent = /** @class */ (function () {
    function PoAdvancedFilterBaseComponent(languageService) {
        var _this = this;
        this._filters = [];
        this.filter = {};
        this.language = poLocaleDefault;
        this.primaryAction = {
            action: (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var models = _this.getValuesFromForm();
                _this.searchEvent.emit(models);
                _this.poModal.close();
            }),
            label: this.literals.confirmLabel
        };
        this.secondaryAction = {
            action: (/**
             * @return {?}
             */
            function () {
                _this.poModal.close();
            }),
            label: this.literals.cancelLabel
        };
        /**
         * Função que será disparada e receberá os valores do formulário ao ser clicado no botão buscar.
         */
        this.searchEvent = new EventEmitter();
        this.language = languageService.getShortLanguage();
    }
    Object.defineProperty(PoAdvancedFilterBaseComponent.prototype, "filters", {
        get: /**
         * @return {?}
         */
        function () {
            return this._filters;
        },
        /**
         * Coleção de objetos que implementam a interface PoDynamicFormField, para definição dos campos que serão criados
         * dinamicamente.
         */
        set: /**
         * Coleção de objetos que implementam a interface PoDynamicFormField, para definição dos campos que serão criados
         * dinamicamente.
         * @param {?} filters
         * @return {?}
         */
        function (filters) {
            this._filters = Array.isArray(filters) ? tslib_1.__spread(filters) : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoAdvancedFilterBaseComponent.prototype, "literals", {
        get: /**
         * @return {?}
         */
        function () {
            return this._literals || poAdvancedFiltersLiteralsDefault[this.language];
        },
        /** Objeto com as literais usadas no `po-advanced-filter`. */
        set: /**
         * Objeto com as literais usadas no `po-advanced-filter`.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, poAdvancedFiltersLiteralsDefault[poLocaleDefault], poAdvancedFiltersLiteralsDefault[this.language], value);
            }
            else {
                this._literals = poAdvancedFiltersLiteralsDefault[this.language];
            }
            this.primaryAction.label = this.literals.confirmLabel;
            this.secondaryAction.label = this.literals.cancelLabel;
        },
        enumerable: true,
        configurable: true
    });
    // Retorna os models dos campos preenchidos
    // Retorna os models dos campos preenchidos
    /**
     * @private
     * @return {?}
     */
    PoAdvancedFilterBaseComponent.prototype.getValuesFromForm = 
    // Retorna os models dos campos preenchidos
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        Object.keys(this.filter).forEach((/**
         * @param {?} property
         * @return {?}
         */
        function (property) {
            if (_this.filter[property] === undefined || _this.filter[property] === '') {
                delete _this.filter[property];
            }
        }));
        return this.filter;
    };
    PoAdvancedFilterBaseComponent.propDecorators = {
        poModal: [{ type: ViewChild, args: [PoModalComponent, { static: true },] }],
        filters: [{ type: Input, args: ['p-filters',] }],
        literals: [{ type: Input, args: ['p-literals',] }],
        searchEvent: [{ type: Output, args: ['p-search-event',] }]
    };
    return PoAdvancedFilterBaseComponent;
}());
export { PoAdvancedFilterBaseComponent };
if (false) {
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.poModal;
    /**
     * @type {?}
     * @private
     */
    PoAdvancedFilterBaseComponent.prototype._filters;
    /**
     * @type {?}
     * @private
     */
    PoAdvancedFilterBaseComponent.prototype._literals;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.filter;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.language;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.primaryAction;
    /** @type {?} */
    PoAdvancedFilterBaseComponent.prototype.secondaryAction;
    /**
     * Função que será disparada e receberá os valores do formulário ao ser clicado no botão buscar.
     * @type {?}
     */
    PoAdvancedFilterBaseComponent.prototype.searchEvent;
}
