// @portinari/portinari-ui/esm2015/lib/components/po-field/po-clean/po-clean.component.js
import * as tslib_1 from "tslib";
import { Component, ViewContainerRef } from '@angular/core';
import { PoCleanBaseComponent } from './po-clean-base.component';
/**
 * @docsPrivate
 *
 * @docsExtends PoCleanBaseComponent
 *
 * @examplePrivate
 *
 * <example-private name="po-clean-labs" title="Portinari Clean Labs">
 *   <file name="sample-po-clean-labs.component.html"> </file>
 *   <file name="sample-po-clean-labs.component.ts"> </file>
 * </example-private>
 */
let PoCleanComponent = class PoCleanComponent extends PoCleanBaseComponent {
    constructor(viewRef) {
        super();
        this.viewRef = viewRef;
        this.parentComponent = this.viewRef['_view']['component'];
    }
    setInputValue(value) {
        if (this.inputRef && this.inputRef.nativeElement) {
            this.inputRef.nativeElement.value = value;
        }
    }
    getInputValue() {
        if (this.inputRef && this.inputRef.nativeElement) {
            return this.inputRef.nativeElement.value;
        }
    }
};
PoCleanComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
PoCleanComponent = tslib_1.__decorate([
    Component({
        selector: 'po-clean',
        template: "<span *ngIf=\"showIcon()\"\n  class=\"po-icon po-field-icon po-icon-close\"\n  (click)=\"clear()\">\n</span>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [ViewContainerRef])
], PoCleanComponent);
export { PoCleanComponent };
