// @portinari/portinari-ui/esm5/lib/components/po-table/po-table-subtitle-circle/po-table-subtitle-circle.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { PoColorPaletteService } from './../../../services/po-color-palette/po-color-palette.service';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para a criação da representação da legenda, em formato de círculo.
 */
var PoTableSubtitleCircleComponent = /** @class */ (function () {
    function PoTableSubtitleCircleComponent(poColorPaletteService) {
        this.poColorPaletteService = poColorPaletteService;
        /** Esconde a tooltip. */
        this.hideTitle = false;
    }
    Object.defineProperty(PoTableSubtitleCircleComponent.prototype, "subtitle", {
        get: function () {
            return this._subtitle;
        },
        /** Objeto com os dados da legenda. */
        set: function (subtitle) {
            if (subtitle) {
                subtitle.color = this.poColorPaletteService.getColor(subtitle);
            }
            this._subtitle = subtitle;
        },
        enumerable: true,
        configurable: true
    });
    PoTableSubtitleCircleComponent.ctorParameters = function () { return [
        { type: PoColorPaletteService }
    ]; };
    tslib_1.__decorate([
        Input('p-subtitle'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoTableSubtitleCircleComponent.prototype, "subtitle", null);
    tslib_1.__decorate([
        Input('p-hide-title'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoTableSubtitleCircleComponent.prototype, "hideTitle", void 0);
    PoTableSubtitleCircleComponent = tslib_1.__decorate([
        Component({
            selector: 'po-table-subtitle-circle',
            template: "<span\n  [title]=\"hideTitle ? '' : subtitle?.label\"\n  class=\"po-table-subtitle-circle po-{{ subtitle?.color }}\">\n  {{ subtitle?.content }}\n</span>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        tslib_1.__metadata("design:paramtypes", [PoColorPaletteService])
    ], PoTableSubtitleCircleComponent);
    return PoTableSubtitleCircleComponent;
}());
export { PoTableSubtitleCircleComponent };
