// @portinari/portinari-ui/esm5/lib/components/po-menu-panel/services/po-menu-panel-items.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço que implementa a comunicação entre os items do po-menu-panel.
 */
var PoMenuPanelItemsService = /** @class */ (function () {
    function PoMenuPanelItemsService() {
        this.subjectChild = new Subject();
        this.subjectParent = new Subject();
    }
    // Recebe do po-menu-panel-item sua informação de click.
    PoMenuPanelItemsService.prototype.receiveFromChildMenuClicked = function () {
        return this.subjectChild.asObservable();
    };
    // Recebe do po-menu-panel as informações processadas do click de um po-menu-panel-item.
    PoMenuPanelItemsService.prototype.receiveFromParentMenuClicked = function () {
        return this.subjectParent.asObservable();
    };
    // Envia informações do click do po-menu-panel-item para o po-menu
    PoMenuPanelItemsService.prototype.sendToParentMenuClicked = function (menu) {
        this.subjectChild.next(menu);
    };
    // Envia para os po-menu-panel-item a resposta do processamento de click de um po-menu-panel-item.
    PoMenuPanelItemsService.prototype.sendToChildMenuClicked = function (menu) {
        this.subjectParent.next(menu);
    };
    PoMenuPanelItemsService = tslib_1.__decorate([
        Injectable()
    ], PoMenuPanelItemsService);
    return PoMenuPanelItemsService;
}());
export { PoMenuPanelItemsService };
