// @portinari/portinari-templates/esm5/lib/components/po-page-login/po-page-login.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PoModule } from '@portinari/portinari-ui';
import { PoModalPasswordRecoveryComponent } from '../po-modal-password-recovery/po-modal-password-recovery.component';
import { PoPageBackgroundModule } from '../po-page-background/index';
import { PoPageLoginComponent } from './po-page-login.component';
import { PoPageLoginPopoverComponent } from './po-page-login-popover/po-page-login-popover.component';
import { PoPageLoginService } from './po-page-login.service';
/**
 * \@description
 *
 * Módulo do template do po-page-login.
 */
var PoPageLoginModule = /** @class */ (function () {
    function PoPageLoginModule() {
    }
    PoPageLoginModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        RouterModule,
                        PoPageBackgroundModule,
                        PoModule
                    ],
                    declarations: [
                        PoPageLoginComponent,
                        PoPageLoginPopoverComponent
                    ],
                    entryComponents: [
                        PoModalPasswordRecoveryComponent
                    ],
                    exports: [
                        PoPageLoginComponent
                    ],
                    providers: [PoPageLoginService]
                },] }
    ];
    return PoPageLoginModule;
}());
export { PoPageLoginModule };
