// @portinari/portinari-ui/esm5/lib/components/po-navbar/po-navbar-base.component.js
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { browserLanguage, convertToBoolean, poLocaleDefault } from '../../utils/util';
import { PoMenuComponent } from '../po-menu';
export var poNavbarLiteralsDefault = {
    en: {
        navbarLinks: 'Navbar links'
    },
    es: {
        navbarLinks: 'Navbar links'
    },
    pt: {
        navbarLinks: 'Navbar links'
    },
    ru: {
        navbarLinks: 'Navbar связи'
    }
};
/**
 * @description
 *
 * O componente `po-navbar` é um cabeçalho fixo que permite apresentar uma lista de links para facilitar a navegação pelas
 * páginas da aplicação. Também possui ícones com ações.
 */
var PoNavbarBaseComponent = /** @class */ (function () {
    function PoNavbarBaseComponent() {
        this._iconActions = [];
        this._items = [];
        this._shadow = false;
    }
    Object.defineProperty(PoNavbarBaseComponent.prototype, "iconActions", {
        get: function () {
            return this._iconActions;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define uma lista de ações apresentadas em ícones no lado direito do `po-navbar`.
         */
        set: function (value) {
            this._iconActions = Array.isArray(value) ? value : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoNavbarBaseComponent.prototype, "items", {
        get: function () {
            return this._items;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define uma lista de items do `po-navbar`.
         */
        set: function (value) {
            this._items = Array.isArray(value) ? value : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoNavbarBaseComponent.prototype, "literals", {
        get: function () {
            return this._literals || poNavbarLiteralsDefault[browserLanguage()];
        },
        /**
         * @optional
         *
         * @description
         *
         * Objeto com a literal usada na propriedade `p-literals`.
         *
         * Para customizar a literal, basta declarar um objeto do tipo `PoNavbarLiterals` conforme exemplo abaixo:
         *
         * ```
         *  const customLiterals: PoNavbarLiterals = {
         *    navbarLinks: 'Itens de navegação'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <po-navbar
         *   [p-literals]="customLiterals">
         * </po-navbar>
         * ```
         *
         *  > O objeto padrão de literais será traduzido de acordo com o idioma do *browser* (pt, en, es).
         */
        set: function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, poNavbarLiteralsDefault[poLocaleDefault], poNavbarLiteralsDefault[browserLanguage()], value);
            }
            else {
                this._literals = poNavbarLiteralsDefault[browserLanguage()];
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoNavbarBaseComponent.prototype, "logo", {
        get: function () {
            return this._logo;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define a logo apresentada `po-navbar`.
         */
        set: function (value) {
            this._logo = value;
            if (this.menu) {
                this.validateMenuLogo();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoNavbarBaseComponent.prototype, "shadow", {
        get: function () {
            return this._shadow;
        },
        /**
         * @optional
         *
         * @description
         *
         * Aplica uma sombra na parte inferior do `po-navbar`.
         *
         * @default `false`
         */
        set: function (value) {
            this._shadow = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input('p-icon-actions'),
        tslib_1.__metadata("design:type", Array),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], PoNavbarBaseComponent.prototype, "iconActions", null);
    tslib_1.__decorate([
        Input('p-items'),
        tslib_1.__metadata("design:type", Array),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], PoNavbarBaseComponent.prototype, "items", null);
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoNavbarBaseComponent.prototype, "literals", null);
    tslib_1.__decorate([
        Input('p-logo'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoNavbarBaseComponent.prototype, "logo", null);
    tslib_1.__decorate([
        Input('p-menu'),
        tslib_1.__metadata("design:type", PoMenuComponent)
    ], PoNavbarBaseComponent.prototype, "menu", void 0);
    tslib_1.__decorate([
        Input('p-shadow'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoNavbarBaseComponent.prototype, "shadow", null);
    return PoNavbarBaseComponent;
}());
export { PoNavbarBaseComponent };
