// @portinari/portinari-ui/esm2015/lib/components/po-tabs/po-tab-button/po-tab-button.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, OnChanges, Output, SimpleChanges } from '@angular/core';
import { convertToBoolean } from '../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente responsável por manipular os botões de aba.
 */
let PoTabButtonComponent = class PoTabButtonComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        // Função sera emitida quando a tab ficar ativada
        this.activated = new EventEmitter();
        // Função sera emitida quando a tab ficar desabilitada ou escondida
        this.changeState = new EventEmitter();
        // Método recebido do usuário para ser disparado quando clicar na aba
        this.click = new EventEmitter();
    }
    // Ativa o botão
    set active(value) {
        this._active = value;
        this.emitActivated();
    }
    get active() {
        return this._active;
    }
    // Oculta o botão
    set hide(value) {
        this._hide = convertToBoolean(value);
        this.setDisplayOnHide();
    }
    get hide() {
        return this._hide;
    }
    ngOnChanges(changes) {
        if ((changes.hide && changes.hide.currentValue) || (changes.disabled && changes.disabled.currentValue)) {
            this.changeState.emit(this);
        }
    }
    onClick() {
        if (!this.disabled) {
            this.click.emit(this.id);
        }
    }
    emitActivated() {
        if (this.active) {
            this.activated.emit(this);
        }
    }
    setDisplayOnHide() {
        this.elementRef.nativeElement.style.display = this.hide ? 'none' : '';
    }
};
PoTabButtonComponent.ctorParameters = () => [
    { type: ElementRef }
];
tslib_1.__decorate([
    Input('p-active'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], PoTabButtonComponent.prototype, "active", null);
tslib_1.__decorate([
    Input('p-disabled'),
    tslib_1.__metadata("design:type", Boolean)
], PoTabButtonComponent.prototype, "disabled", void 0);
tslib_1.__decorate([
    Input('p-hide'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], PoTabButtonComponent.prototype, "hide", null);
tslib_1.__decorate([
    Input('p-id'),
    tslib_1.__metadata("design:type", String)
], PoTabButtonComponent.prototype, "id", void 0);
tslib_1.__decorate([
    Input('p-label'),
    tslib_1.__metadata("design:type", String)
], PoTabButtonComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Input('p-small'),
    tslib_1.__metadata("design:type", Boolean)
], PoTabButtonComponent.prototype, "small", void 0);
tslib_1.__decorate([
    Output('p-activated'),
    tslib_1.__metadata("design:type", Object)
], PoTabButtonComponent.prototype, "activated", void 0);
tslib_1.__decorate([
    Output('p-change-state'),
    tslib_1.__metadata("design:type", Object)
], PoTabButtonComponent.prototype, "changeState", void 0);
tslib_1.__decorate([
    Output('p-click'),
    tslib_1.__metadata("design:type", Object)
], PoTabButtonComponent.prototype, "click", void 0);
PoTabButtonComponent = tslib_1.__decorate([
    Component({
        selector: 'po-tab-button',
        template: "<div\n  class=\"po-tab-button-md\"\n  [attr.tabindex]=\"disabled ? null : 0\"\n  [class.po-tab-button-active]=\"active\"\n  [class.po-tab-button-disabled]=\"disabled\"\n  [class.po-tab-button-sm]=\"small\"\n  (click)=\"onClick()\"\n  (keyup.enter)=\"onClick()\">\n\n  <span class=\"po-tab-button-label po-text-ellipsis\">\n    {{label}}\n  </span>\n</div>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef])
], PoTabButtonComponent);
export { PoTabButtonComponent };
