// @portinari/portinari-ui/esm5/lib/components/po-table/po-table-subtitle-footer/po-table-subtitle-footer.component.js
import * as tslib_1 from "tslib";
import { AfterViewInit, Component, DoCheck, ElementRef, Input, OnDestroy, Renderer2 } from '@angular/core';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para a criação de um conjunto de legendas.
 */
var PoTableSubtitleFooterComponent = /** @class */ (function () {
    function PoTableSubtitleFooterComponent(element, renderer) {
        this.element = element;
        this.renderer = renderer;
    }
    PoTableSubtitleFooterComponent.prototype.ngAfterViewInit = function () {
        this.initializeResizeListener();
        this.debounceResize();
    };
    PoTableSubtitleFooterComponent.prototype.ngDoCheck = function () {
        if (!this.isVisible && this.getContainerSize() > 0) {
            this.toggleShowCompleteSubtitle();
            this.isVisible = true;
        }
    };
    PoTableSubtitleFooterComponent.prototype.ngOnDestroy = function () {
        this.removeResizeListener();
    };
    PoTableSubtitleFooterComponent.prototype.debounceResize = function () {
        var _this = this;
        clearTimeout(this.timeoutResize);
        this.timeoutResize = setTimeout(function () {
            _this.toggleShowCompleteSubtitle();
        });
    };
    PoTableSubtitleFooterComponent.prototype.getContainerSize = function () {
        return this.element.nativeElement.querySelector('.po-table-subtitle-footer-container').offsetWidth;
    };
    PoTableSubtitleFooterComponent.prototype.getItemsSize = function () {
        var items = this.element.nativeElement.querySelectorAll('.po-table-subtitle-footer');
        return Array.from(items).map(function (item) { return item['offsetWidth']; }).reduce(function (a, b) { return a + b; }, 16);
    };
    PoTableSubtitleFooterComponent.prototype.initializeResizeListener = function () {
        var _this = this;
        this.resizeListener = this.renderer.listen('window', 'resize', function (event) {
            _this.debounceResize();
        });
    };
    PoTableSubtitleFooterComponent.prototype.removeResizeListener = function () {
        this.resizeListener();
    };
    PoTableSubtitleFooterComponent.prototype.toggleShowCompleteSubtitle = function () {
        var containerSize = this.getContainerSize();
        var itemsSize = this.getItemsSize();
        this.showSubtitle = itemsSize > containerSize;
    };
    PoTableSubtitleFooterComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object)
    ], PoTableSubtitleFooterComponent.prototype, "literals", void 0);
    tslib_1.__decorate([
        Input('p-subtitles'),
        tslib_1.__metadata("design:type", Array)
    ], PoTableSubtitleFooterComponent.prototype, "subtitles", void 0);
    PoTableSubtitleFooterComponent = tslib_1.__decorate([
        Component({
            selector: 'po-table-subtitle-footer',
            template: "<div class=\"po-table-subtitle-footer-container\">\n  <div *ngFor=\"let subtitle of subtitles\" class=\"po-table-subtitle-footer\">\n    <po-table-subtitle-circle\n      p-hide-title=\"true\"\n      [p-subtitle]=\"subtitle\">\n    </po-table-subtitle-circle>\n    <span class=\"po-table-subtitle-text\">{{ subtitle.label }}</span>\n  </div>\n  <po-table-show-subtitle *ngIf=\"showSubtitle\"\n    [p-literals]=\"literals\"\n    [p-subtitles]=\"subtitles\">\n  </po-table-show-subtitle>\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], PoTableSubtitleFooterComponent);
    return PoTableSubtitleFooterComponent;
}());
export { PoTableSubtitleFooterComponent };
