// @portinari/portinari-templates/esm5/lib/components/po-page-job-scheduler/po-page-job-scheduler.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ActivatedRoute, Router } from '@angular/router';
import { Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { PoDialogService, PoNotificationService, PoStepperStatus } from '@portinari/portinari-ui';
import * as util from './../../utils/util';
import { PoPageJobSchedulerInternal } from './po-page-job-scheduler-internal';
import { PoPageJobSchedulerBaseComponent } from './po-page-job-scheduler-base.component';
import { poPageJobSchedulerLiteralsDefault } from './po-page-job-scheduler-literals';
import { PoPageJobSchedulerLookupService } from './po-page-job-scheduler-lookup.service';
import { PoPageJobSchedulerService } from './po-page-job-scheduler.service';
/**
 * \@docsExtends PoPageJobSchedulerBaseComponent
 *
 * \@example
 *
 * <example name="po-page-job-scheduler-basic" title="Portinari Page Job Scheduler Basic">
 *  <file name="sample-po-page-job-scheduler-basic/sample-po-page-job-scheduler-basic.component.html"> </file>
 *  <file name="sample-po-page-job-scheduler-basic/sample-po-page-job-scheduler-basic.component.ts"> </file>
 * </example>
 *
 */
var PoPageJobSchedulerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoPageJobSchedulerComponent, _super);
    function PoPageJobSchedulerComponent(poPageDynamicLookupService, activatedRoute, poDialogService, poNotification, router, poPageJobSchedulerService) {
        var _this = _super.call(this, poPageJobSchedulerService) || this;
        _this.poPageDynamicLookupService = poPageDynamicLookupService;
        _this.activatedRoute = activatedRoute;
        _this.poDialogService = poDialogService;
        _this.poNotification = poNotification;
        _this.router = router;
        _this.isEdit = false;
        _this.literals = tslib_1.__assign({}, poPageJobSchedulerLiteralsDefault[util.poLocaleDefault], poPageJobSchedulerLiteralsDefault[util.browserLanguage()]);
        _this.parameters = [];
        _this.step = 1;
        _this.backPageAction = {
            label: _this.literals.back,
            action: _this.nextStepOperation.bind(_this, 'back'),
            disabled: _this.isDisabledBack.bind(_this)
        };
        _this.concludePageActions = [
            {
                label: _this.literals.conclude,
                action: _this.confirmJobScheduler.bind(_this)
            },
            tslib_1.__assign({}, _this.backPageAction)
        ];
        _this.nextPageActions = [
            {
                label: _this.literals.next,
                action: _this.nextStepOperation.bind(_this, 'next'),
                disabled: _this.isDisabledAdvance.bind(_this)
            },
            tslib_1.__assign({}, _this.backPageAction)
        ];
        _this.jobSchedulerActions = tslib_1.__spread(_this.nextPageActions);
        _this.steps = [
            { label: _this.literals.scheduling },
            { label: _this.literals.parameterization },
            { label: _this.literals.conclude }
        ];
        return _this;
    }
    Object.defineProperty(PoPageJobSchedulerComponent.prototype, "stepperOrientation", {
        get: /**
         * @return {?}
         */
        function () {
            return window.innerWidth > 481 && window.innerWidth < 960 ? 'horizontal' : 'vertical';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var paramId = this.activatedRoute.snapshot.params['id'];
        this.isEdit = !!paramId;
        this.poPageJobSchedulerService.configServiceApi({ endpoint: this.serviceApi });
        this.loadData(paramId);
    };
    /**
     * @param {?} currentStep
     * @param {?} nextStep
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.changePageActionsBySteps = /**
     * @param {?} currentStep
     * @param {?} nextStep
     * @return {?}
     */
    function (currentStep, nextStep) {
        /** @type {?} */
        var stepsLength = this.steps.length;
        if (nextStep === stepsLength) {
            this.jobSchedulerActions = tslib_1.__spread(this.concludePageActions);
        }
        else if (currentStep === stepsLength && nextStep < currentStep) {
            this.jobSchedulerActions = tslib_1.__spread(this.nextPageActions);
        }
    };
    /**
     * @param {?} stepNumber
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.nextStep = /**
     * @param {?} stepNumber
     * @return {?}
     */
    function (stepNumber) {
        if (stepNumber > 1 && this.schedulerExecution.form.invalid) {
            this.markAsDirtyInvalidControls(this.schedulerExecution.form.controls);
            return;
        }
        if (stepNumber > 2 && (this.schedulerParameters && this.schedulerParameters.form && this.schedulerParameters.form.invalid)) {
            this.markAsDirtyInvalidControls(this.schedulerParameters.form.controls);
            return;
        }
        this.changePageActionsBySteps(this.step, stepNumber);
        /** @type {?} */
        var steps = this.steps[this.step - 1];
        this.step = stepNumber;
        if (steps) {
            steps.status = PoStepperStatus.Done;
        }
    };
    /**
     * @param {?} process
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.onChangeProcess = /**
     * @param {?} process
     * @return {?}
     */
    function (process) {
        if (process.existAPI && process.processId) {
            this.getParametersByProcess(process.processId);
            if (!this.isEdit) {
                this.model.executionParameter = {};
            }
            return;
        }
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.confirmJobScheduler = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var paramId = this.activatedRoute.snapshot.params['id'];
        /** @type {?} */
        var confirmMessage = paramId ? this.literals.confirmUpdateMessage : this.literals.confirmSaveMessage;
        this.poDialogService.confirm({
            title: this.literals.confirmation,
            message: confirmMessage,
            confirm: (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var model = Object.assign({}, _this.model);
                _this.save(model, paramId);
            })
        });
    };
    /**
     * @private
     * @param {?} msgSuccess
     * @param {?} saveOperation
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.emitSuccessMessage = /**
     * @private
     * @param {?} msgSuccess
     * @param {?} saveOperation
     * @return {?}
     */
    function (msgSuccess, saveOperation) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, saveOperation.toPromise()];
                    case 1:
                        _a.sent();
                        this.poNotification.success(msgSuccess);
                        this.resetJobSchedulerForm();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @private
     * @param {?} process
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.getParametersByProcess = /**
     * @private
     * @param {?} process
     * @return {?}
     */
    function (process) {
        var _this = this;
        this.poPageJobSchedulerService.getParametersByProcess(process).subscribe((/**
         * @param {?} parameters
         * @return {?}
         */
        function (parameters) {
            _this.parameters = parameters;
        }));
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.isDisabledAdvance = /**
     * @private
     * @return {?}
     */
    function () {
        return this.schedulerExecution ? this.schedulerExecution.form.invalid : false;
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.isDisabledBack = /**
     * @private
     * @return {?}
     */
    function () {
        return this.step === 1;
    };
    /**
     * @private
     * @param {?=} operation
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.nextStepOperation = /**
     * @private
     * @param {?=} operation
     * @return {?}
     */
    function (operation) {
        /** @type {?} */
        var stepNumber = operation === 'back' ? this.step - 1 : this.step + 1;
        this.nextStep(stepNumber);
    };
    /**
     * @private
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.resetJobSchedulerForm = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.schedulerExecution.form.reset();
        // radiogroup não estava atribuindo novo valor, fica vermelho sem o timetout.
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.model = new PoPageJobSchedulerInternal();
            _this.step = 1;
            _this.steps.forEach((/**
             * @param {?} step
             * @return {?}
             */
            function (step) {
                step.status = PoStepperStatus.Default;
            }));
            _this.jobSchedulerActions = tslib_1.__spread(_this.nextPageActions);
        }));
    };
    /**
     * @private
     * @param {?} model
     * @param {?} paramId
     * @return {?}
     */
    PoPageJobSchedulerComponent.prototype.save = /**
     * @private
     * @param {?} model
     * @param {?} paramId
     * @return {?}
     */
    function (model, paramId) {
        /** @type {?} */
        var saveOperation = paramId
            ? this.poPageJobSchedulerService.updateResource(paramId, model)
            : this.poPageJobSchedulerService.createResource(model);
        /** @type {?} */
        var msgSuccess = paramId ? this.literals.saveNotificationSuccessUpdate : this.literals.saveNotificationSuccessSave;
        this.emitSuccessMessage(msgSuccess, saveOperation);
    };
    PoPageJobSchedulerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-job-scheduler',
                    template: "<po-page-default\r\n  [p-actions]=\"jobSchedulerActions\"\r\n  [p-breadcrumb]=\"breadcrumb\"\r\n  [p-title]=\"title\">\r\n\r\n  <div class=\"po-row\">\r\n\r\n    <po-stepper\r\n      class=\"po-lg-3 po-xl-2\"\r\n      p-sequential=\"true\"\r\n      [p-orientation]=\"stepperOrientation\"\r\n      [p-step]=\"step\"\r\n      [p-steps]=\"steps\"\r\n      (p-change-step)=\"nextStep($event)\">\r\n    </po-stepper>\r\n\r\n    <po-container class=\"po-lg-8 po-xl-6\">\r\n      <form #formScheduler=\"ngForm\">\r\n        <div class=\"po-row\">\r\n          <po-page-job-scheduler-execution [hidden]=\"step !== 1\"\r\n            #schedulerExecution\r\n            class=\"po-md-12\"\r\n            [p-is-edit]=\"isEdit\"\r\n            [p-literals]=\"literals\"\r\n            [p-value]=\"model\"\r\n            (p-change-process)=\"onChangeProcess($event)\">\r\n          </po-page-job-scheduler-execution>\r\n\r\n          <po-page-job-scheduler-parameters *ngIf=\"step === 2\"\r\n            #schedulerParameters\r\n            class=\"po-md-12\"\r\n            [p-literals]=\"literals\"\r\n            [p-parameters]=\"parameters || []\"\r\n            [(p-value)]=\"model.executionParameter\">\r\n          </po-page-job-scheduler-parameters>\r\n\r\n          <po-page-job-scheduler-summary *ngIf=\"step === 3\"\r\n            class=\"po-md-12\"\r\n            [p-literals]=\"literals\"\r\n            [p-parameters]=\"parameters\"\r\n            [p-value]=\"model\">\r\n          </po-page-job-scheduler-summary>\r\n        </div>\r\n      </form>\r\n    </po-container>\r\n  </div>\r\n\r\n</po-page-default>\r\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["\n    po-container .po-container {\n      overflow-y: unset;\n    }\n  "]
                }] }
    ];
    /** @nocollapse */
    PoPageJobSchedulerComponent.ctorParameters = function () { return [
        { type: PoPageJobSchedulerLookupService },
        { type: ActivatedRoute },
        { type: PoDialogService },
        { type: PoNotificationService },
        { type: Router },
        { type: PoPageJobSchedulerService }
    ]; };
    PoPageJobSchedulerComponent.propDecorators = {
        schedulerExecution: [{ type: ViewChild, args: ['schedulerExecution', { static: true },] }],
        schedulerParameters: [{ type: ViewChild, args: ['schedulerParameters', { static: true },] }]
    };
    return PoPageJobSchedulerComponent;
}(PoPageJobSchedulerBaseComponent));
export { PoPageJobSchedulerComponent };
if (false) {
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.isEdit;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.literals;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.parameters;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.saveOperation;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.step;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerComponent.prototype.backPageAction;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerComponent.prototype.concludePageActions;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerComponent.prototype.nextPageActions;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.jobSchedulerActions;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.steps;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.schedulerExecution;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.schedulerParameters;
    /** @type {?} */
    PoPageJobSchedulerComponent.prototype.poPageDynamicLookupService;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerComponent.prototype.poDialogService;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerComponent.prototype.poNotification;
    /**
     * @type {?}
     * @private
     */
    PoPageJobSchedulerComponent.prototype.router;
}
