// @portinari/portinari-templates/esm5/lib/components/po-page-job-scheduler/po-page-job-scheduler-internal.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var PoPageJobSchedulerInternal = /** @class */ (function () {
    function PoPageJobSchedulerInternal() {
        this.periodicity = 'single';
        this.firstExecution = new Date();
        this.firstExecutionHour = this.getCurrentHour(this.firstExecution);
        this.recurrent = true;
    }
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    PoPageJobSchedulerInternal.prototype.getCurrentHour = /**
     * @private
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var hours = date.getHours() < 10 ? '0' + date.getHours() : date.getHours();
        /** @type {?} */
        var minutes = date.getMinutes() < 10 ? '0' + date.getMinutes() : date.getMinutes();
        return hours + ":" + minutes;
    };
    return PoPageJobSchedulerInternal;
}());
export { PoPageJobSchedulerInternal };
if (false) {
    /** @type {?} */
    PoPageJobSchedulerInternal.prototype.periodicity;
    /** @type {?} */
    PoPageJobSchedulerInternal.prototype.firstExecution;
    /** @type {?} */
    PoPageJobSchedulerInternal.prototype.firstExecutionHour;
    /** @type {?} */
    PoPageJobSchedulerInternal.prototype.recurrent;
}
