// @portinari/portinari-ui/esm5/lib/components/po-button/po-button.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, ViewChild } from '@angular/core';
import { PoButtonBaseComponent } from './po-button-base.component';
/**
 * @docsExtends PoButtonBaseComponent
 *
 * @example
 *
 * <example name="po-button-basic" title="Portinari Button Basic">
 *  <file name="sample-po-button-basic/sample-po-button-basic.component.html"> </file>
 *  <file name="sample-po-button-basic/sample-po-button-basic.component.ts"> </file>
 *  <file name="sample-po-button-basic/sample-po-button-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-button-basic/sample-po-button-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-button-labs" title="Portinari Button Labs">
 *  <file name="sample-po-button-labs/sample-po-button-labs.component.html"> </file>
 *  <file name="sample-po-button-labs/sample-po-button-labs.component.ts"> </file>
 *  <file name="sample-po-button-labs/sample-po-button-labs.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-button-labs/sample-po-button-labs.component.po.ts"> </file>
 * </example>
 */
var PoButtonComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoButtonComponent, _super);
    function PoButtonComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Função que atribui foco ao componente.
     *
     * Para utilizá-la é necessário ter a instância do componente no DOM, podendo ser utilizado o ViewChild da seguinte forma:
     *
     * ```
     * import { PoButtonComponent } from '@portinari/portinari-ui';
     *
     * ...
     *
     * @ViewChild(PoButtonComponent, { static: true }) button: PoButtonComponent;
     *
     * focusButton() {
     *   this.button.focus();
     * }
     * ```
     */
    PoButtonComponent.prototype.focus = function () {
        if (!this.disabled) {
            this.buttonElement.nativeElement.focus();
        }
    };
    PoButtonComponent.prototype.onClick = function () {
        this.click.emit(null);
    };
    tslib_1.__decorate([
        ViewChild('button', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoButtonComponent.prototype, "buttonElement", void 0);
    PoButtonComponent = tslib_1.__decorate([
        Component({
            selector: 'po-button',
            template: "<button\n  #button\n  class=\"po-button po-text-ellipsis\"\n  type=\"button\"\n  [class.po-button-danger]=\"type === 'danger'\"\n  [class.po-button-link]=\"type === 'link'\"\n  [class.po-button-primary]=\"type === 'primary'\"\n  [class.po-button-sm] = \"small\"\n  [class.po-clickable]=\"type === 'link'\"\n  [disabled]=\"disabled || loading\"\n  (click)=\"onClick()\">\n\n  <div *ngIf=\"loading\" class=\"po-button-loading-icon\">\n    <po-loading-icon p-neutral-color></po-loading-icon>\n  </div>\n  <span *ngIf=\"icon\" class=\"po-icon {{ icon }}\" aria-hidden=\"true\"></span>\n  <span *ngIf=\"label\" class=\"po-button-label\">{{ label }}</span>\n</button>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], PoButtonComponent);
    return PoButtonComponent;
}(PoButtonBaseComponent));
export { PoButtonComponent };
