// @portinari/portinari-ui/esm2015/lib/components/po-menu-panel/services/po-menu-panel-items.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço que implementa a comunicação entre os items do po-menu-panel.
 */
let PoMenuPanelItemsService = class PoMenuPanelItemsService {
    /**
     * @docsPrivate
     *
     * @description
     *
     * Serviço que implementa a comunicação entre os items do po-menu-panel.
     */
    constructor() {
        this.subjectChild = new Subject();
        this.subjectParent = new Subject();
    }
    // Recebe do po-menu-panel-item sua informação de click.
    receiveFromChildMenuClicked() {
        return this.subjectChild.asObservable();
    }
    // Recebe do po-menu-panel as informações processadas do click de um po-menu-panel-item.
    receiveFromParentMenuClicked() {
        return this.subjectParent.asObservable();
    }
    // Envia informações do click do po-menu-panel-item para o po-menu
    sendToParentMenuClicked(menu) {
        this.subjectChild.next(menu);
    }
    // Envia para os po-menu-panel-item a resposta do processamento de click de um po-menu-panel-item.
    sendToChildMenuClicked(menu) {
        this.subjectParent.next(menu);
    }
};
PoMenuPanelItemsService = tslib_1.__decorate([
    Injectable()
], PoMenuPanelItemsService);
export { PoMenuPanelItemsService };
