// @portinari/portinari-ui/esm2015/lib/components/po-navbar/po-navbar-items/po-navbar-items.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, Input, QueryList, ViewChild, ViewChildren } from '@angular/core';
import { NavigationCancel, NavigationEnd, Router } from '@angular/router';
import { PoNavbarItemComponent } from './po-navbar-item/po-navbar-item.component';
let PoNavbarItemsComponent = class PoNavbarItemsComponent {
    constructor(router) {
        this.router = router;
    }
    ngOnDestroy() {
        this.routeSubscription.unsubscribe();
    }
    ngOnInit() {
        this.subscribeToRoute();
    }
    selectItem(item) {
        this.selectedItem = item;
    }
    checkActiveItemByUrl(urlRouter) {
        this.selectedItem = this.items.find(item => item.link === urlRouter);
    }
    checkRouterChildrenFragments() {
        const childrenPrimary = this.router.parseUrl(this.router.url).root.children['primary'];
        return childrenPrimary ? `/${childrenPrimary.segments.map(it => it.path).join('/')}` : '';
    }
    subscribeToRoute() {
        this.routeSubscription = this.router.events.subscribe(val => {
            if (val instanceof NavigationEnd || val instanceof NavigationCancel) {
                const urlRouter = this.checkRouterChildrenFragments();
                this.checkActiveItemByUrl(urlRouter);
            }
        });
    }
};
PoNavbarItemsComponent.ctorParameters = () => [
    { type: Router }
];
tslib_1.__decorate([
    ViewChild('navbarItemsContainer', { read: ElementRef, static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], PoNavbarItemsComponent.prototype, "navbarItemsContainer", void 0);
tslib_1.__decorate([
    ViewChildren(PoNavbarItemComponent, { read: ElementRef }),
    tslib_1.__metadata("design:type", QueryList)
], PoNavbarItemsComponent.prototype, "allNavbarItems", void 0);
tslib_1.__decorate([
    Input('p-items'),
    tslib_1.__metadata("design:type", Array)
], PoNavbarItemsComponent.prototype, "items", void 0);
PoNavbarItemsComponent = tslib_1.__decorate([
    Component({
        selector: 'po-navbar-items',
        template: "<nav>\n  <ul class=\"po-navbar-items-container\" #navbarItemsContainer>\n    <li *ngFor=\"let item of items\"\n      class=\"po-navbar-item\"\n      [class.po-navbar-item-selected]=\"selectedItem === item\">\n      <po-navbar-item\n        [p-action]=\"item.action\"\n        [p-clickable]=\"selectedItem !== item\"\n        [p-label]=\"item.label\"\n        [p-link]=\"item.link\"\n        (p-click)=\"selectItem(item)\">\n      </po-navbar-item>\n    </li>\n  </ul>\n</nav>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [Router])
], PoNavbarItemsComponent);
export { PoNavbarItemsComponent };
