// @portinari/portinari-ui/esm5/lib/components/po-widget/po-widget-base.component.js
import * as tslib_1 from "tslib";
import { Input, Output, EventEmitter } from '@angular/core';
import { convertToBoolean, isTypeof } from '../../utils/util';
/**
 *
 * @description
 *
 * O componente `po-widget` é recomendado para exibição de *dashboards*, podendo ser utilizado
 * para incluir vários tipos de conteúdo como: gráficos, tabelas, grids e imagens.
 *
 * Além da exibição de conteúdos, este componente possibilita adicionar ações e um link
 * para ajuda, como também possibilita ser utilizado com ou sem sombra.
 *
 * Para controlar sua largura, é possível utilizar o [Grid System](/guides/grid-system) para um maior
 * controle de seu redimensionamento, assim possibilitando o tratamento para diferentes resoluções.
 */
var PoWidgetBaseComponent = /** @class */ (function () {
    function PoWidgetBaseComponent() {
        this._disabled = false;
        this._noShadow = false;
        this._primary = false;
        this.containerHeight = 'auto';
        /** Ação que será executada quando o usuário clicar sobre a área total do `po-widget`. */
        this.click = new EventEmitter();
        /** Função que será disparada com o valor do `p-disabled` quando esta propriedade for alterada. */
        this.onDisabled = new EventEmitter();
        /** Função que será chamada na primeira ação. */
        this.primaryAction = new EventEmitter();
        /** Função que será chamada na segunda ação. */
        this.secondaryAction = new EventEmitter();
        /** Função chamada ao clicar no ícone de configuração */
        this.setting = new EventEmitter();
        /** Função que será chamada ao clicar no título. */
        this.titleAction = new EventEmitter();
    }
    Object.defineProperty(PoWidgetBaseComponent.prototype, "background", {
        get: function () {
            return this._background;
        },
        /**
         * @optional
         *
         * @description
         *
         * Aplicação de imagem de fundo.
         * > Se a imagem escolhida intervir na legibilidade do texto contido no `p-widget`,
         * pode-se utilizar a propriedade `p-primary` em conjunto para que os textos fiquem na cor branca.
         *
         */
        set: function (value) {
            this._background = (value && typeof value === 'string') ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "disabled", {
        get: function () {
            return this._disabled;
        },
        /**
         * @optional
         *
         * @description
         *
         * Desabilita todas as ações do componente.
         *
         * @default `false`
         */
        set: function (value) {
            this._disabled = convertToBoolean(value);
            this.onDisabled.emit(this.disabled);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "height", {
        get: function () {
            return this._height;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define a altura do `po-widget`.
         * A altura mínima para o `po-widget` depende do que será exibido através das propriedades `p-primary-label`,
         * `p-setting`, `p-help` e `p-title`.
         * > Caso não seja informado valor, a propriedade irá assumir o tamanho do conteúdo.
         */
        set: function (value) {
            this._height = parseInt(value, 10);
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "help", {
        get: function () {
            return this._help;
        },
        /**
         * @optional
         *
         * @description
         *
         * Link de ajuda
         */
        set: function (value) {
            this._help = isTypeof(value, 'string') ? value : '';
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "noShadow", {
        get: function () {
            return this._noShadow;
        },
        /**
         * @optional
         *
         * @description
         *
         * Desabilita a sombra do `po-widget`.
         *
         * @default `true`
         */
        set: function (value) {
            this._noShadow = value === '' ? true : convertToBoolean(value);
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "primary", {
        get: function () {
            return this._primary;
        },
        /**
         * @optional
         *
         * @description
         *
         * Opção para que o `po-widget` fique em destaque.
         *
         * @default `false`
         */
        set: function (value) {
            this._primary = value === '' ? true : convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "primaryLabel", {
        get: function () {
            return this._primaryLabel;
        },
        /**
         * @optional
         *
         * @description
         *
         * Descrição da primeira ação.
         *
         * @default `false`
         */
        set: function (value) {
            this._primaryLabel = isTypeof(value, 'string') ? value : '';
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "title", {
        get: function () {
            return this._title;
        },
        /**
         * @optional
         *
         * @description
         *
         * Título do `po-widget`.
         *
         * @default `false`
         */
        set: function (value) {
            this._title = isTypeof(value, 'string') ? value : '';
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input('p-background'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoWidgetBaseComponent.prototype, "background", null);
    tslib_1.__decorate([
        Input('p-disabled'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoWidgetBaseComponent.prototype, "disabled", null);
    tslib_1.__decorate([
        Input('p-height'),
        tslib_1.__metadata("design:type", Number),
        tslib_1.__metadata("design:paramtypes", [Number])
    ], PoWidgetBaseComponent.prototype, "height", null);
    tslib_1.__decorate([
        Input('p-help'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoWidgetBaseComponent.prototype, "help", null);
    tslib_1.__decorate([
        Input('p-no-shadow'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoWidgetBaseComponent.prototype, "noShadow", null);
    tslib_1.__decorate([
        Input('p-primary'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoWidgetBaseComponent.prototype, "primary", null);
    tslib_1.__decorate([
        Input('p-primary-label'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoWidgetBaseComponent.prototype, "primaryLabel", null);
    tslib_1.__decorate([
        Input('p-secondary-label'),
        tslib_1.__metadata("design:type", String)
    ], PoWidgetBaseComponent.prototype, "secondaryLabel", void 0);
    tslib_1.__decorate([
        Input('p-title'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoWidgetBaseComponent.prototype, "title", null);
    tslib_1.__decorate([
        Output('p-click'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoWidgetBaseComponent.prototype, "click", void 0);
    tslib_1.__decorate([
        Output('p-on-disabled'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoWidgetBaseComponent.prototype, "onDisabled", void 0);
    tslib_1.__decorate([
        Output('p-primary-action'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoWidgetBaseComponent.prototype, "primaryAction", void 0);
    tslib_1.__decorate([
        Output('p-secondary-action'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoWidgetBaseComponent.prototype, "secondaryAction", void 0);
    tslib_1.__decorate([
        Output('p-setting'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoWidgetBaseComponent.prototype, "setting", void 0);
    tslib_1.__decorate([
        Output('p-title-action'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoWidgetBaseComponent.prototype, "titleAction", void 0);
    return PoWidgetBaseComponent;
}());
export { PoWidgetBaseComponent };
