// @portinari/portinari-templates/esm5/lib/components/po-page-change-password/po-page-change-password.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PoModule } from '@portinari/portinari-ui';
import { PoModalPasswordRecoveryComponent } from '../po-modal-password-recovery/po-modal-password-recovery.component';
import { PoPageBackgroundModule } from '../po-page-background/index';
import { PoPageChangePasswordComponent } from './po-page-change-password.component';
import { PoPageChangePasswordService } from './po-page-change-password.service';
/**
 * \@description
 *
 * Módulo do template do po-page-change-password.
 */
var PoPageChangePasswordModule = /** @class */ (function () {
    function PoPageChangePasswordModule() {
    }
    PoPageChangePasswordModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        RouterModule,
                        PoModule,
                        PoPageBackgroundModule
                    ],
                    declarations: [
                        PoPageChangePasswordComponent
                    ],
                    entryComponents: [
                        PoModalPasswordRecoveryComponent
                    ],
                    providers: [
                        PoPageChangePasswordService
                    ],
                    exports: [
                        PoPageChangePasswordComponent
                    ]
                },] }
    ];
    return PoPageChangePasswordModule;
}());
export { PoPageChangePasswordModule };
