// @portinari/portinari-templates/esm5/lib/components/po-page-change-password/po-page-change-password.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild, ViewContainerRef } from '@angular/core';
import { NgForm } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { PoComponentInjectorService, PoModalComponent } from '@portinari/portinari-ui';
import { browserLanguage, isExternalLink, isTypeof, poLocaleDefault } from '../../utils/util';
import { PoModalPasswordRecoveryComponent } from '../po-modal-password-recovery/po-modal-password-recovery.component';
import { PoModalPasswordRecoveryType } from '../po-modal-password-recovery/enums/po-modal-password-recovery-type.enum';
import { PoPageChangePasswordBaseComponent } from './po-page-change-password-base.component';
import { poPageChangePasswordLiterals } from './literals/i18n/po-page-change-password-literals';
import { PoPageChangePasswordService } from './po-page-change-password.service';
/**
 * \@docsExtends PoPageChangePasswordBaseComponent
 *
 * \@example
 *
 * <example name="po-page-change-password-basic" title="Portinari Page Change Password Basic">
 *  <file name="sample-po-page-change-password-basic/sample-po-page-change-password-basic.component.html"> </file>
 *  <file name="sample-po-page-change-password-basic/sample-po-page-change-password-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-labs" title="Portinari Page Change Password Labs">
 *  <file name="sample-po-page-change-password-labs/sample-po-page-change-password-labs.component.html"> </file>
 *  <file name="sample-po-page-change-password-labs/sample-po-page-change-password-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-modify" title="Portinari Page Change Password Modify">
 *  <file name="sample-po-page-change-password-modify/sample-po-page-change-password-modify.component.html"> </file>
 *  <file name="sample-po-page-change-password-modify/sample-po-page-change-password-modify.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-create" title="Portinari Page Change Password Create">
 *  <file name="sample-po-page-change-password-create/sample-po-page-change-password-create.component.html"> </file>
 *  <file name="sample-po-page-change-password-create/sample-po-page-change-password-create.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-request" title="Portinari Page Change Password Request">
 *  <file name="sample-po-page-change-password-request/sample-po-page-change-password-request.component.html"> </file>
 *  <file name="sample-po-page-change-password-request/sample-po-page-change-password-request.component.ts"> </file>
 * </example>
 */
var PoPageChangePasswordComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoPageChangePasswordComponent, _super);
    function PoPageChangePasswordComponent(activatedRoute, route, router, service, poComponentInjector, viewRef) {
        var _this = _super.call(this) || this;
        _this.activatedRoute = activatedRoute;
        _this.route = route;
        _this.router = router;
        _this.service = service;
        _this.poComponentInjector = poComponentInjector;
        _this.literals = tslib_1.__assign({}, poPageChangePasswordLiterals[poLocaleDefault], poPageChangePasswordLiterals[browserLanguage()]);
        _this.componentRef = null;
        _this.modalAction = {
            action: _this.navigateTo.bind(_this, _this.urlHome),
            label: _this.literals.enterSystemButton
        };
        _this.parentRef = viewRef['_view']['component'];
        return _this;
    }
    /**
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.urlNewPassword) {
            this.subscribeToTokenParameter();
        }
    };
    /**
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.newPasswordSubscription) {
            this.newPasswordSubscription.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.navigateTo = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        isExternalLink(url) ? window.open(url) : this.router.navigate([url || '/']);
    };
    /**
     * @param {?} recovery
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.onForgotPasswordClick = /**
     * @param {?} recovery
     * @return {?}
     */
    function (recovery) {
        if (isTypeof(recovery, 'function')) {
            recovery();
        }
        else {
            this.createModalPasswordRecoveryComponent(recovery);
        }
    };
    /**
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.onLoginSubmit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var form = this.getLoginForm();
        if (this.urlNewPassword) {
            this.postUrlNewPassword(form);
        }
        else if (this.submit && this.submit.observers.length) {
            this.emitSubmit(form);
        }
    };
    /**
     * Abre uma modal de confirmação com texto, imagem e botão que redireciona para o link definido na propriedade `p-url-home`
     */
    /**
     * Abre uma modal de confirmação com texto, imagem e botão que redireciona para o link definido na propriedade `p-url-home`
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.openConfirmation = /**
     * Abre uma modal de confirmação com texto, imagem e botão que redireciona para o link definido na propriedade `p-url-home`
     * @return {?}
     */
    function () {
        this.modal.open();
    };
    /**
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.validatePassword = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var controls = this.passwordForm.form.controls;
        /** @type {?} */
        var controlConfirmPassword = controls['confirmPassword'];
        /** @type {?} */
        var controlNewPassword = controls['newPassword'];
        if (!this.newPassword) {
            this.setFormErrors({ 'required': true }, [controlNewPassword]);
        }
        else if (!this.confirmPassword) {
            this.setFormErrors({ 'required': true }, [controlConfirmPassword]);
        }
        else if (this.newPassword && this.confirmPassword && this.newPassword !== this.confirmPassword) {
            this.setFormErrors({ 'equalPassword': true }, [controlNewPassword, controlConfirmPassword]);
        }
        else {
            this.setFormErrors(null, [controlConfirmPassword, controlNewPassword]);
        }
        if (this.requirements.length && this.requirements.find((/**
         * @param {?} requirement
         * @return {?}
         */
        function (requirement) { return _this.validateRequirement(requirement) === false; }))) {
            this.setFormErrors({ 'requirement': true }, [controlNewPassword]);
        }
    };
    /**
     * @param {?} requirement
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.validateRequirement = /**
     * @param {?} requirement
     * @return {?}
     */
    function (requirement) {
        return typeof requirement.status === 'function' ? requirement.status.call(this.parentRef, this.newPassword) : requirement.status;
    };
    /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.checkingForMetadataProperty = /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    function (object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    };
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.checkingForRouteMetadata = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (Object.keys(data).length !== 0) {
            this.urlNewPassword = this.checkingForMetadataProperty(data, 'serviceApi') || this.urlNewPassword;
            this.recovery = this.checkingForMetadataProperty(data, 'recovery') || this.recovery;
            this.hideCurrentPassword = this.checkingForMetadataProperty(data, 'hideCurrentPassword') || this.hideCurrentPassword;
        }
    };
    /**
     * @private
     * @param {?} recovery
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.createModalPasswordRecoveryComponent = /**
     * @private
     * @param {?} recovery
     * @return {?}
     */
    function (recovery) {
        var _this = this;
        if (this.componentRef) {
            this.poComponentInjector.destroyComponentInApplication(this.componentRef);
        }
        this.componentRef = this.poComponentInjector.createComponentInApplication(PoModalPasswordRecoveryComponent);
        this.componentRef.instance.recovery = recovery.url;
        this.componentRef.instance.contactEmail = recovery.contactMail;
        this.componentRef.instance.phoneMask = recovery.phoneMask;
        this.componentRef.instance.type = recovery.type || PoModalPasswordRecoveryType.Email;
        this.componentRef.changeDetectorRef.detectChanges();
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.componentRef.instance.open();
        }));
    };
    /**
     * @private
     * @param {?} form
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.emitSubmit = /**
     * @private
     * @param {?} form
     * @return {?}
     */
    function (form) {
        this.submit.emit(form);
    };
    /**
     * @private
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.getLoginForm = /**
     * @private
     * @return {?}
     */
    function () {
        return {
            currentPassword: this.currentPassword,
            newPassword: this.newPassword
        };
    };
    /**
     * @private
     * @param {?} form
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.postUrlNewPassword = /**
     * @private
     * @param {?} form
     * @return {?}
     */
    function (form) {
        var _this = this;
        form['token'] = this.token;
        this.service.post(this.urlNewPassword, form).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            if (response.status === 204) {
                _this.openConfirmation();
            }
        }));
    };
    /**
     * @private
     * @param {?} error
     * @param {?=} controls
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.setFormErrors = /**
     * @private
     * @param {?} error
     * @param {?=} controls
     * @return {?}
     */
    function (error, controls) {
        controls.forEach((/**
         * @param {?} control
         * @return {?}
         */
        function (control) {
            control.setErrors(error);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    PoPageChangePasswordComponent.prototype.subscribeToTokenParameter = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.route.queryParams.subscribe((/**
         * @param {?} params
         * @return {?}
         */
        function (params) {
            /** @type {?} */
            var token = params['token'];
            if (token) {
                _this.token = token;
            }
        }));
    };
    PoPageChangePasswordComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-change-password',
                    template: "<div class=\"po-page-change-password-wrapper\" #pageChangePassword>\r\n  <div class=\"po-page-change-password-container\">\r\n    <po-page-background class=\"po-page-change-password\"\r\n      [p-logo]=\"logo\"\r\n      [p-secondary-logo]=\"secondaryLogo\">\r\n\r\n      <div class=\"po-page-blocked-user-header po-row\">\r\n        <div class=\"po-md-12\">{{literals.createNewPassword}}</div>\r\n        <div class=\"po-md-12 po-text-center po-font-text\">\r\n          {{literals.createNewPasswordPhrase}}\r\n        </div>\r\n      </div>\r\n\r\n      <form #passwordForm=\"ngForm\">\r\n\r\n        <div class=\"po-row po-mt-3 po-mt-sm-1\">\r\n          <div class=\"po-md-6\"\r\n            [ngClass]=\"{'po-offset-md-3 po-offset-lg-3 po-offset-xl-3': !showRequirements}\">\r\n            <div *ngIf=\"!hideCurrentPassword\">\r\n              <po-password\r\n                name=\"currentPassword\"\r\n                [(ngModel)]=\"currentPassword\"\r\n                p-required\r\n                [p-label]=\"literals.currentPassword\">\r\n              </po-password>\r\n            </div>\r\n            <div *ngIf=\"!hideCurrentPassword && recovery\" class=\"po-text-center po-mb-sm-1 po-mb-2\">\r\n              <a *ngIf=\"recoveryUrlType === 'internalLink'\"\r\n                class=\"po-font-text-bold po-page-change-password-forgot-password\"\r\n                [routerLink]=\"recovery\">\r\n                {{literals.forgotPassword}}\r\n              </a>\r\n              <a *ngIf=\"recoveryUrlType === 'externalLink'\"\r\n                class=\"po-font-text-bold po-page-change-password-forgot-password\"\r\n                [href]=\"recovery\">\r\n                {{literals.forgotPassword}}\r\n              </a>\r\n              <a *ngIf=\"!recoveryUrlType\"\r\n                class=\"po-font-text-bold po-page-change-password-forgot-password po-clickable\"\r\n                (click)=\"onForgotPasswordClick(recovery)\">\r\n                {{literals.forgotPassword}}\r\n              </a>\r\n            </div>\r\n            <div>\r\n              <po-password\r\n                name=\"newPassword\"\r\n                [(ngModel)]=\"newPassword\"\r\n                p-required\r\n                [p-label]=\"literals.newPassword\"\r\n                (p-change-model)=\"validatePassword()\">\r\n              </po-password>\r\n            </div>\r\n            <div>\r\n              <po-password\r\n                name=\"confirmPassword\"\r\n                [(ngModel)]=\"confirmPassword\"\r\n                p-required\r\n                [p-label]=\"literals.confirmPassword\"\r\n                (p-change-model)=\"validatePassword()\">\r\n              </po-password>\r\n            </div>\r\n          </div>\r\n          <div *ngIf=\"showRequirements\"\r\n            class=\"po-md-6 po-page-change-password-required-container\">\r\n            <po-container>\r\n              <ul class=\"po-page-change-password-required-list\">\r\n                <li class=\"po-page-change-password-required-item\">\r\n                  <p class=\"po-mb-1 po-font-text-bold po-page-change-password-required-title\">\r\n                    {{literals?.requirements}}\r\n                  </p>\r\n                </li>\r\n                <li *ngFor=\"let requirement of requirements\"\r\n                  class=\"po-page-change-password-required-item\">\r\n                  <span class=\"po-icon po-page-change-password-required-icon\"\r\n                    [ngClass]=\"validateRequirement(requirement) ? 'po-icon-ok po-page-change-password-required-ok' : 'po-icon-minus po-page-change-password-required-minus'\">\r\n                  </span>\r\n                  <p class=\"po-page-change-password-required-text po-font-text\">\r\n                    {{requirement.requirement}}\r\n                  </p>\r\n                </li>\r\n              </ul>\r\n            </po-container>\r\n          </div>\r\n        </div>\r\n      </form>\r\n\r\n      <po-divider class=\"po-sm-12\"></po-divider>\r\n\r\n      <div class=\"po-page-change-password-tips po-row\">\r\n        <div class=\"po-md-2 po-mr-3\">\r\n          <img\r\n            class=\"po-page-change-password-lock-image po-mb-3\"\r\n            alt=\"Blocked user image\"\r\n            src=\"./assets/images/big-lock.svg\"\r\n          />\r\n        </div>\r\n        <div class=\"po-md-9 po-page-change-password-tips-text\">\r\n          <div class=\"po-font-text-bold po-pb-1\">{{literals.safetyTips}}</div>\r\n          <div class=\"po-font-text-small\">\r\n            <div class=\"po-row po-pb-1\">{{literals.safetyTipsPhrase}}</div>\r\n            <div class=\"po-row po-pb-1\">\u2022 {{literals.safetyTipsFirst}}</div>\r\n            <div class=\"po-row po-pb-1\">\u2022 {{literals.safetyTipsSecond}}</div>\r\n            <div class=\"po-row po-pb-1\">\u2022 {{literals.safetyTipsThird}}</div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n\r\n      <div class=\"po-page-change-password-buttons po-row\">\r\n\r\n        <po-divider class=\"po-sm-12 po-mobile-only\"></po-divider>\r\n\r\n        <po-button *ngIf=\"!hideCurrentPassword\"\r\n          class=\"po-sm-6 po-mb-sm-1\"\r\n          [p-label]=\"literals.backButton\"\r\n          (p-click)=\"navigateTo(urlBack)\">\r\n        </po-button>\r\n\r\n        <po-button\r\n          [ngClass]=\"hideCurrentPassword ? 'po-md-12':'po-sm-6'\"\r\n          p-type=\"primary\"\r\n          [p-disabled]=\"passwordForm.invalid\"\r\n          [p-label]=\"literals.saveButton\"\r\n          (p-click)=\"onLoginSubmit()\">\r\n        </po-button>\r\n\r\n      </div>\r\n\r\n    </po-page-background>\r\n\r\n  </div>\r\n</div>\r\n\r\n<po-modal\r\n  [p-primary-action]=\"modalAction\"\r\n  p-hide-close\r\n  p-size=\"auto\"\r\n  [p-title]=\"hideCurrentPassword ? literals.passwordSuccessfullyCreated : literals.passwordSuccessfullyUpdated\">\r\n\r\n  <div class=\"po-text-center\">\r\n    <img\r\n      alt=\"Blocked user image\"\r\n      class=\"po-page-change-password-lock-image po-mb-3\"\r\n      src=\"./assets/images/success.svg\"\r\n    />\r\n  </div>\r\n</po-modal>\r\n"
                }] }
    ];
    /** @nocollapse */
    PoPageChangePasswordComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: ActivatedRoute },
        { type: Router },
        { type: PoPageChangePasswordService },
        { type: PoComponentInjectorService },
        { type: ViewContainerRef }
    ]; };
    PoPageChangePasswordComponent.propDecorators = {
        modal: [{ type: ViewChild, args: [PoModalComponent, { static: true },] }],
        pageChangePassword: [{ type: ViewChild, args: ['pageChangePassword', { read: ViewContainerRef, static: true },] }],
        passwordForm: [{ type: ViewChild, args: ['passwordForm', { read: NgForm, static: true },] }]
    };
    return PoPageChangePasswordComponent;
}(PoPageChangePasswordBaseComponent));
export { PoPageChangePasswordComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.newPasswordSubscription;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.literals;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.componentRef;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.modalAction;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.parentRef;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.modal;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.pageChangePassword;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.passwordForm;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.service;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.poComponentInjector;
}
