// @portinari/portinari-ui/esm5/lib/components/po-field/po-rich-text/po-rich-text-toolbar/po-rich-text-toolbar.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { isIE } from '../../../../utils/util';
import { PoLanguageService } from '../../../../services/po-language/po-language.service';
import { poRichTextLiteralsDefault } from '../po-rich-text-literals';
import { PoRichTextModalType } from '../enums/po-rich-text-modal-type.enum';
var poRichTextDefaultColor = '#000000';
var PoRichTextToolbarComponent = /** @class */ (function () {
    function PoRichTextToolbarComponent(languageService) {
        var _this = this;
        this.languageService = languageService;
        this.literals = tslib_1.__assign({}, poRichTextLiteralsDefault[this.languageService.getShortLanguage()]);
        this.alignButtons = [
            {
                command: 'justifyleft',
                icon: 'po-icon-align-left',
                tooltip: this.literals.left,
                action: this.emitAlignCommand.bind(this, 'justifyleft'),
                selected: true
            },
            {
                command: 'justifycenter',
                icon: 'po-icon-align-center',
                tooltip: this.literals.center,
                action: this.emitAlignCommand.bind(this, 'justifycenter')
            },
            {
                command: 'justifyright',
                icon: 'po-icon-align-right',
                tooltip: this.literals.right,
                action: this.emitAlignCommand.bind(this, 'justifyright')
            },
            {
                command: 'justifyfull',
                icon: 'po-icon-align-justify',
                tooltip: this.literals.justify,
                action: this.emitAlignCommand.bind(this, 'justifyfull')
            }
        ];
        this.formatButtons = [
            {
                command: 'bold',
                icon: 'po-icon-text-bold',
                tooltip: this.literals.bold,
                action: this.emitCommand.bind(this, 'bold')
            },
            {
                command: 'italic',
                icon: 'po-icon-text-italic',
                tooltip: this.literals.italic,
                action: this.emitCommand.bind(this, 'italic')
            },
            {
                command: 'underline',
                icon: 'po-icon-text-underline',
                tooltip: this.literals.underline,
                action: this.emitCommand.bind(this, 'underline')
            }
        ];
        this.listButtons = [
            {
                command: 'insertUnorderedList',
                icon: 'po-icon-list',
                tooltip: this.literals.unorderedList,
                action: this.emitCommand.bind(this, 'insertUnorderedList')
            }
        ];
        this.linkButtons = [
            {
                command: 'Createlink',
                icon: 'po-icon-link',
                tooltip: this.literals.insertLink + " (Ctrl + K)",
                action: function () { return _this.modal.emit(PoRichTextModalType.Link); }
            }
        ];
        this.mediaButtons = [
            {
                tooltip: this.literals.insertImage,
                icon: 'po-icon-picture',
                action: function () { return _this.modal.emit(PoRichTextModalType.Image); }
            }
        ];
        this.command = new EventEmitter();
        this.modal = new EventEmitter();
    }
    Object.defineProperty(PoRichTextToolbarComponent.prototype, "readonly", {
        get: function () {
            return this._readonly;
        },
        set: function (value) {
            this._readonly = value;
            this.toggleDisableButtons(this._readonly);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoRichTextToolbarComponent.prototype, "isInternetExplorer", {
        get: function () {
            return isIE();
        },
        enumerable: true,
        configurable: true
    });
    PoRichTextToolbarComponent.prototype.ngAfterViewInit = function () {
        this.removeButtonFocus();
        this.setColorInColorPicker(poRichTextDefaultColor);
    };
    PoRichTextToolbarComponent.prototype.changeTextColor = function (value) {
        var command = 'foreColor';
        this.command.emit({ command: command, value: value });
    };
    PoRichTextToolbarComponent.prototype.setButtonsStates = function (obj) {
        if (!this.readonly) {
            this.alignButtons.forEach(function (button) { return button.selected = obj.commands.includes(button.command); });
            this.formatButtons.forEach(function (button) { return button.selected = obj.commands.includes(button.command); });
            this.listButtons[0].selected = obj.commands.includes(this.listButtons[0].command);
            this.linkButtons[0].selected = obj.commands.includes(this.linkButtons[0].command);
            this.setColorInColorPicker(obj.hexColor);
        }
    };
    PoRichTextToolbarComponent.prototype.shortcutTrigger = function () {
        this.modal.emit(PoRichTextModalType.Link);
    };
    PoRichTextToolbarComponent.prototype.emitAlignCommand = function (command) {
        var index = this.alignButtons.findIndex(function (btn) { return btn.command === command; });
        if (this.alignButtons[index].selected) {
            this.alignButtons[index].selected = false;
        }
        this.command.emit(command);
    };
    PoRichTextToolbarComponent.prototype.emitCommand = function (command) {
        this.command.emit(command);
    };
    PoRichTextToolbarComponent.prototype.removeButtonFocus = function () {
        var buttons = this.toolbarElement.nativeElement.querySelectorAll('button');
        buttons.forEach(function (button) { return button.setAttribute('tabindex', '-1'); });
    };
    PoRichTextToolbarComponent.prototype.setColorInColorPicker = function (color) {
        this.colorPickerInput.nativeElement.value = color;
    };
    PoRichTextToolbarComponent.prototype.toggleDisableButtons = function (state) {
        this.alignButtons.forEach(function (button) { return button.disabled = state; });
        this.formatButtons.forEach(function (button) { return button.disabled = state; });
        this.listButtons[0].disabled = state;
        this.linkButtons[0].disabled = state;
        this.mediaButtons[0].disabled = state;
    };
    PoRichTextToolbarComponent.ctorParameters = function () { return [
        { type: PoLanguageService }
    ]; };
    tslib_1.__decorate([
        ViewChild('colorPickerInput', { read: ElementRef, static: false }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoRichTextToolbarComponent.prototype, "colorPickerInput", void 0);
    tslib_1.__decorate([
        ViewChild('toolbarElement', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoRichTextToolbarComponent.prototype, "toolbarElement", void 0);
    tslib_1.__decorate([
        Input('p-readonly'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoRichTextToolbarComponent.prototype, "readonly", null);
    tslib_1.__decorate([
        Output('p-command'),
        tslib_1.__metadata("design:type", Object)
    ], PoRichTextToolbarComponent.prototype, "command", void 0);
    tslib_1.__decorate([
        Output('p-modal'),
        tslib_1.__metadata("design:type", Object)
    ], PoRichTextToolbarComponent.prototype, "modal", void 0);
    PoRichTextToolbarComponent = tslib_1.__decorate([
        Component({
            selector: 'po-rich-text-toolbar',
            template: "<div class=\"po-rich-text-toolbar\" #toolbarElement>\n  <div class=\"po-rich-text-toolbar-button-align\">\n    <po-button-group p-toggle=\"multiple\" [p-buttons]=\"formatButtons\">\n    </po-button-group>\n  </div>\n\n  <div *ngIf=\"!isInternetExplorer\" class=\"po-rich-text-toolbar-button-align\">\n    <div class=\"po-rich-text-toolbar-color-picker-container\">\n      <button class=\"po-button po-text-ellipsis po-rich-text-toolbar-color-picker-button\"\n        [disabled]=\"readonly\"\n        [p-tooltip]=\"literals.textColor\">\n        <input\n          #colorPickerInput\n          class=\"po-rich-text-toolbar-color-picker-input\"\n          type=\"color\"\n          [disabled]=\"readonly\"\n          (change)=\"changeTextColor($event.target.value)\">\n      </button>\n    </div>\n  </div>\n\n  <div class=\"po-rich-text-toolbar-button-align\">\n    <po-button-group p-toggle=\"single\" [p-buttons]=\"alignButtons\">\n    </po-button-group>\n  </div>\n\n  <div class=\"po-rich-text-toolbar-button-align\">\n    <po-button-group p-toggle=\"single\" [p-buttons]=\"listButtons\">\n    </po-button-group>\n  </div>\n\n  <div class=\"po-rich-text-toolbar-button-align\">\n    <po-button-group [p-buttons]=\"linkButtons\">\n    </po-button-group>\n  </div>\n\n  <div class=\"po-rich-text-toolbar-button-align\">\n    <po-button-group [p-buttons]=\"mediaButtons\">\n    </po-button-group>\n  </div>\n\n</div>\n\n\n"
        }),
        tslib_1.__metadata("design:paramtypes", [PoLanguageService])
    ], PoRichTextToolbarComponent);
    return PoRichTextToolbarComponent;
}());
export { PoRichTextToolbarComponent };
