// @portinari/portinari-ui/esm5/lib/components/po-field/po-checkbox/po-checkbox-base.component.js
import * as tslib_1 from "tslib";
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean, uuid } from './../../../utils/util';
import { InputBoolean } from '../../../decorators';
/**
 * @description
 *
 * O componente `po-checkbox` exibe uma caixa de opção com um texto ao lado, na qual é possível marcar e desmarcar através tanto
 * no *click* do *mouse* quanto por meio da tecla *space* quando estiver com foco.
 *
 * Cada opção poderá receber um estado de marcado, desmarcado, indeterminado e desabilitado, como também uma ação que será disparada quando
 * ocorrer mudanças do valor.
 *
 * > O *model* deste componente aceitará valores igual à `true`, `false` ou `null` para quando for indeterminado.
 */
var PoCheckboxBaseComponent = /** @class */ (function () {
    function PoCheckboxBaseComponent() {
        this.id = uuid();
        this._disabled = false;
        /**
         * @optional
         *
         * @description
         *
         * Aplica foco no elemento ao ser iniciado.
         *
         * > Caso mais de um elemento seja configurado com essa propriedade, apenas o último elemento declarado com ela terá o foco.
         *
         * @default `false`
         */
        this.autoFocus = false;
        /** Evento disparado quando o valor do *checkbox* for alterado. */
        this.change = new EventEmitter();
    }
    Object.defineProperty(PoCheckboxBaseComponent.prototype, "disabled", {
        get: function () {
            return this._disabled;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define o estado do *checkbox* como desabilitado.
         *
         * @default `false`
         */
        set: function (value) {
            this._disabled = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    PoCheckboxBaseComponent.prototype.changeValue = function () {
        if (this.propagateChange) {
            this.propagateChange(this.checkboxValue);
        }
        this.change.emit(this.checkboxValue);
    };
    PoCheckboxBaseComponent.prototype.checkOption = function (value) {
        if (!this.disabled) {
            this.changeModelValue(!value);
            this.changeValue();
        }
    };
    PoCheckboxBaseComponent.prototype.registerOnChange = function (fn) {
        this.propagateChange = fn;
    };
    PoCheckboxBaseComponent.prototype.registerOnTouched = function (fn) { };
    PoCheckboxBaseComponent.prototype.writeValue = function (value) {
        if (value !== this.checkboxValue) {
            this.changeModelValue(value);
        }
    };
    tslib_1.__decorate([
        Input('name'),
        tslib_1.__metadata("design:type", String)
    ], PoCheckboxBaseComponent.prototype, "name", void 0);
    tslib_1.__decorate([
        Input('p-auto-focus'), InputBoolean(),
        tslib_1.__metadata("design:type", Boolean)
    ], PoCheckboxBaseComponent.prototype, "autoFocus", void 0);
    tslib_1.__decorate([
        Input('p-disabled'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoCheckboxBaseComponent.prototype, "disabled", null);
    tslib_1.__decorate([
        Input('p-label'),
        tslib_1.__metadata("design:type", String)
    ], PoCheckboxBaseComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Output('p-change'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoCheckboxBaseComponent.prototype, "change", void 0);
    return PoCheckboxBaseComponent;
}());
export { PoCheckboxBaseComponent };
