// @portinari/portinari-ui/esm5/lib/pipes/po-time/po-time.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { isTypeof } from '../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Pipe responsável por apresentar um horário baseado no formato de entrada definido na API da Portinari.
 */
var PoTimePipe = /** @class */ (function () {
    function PoTimePipe() {
    }
    PoTimePipe.prototype.transform = function (time, format) {
        var hourRegex = /^(([0-1][0-9])|(2[0-3])):[0-5][0-9]:[0-5][0-9][\.]?([0-9]{1,6})?$/g;
        if (isTypeof(time, 'string') && hourRegex.test(time)) {
            var amountOfF = format ? format.lastIndexOf('f') - format.indexOf('f') + 1 : 0;
            var miliseconds = this.addDotMiliseconds(time.substring(9, 9 + amountOfF), amountOfF);
            var formatMiliseconds = this.getFormatMiliseconds(amountOfF);
            return this.formatValue(time, format, formatMiliseconds, miliseconds);
        }
        return null;
    };
    PoTimePipe.prototype.addDotMiliseconds = function (miliseconds, amountOfF) {
        if (miliseconds && amountOfF) {
            miliseconds = '.' + miliseconds;
        }
        return miliseconds;
    };
    PoTimePipe.prototype.formatValue = function (time, format, formatMiliseconds, miliseconds) {
        if (time && format) {
            var hour = time.substring(0, 2);
            var minutes = time.substring(3, 5);
            var seconds = time.substring(6, 8);
            format = format.replace('HH', hour);
            format = format.replace('mm', minutes);
            format = format.replace('ss', seconds);
            format = format.replace("." + formatMiliseconds, miliseconds);
            return format;
        }
        else {
            return time;
        }
    };
    PoTimePipe.prototype.getFormatMiliseconds = function (amountOfF) {
        var formatMiliseconds = '';
        for (var i = 0; i < amountOfF; i++) {
            formatMiliseconds += 'f';
        }
        return formatMiliseconds;
    };
    PoTimePipe = tslib_1.__decorate([
        Pipe({
            name: 'po_time'
        })
    ], PoTimePipe);
    return PoTimePipe;
}());
export { PoTimePipe };
