// @portinari/portinari-ui/esm2015/lib/components/po-page/po-page-detail/po-page-detail-base.component.js
import * as tslib_1 from "tslib";
import { Input, ViewChild } from '@angular/core';
import { browserLanguage, poLocaleDefault } from './../../../utils/util';
import { PoPageContentComponent } from '../po-page-content/po-page-content.component';
export const poPageDetailLiteralsDefault = {
    en: {
        back: 'Back',
        edit: 'Edit',
        remove: 'Remove'
    },
    es: {
        back: 'Volver',
        edit: 'Editar',
        remove: 'Eliminar'
    },
    pt: {
        back: 'Voltar',
        edit: 'Editar',
        remove: 'Remover'
    },
    ru: {
        back: 'возвращение',
        edit: 'редактировать',
        remove: 'удаление'
    }
};
/**
 * @description
 *
 * O componente **po-page-detail** é utilizado como container principal para a tela de detalhamento de um registro.
 * Por padrão possui 3 ações, cada ação na tela executa uma função no componente que está utilizando o po-page-detail,
 * são elas:
 *  - Voltar (função: back);
 *  - Editar (função: edit);
 *  - Remover (função: remove);
 *
 * Caso não estiver implementado alguma função, listado anteriormente, o mesmo não será apresentado.
 */
export class PoPageDetailBaseComponent {
    /**
     * @optional
     *
     * @description
     *
     * Objeto com as literais usadas no `po-page-detail`.
     *
     * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
     *
     * ```
     *  const customLiterals: PoPageDetailLiterals = {
     *    edit: 'Edição',
     *    remove: 'Exclusão',
     *    back: 'Menu'
     *  };
     * ```
     *
     * Ou passando apenas as literais que deseja customizar:
     *
     * ```
     *  const customLiterals: PoPageDetailLiterals = {
     *    remove: 'Excluir registro permanentemente'
     *  };
     * ```
     *
     * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
     *
     * ```
     * <po-page-detail
     *   [p-literals]="customLiterals">
     * </po-page-detail>
     * ```
     *
     *  > O objeto padrão de literais será traduzido de acordo com o idioma do browser (pt, en, es).
     */
    set literals(value) {
        if (value instanceof Object && !(value instanceof Array)) {
            this._literals = Object.assign({}, poPageDetailLiteralsDefault[poLocaleDefault], poPageDetailLiteralsDefault[browserLanguage()], value);
        }
        else {
            this._literals = poPageDetailLiteralsDefault[browserLanguage()];
        }
    }
    get literals() {
        return this._literals || poPageDetailLiteralsDefault[browserLanguage()];
    }
    /** Título da página. */
    set title(title) {
        this._title = title;
        this.poPageContent.recalculateHeaderSize();
    }
    get title() {
        return this._title;
    }
}
tslib_1.__decorate([
    ViewChild(PoPageContentComponent, { static: true }),
    tslib_1.__metadata("design:type", PoPageContentComponent)
], PoPageDetailBaseComponent.prototype, "poPageContent", void 0);
tslib_1.__decorate([
    Input('p-breadcrumb'),
    tslib_1.__metadata("design:type", Object)
], PoPageDetailBaseComponent.prototype, "breadcrumb", void 0);
tslib_1.__decorate([
    Input('p-literals'),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], PoPageDetailBaseComponent.prototype, "literals", null);
tslib_1.__decorate([
    Input('p-title'),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], PoPageDetailBaseComponent.prototype, "title", null);
