// @portinari/portinari-templates/esm2015/lib/components/po-page-background/po-page-background.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, convertToBoolean, isTypeof } from './../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para definição de cor de fundo e dos logotipos primário e secundário para os templates
 * de `po-page-login` e demais templates de login.
 */
export class PoPageBackgroundComponent {
    constructor() {
        this._showSelectLanguage = false;
        this.selectLanguageOptions = [
            { label: 'English', value: 'en' },
            { label: 'Español', value: 'es' },
            { label: 'Português', value: 'pt' },
            { label: 'Pусский', value: 'ru' }
        ];
        /**
         * Evento disparado ao selecionar alguma opção no seletor de idiomas.
         * Para este evento será passado como parâmetro o valor de idioma selecionado.
         */
        this.selectedLanguage = new EventEmitter();
    }
    /**
     * Caminho para a logomarca localizada na parte superior.
     * @param {?} value
     * @return {?}
     */
    set logo(value) {
        this._logo = isTypeof(value, 'string') && value.trim() ? value : undefined;
    }
    /**
     * @return {?}
     */
    get logo() {
        return this._logo;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Caminho para a logomarca localizada no rodapé.
     * @param {?} value
     * @return {?}
     */
    set secondaryLogo(value) {
        this._secondaryLogo = isTypeof(value, 'string') && value.trim() ? value : undefined;
    }
    /**
     * @return {?}
     */
    get secondaryLogo() {
        return this._secondaryLogo;
    }
    /**
     * Define se o seletor de idiomas deve ser exibido.
     * @param {?} showSelectLanguage
     * @return {?}
     */
    set showSelectLanguage(showSelectLanguage) {
        this._showSelectLanguage = convertToBoolean(showSelectLanguage);
    }
    /**
     * @return {?}
     */
    get showSelectLanguage() {
        return this._showSelectLanguage;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.selectedLanguageOption = browserLanguage();
    }
    /**
     * @return {?}
     */
    onChangeLanguage() {
        this.selectedLanguage.emit(this.selectedLanguageOption);
    }
}
PoPageBackgroundComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-page-background',
                template: "<div class=\"po-page-login-container\">\r\n  <div class=\"po-page-login-panel\">\r\n    <img *ngIf=\"logo\"\r\n      class=\"po-page-background-main-logo\"\r\n      [class.po-page-background-hide-logo]=\"hideLogo\"\r\n      alt=\"main-logo\"\r\n      [src]=\"logo\">\r\n\r\n    <div class=\"po-page-login-body\">\r\n      <div class=\"po-page-login-panel-content\">\r\n        <ng-content></ng-content>\r\n      </div>\r\n    </div>\r\n    <div class=\"po-page-background-footer po-sm-12\">\r\n\r\n      <po-divider class=\"po-page-background-footer-mobile-only\"></po-divider>\r\n\r\n      <div class=\"po-page-background-footer-content\">\r\n        <div *ngIf=\"showSelectLanguage\" class=\"po-page-background-footer-select\">\r\n          <po-select\r\n            name=\"selectedLanguageOption\"\r\n            [(ngModel)]=\"selectedLanguageOption\"\r\n            [p-options]=\"selectLanguageOptions\"\r\n            (p-change) = \"onChangeLanguage()\">\r\n          </po-select>\r\n        </div>\r\n\r\n        <div class=\"po-page-background-secondary-logo\"\r\n          [ngClass]=\"showSelectLanguage ? 'po-page-background-secondary-logo-right' : 'po-page-background-secondary-logo-centered'\">\r\n\r\n          <img *ngIf=\"secondaryLogo\"\r\n            class=\"po-page-background-secondary-logo-image\"\r\n            alt=\"secondary-logo\"\r\n            [src]=\"secondaryLogo\">\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n\r\n  <div *ngIf=\"background\"\r\n    [ngClass]=\"background ? 'po-page-login-highlight-image' : 'po-page-login-highlight-image-off'\"\r\n    [style.background-image]=\"'url(' + background + ')'\">\r\n    <div class=\"po-page-login-highlight-text\">\r\n      <div class=\"po-font-display\">{{ highlightInfo }}</div>\r\n    </div>\r\n  </div>\r\n</div>\r\n"
            }] }
];
PoPageBackgroundComponent.propDecorators = {
    background: [{ type: Input, args: ['p-background',] }],
    hideLogo: [{ type: Input, args: ['p-hide-logo',] }],
    highlightInfo: [{ type: Input, args: ['p-highlight-info',] }],
    logo: [{ type: Input, args: ['p-logo',] }],
    secondaryLogo: [{ type: Input, args: ['p-secondary-logo',] }],
    showSelectLanguage: [{ type: Input, args: ['p-show-select-language',] }],
    selectedLanguage: [{ type: Output, args: ['p-selected-language',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageBackgroundComponent.prototype._logo;
    /**
     * @type {?}
     * @private
     */
    PoPageBackgroundComponent.prototype._secondaryLogo;
    /**
     * @type {?}
     * @private
     */
    PoPageBackgroundComponent.prototype._showSelectLanguage;
    /** @type {?} */
    PoPageBackgroundComponent.prototype.selectedLanguageOption;
    /** @type {?} */
    PoPageBackgroundComponent.prototype.selectLanguageOptions;
    /**
     * Insere uma imagem de destaque ao lado direito do container.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.background;
    /**
     * Designa se o logotipo deve desaparecer em resoluções menores.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.hideLogo;
    /**
     * Texto de destaque sobreposto à imagem de destaque. Essa opção é utilizada em conjunto com o atributo `p-background`.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.highlightInfo;
    /**
     * Evento disparado ao selecionar alguma opção no seletor de idiomas.
     * Para este evento será passado como parâmetro o valor de idioma selecionado.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.selectedLanguage;
}
