// @portinari/portinari-templates/esm2015/lib/components/po-page-blocked-user/po-page-blocked-user-contacts/po-page-blocked-user-contacts.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, Input, ViewChild } from '@angular/core';
/** @type {?} */
const poPageBlockedUserContactItemMargin = 16;
export class PoPageBlockedUserContactsComponent {
    /**
     * @param {?} changeDetector
     */
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.overflowItem = false;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.checkContactItemWidth();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.email || changes.phone) {
            this.checkContactItemWidth();
        }
    }
    /**
     * @private
     * @return {?}
     */
    checkContactItemWidth() {
        if (!this.email || !this.phone) {
            this.overflowItem = true;
            return;
        }
        else {
            setTimeout((/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const phoneWidth = this.phoneItem.nativeElement.offsetWidth;
                /** @type {?} */
                const mailWidth = this.mailItem.nativeElement.offsetWidth;
                /** @type {?} */
                const contactGroupHalfWidth = this.contactGroup.nativeElement.offsetWidth / 2;
                this.overflowItem = phoneWidth > contactGroupHalfWidth || mailWidth > contactGroupHalfWidth - poPageBlockedUserContactItemMargin;
            }));
        }
        this.changeDetector.detectChanges();
    }
}
PoPageBlockedUserContactsComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-page-blocked-user-contacts',
                template: "<div #contactGroup class=\"po-page-blocked-user-contact-group po-row po-mb-3\">\r\n  <div *ngIf=\"phone\"\r\n    class=\"po-page-blocked-user-contact-item\"\r\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\">\r\n    <a\r\n      class=\"po-page-blocked-user-link po-clickable\"\r\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\r\n      href=\"tel:{{ phone }}\"\r\n      target=\"_self\">\r\n      <div class=\"po-page-blocked-user-contact-group-item\">\r\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-telephone po-pr-1\"></span>\r\n        <span #phoneItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ phone }}</span>\r\n      </div>\r\n    </a>\r\n  </div>\r\n  <div *ngIf=\"email\"\r\n    class=\"po-page-blocked-user-contact-item po-page-blocked-user-contact-mail\"\r\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\">\r\n    <a\r\n      class=\"po-page-blocked-user-link po-clickable\"\r\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\r\n      href=\"mailto:{{ email }}\"\r\n      target=\"_self\">\r\n      <div class=\"po-page-blocked-user-contact-group-item\">\r\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-mail po-pr-1 \"></span>\r\n        <span #mailItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ email }}</span>\r\n      </div>\r\n    </a>\r\n  </div>\r\n</div>"
            }] }
];
/** @nocollapse */
PoPageBlockedUserContactsComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PoPageBlockedUserContactsComponent.propDecorators = {
    email: [{ type: Input, args: ['p-email',] }],
    phone: [{ type: Input, args: ['p-phone',] }],
    contactGroup: [{ type: ViewChild, args: ['contactGroup', { static: true },] }],
    mailItem: [{ type: ViewChild, args: ['mailItem', { static: true },] }],
    phoneItem: [{ type: ViewChild, args: ['phoneItem', { static: true },] }]
};
if (false) {
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.literals;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.overflowItem;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserContactsComponent.prototype.mailText;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserContactsComponent.prototype.phoneText;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.email;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.phone;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.contactGroup;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.mailItem;
    /** @type {?} */
    PoPageBlockedUserContactsComponent.prototype.phoneItem;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserContactsComponent.prototype.changeDetector;
}
