// @portinari/portinari-ui/esm5/lib/components/po-menu/po-menu-filter/po-menu-filter.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que apresenta o campo de pesquisa no po-menu.
 */
var PoMenuFilterComponent = /** @class */ (function () {
    function PoMenuFilterComponent() {
        // Variável necessária para o po-clean identificar que deve ser criado.
        this.clean = true;
        this.filter = new EventEmitter();
    }
    PoMenuFilterComponent.prototype.filterItems = function (search) {
        this.filter.emit(search);
    };
    tslib_1.__decorate([
        Input('p-loading'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoMenuFilterComponent.prototype, "loading", void 0);
    tslib_1.__decorate([
        ViewChild('inputFilter', { read: ElementRef, static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoMenuFilterComponent.prototype, "inputFilterElement", void 0);
    tslib_1.__decorate([
        Output('p-filter'),
        tslib_1.__metadata("design:type", Object)
    ], PoMenuFilterComponent.prototype, "filter", void 0);
    PoMenuFilterComponent = tslib_1.__decorate([
        Component({
            selector: 'po-menu-filter',
            template: "<div class=\"po-menu-filter-container\">\n\n  <input #inputFilter\n    type=\"text\"\n    class=\"po-menu-filter\"\n    placeholder=\"Pesquisar\"\n    (keyup)=\"filterItems(inputFilter.value)\">\n\n  <div class=\"po-menu-filter-search-icon-container\">\n    <span *ngIf=\"!loading\" class=\"po-icon po-menu-filter-icon po-icon-search\"></span>\n    <po-loading-icon *ngIf=\"loading\" ></po-loading-icon>\n  </div>\n  <div class=\"po-menu-filter-close-icon-container\">\n    <po-clean [p-element-ref]=\"inputFilterElement\" (p-change-event)=\"filterItems(inputFilter.value)\"></po-clean>\n  </div>\n\n</div>\n"
        })
    ], PoMenuFilterComponent);
    return PoMenuFilterComponent;
}());
export { PoMenuFilterComponent };
