// @portinari/portinari-ui/esm2015/lib/components/po-page/po-page-default/po-page-default.component.js
import * as tslib_1 from "tslib";
import { AfterContentInit, Component, OnChanges, Renderer2, SimpleChange, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { callFunction, isTypeof } from '../../../utils/util';
import { PoPageDefaultBaseComponent } from './po-page-default-base.component';
/**
 * @docsExtends PoPageDefaultBaseComponent
 *
 * @example
 *
 * <example name="po-page-default-basic" title="Portinari Page Default Basic">
 *  <file name="sample-po-page-default-basic/sample-po-page-default-basic.component.html"> </file>
 *  <file name="sample-po-page-default-basic/sample-po-page-default-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-default-labs" title="Portinari Page Default Labs">
 *  <file name="sample-po-page-default-labs/sample-po-page-default-labs.component.html"> </file>
 *  <file name="sample-po-page-default-labs/sample-po-page-default-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-default-dashboard" title="Portinari Page Default - Dashboard">
 *  <file name="sample-po-page-default-dashboard/sample-po-page-default-dashboard.component.html"> </file>
 *  <file name="sample-po-page-default-dashboard/sample-po-page-default-dashboard.component.ts"> </file>
 *  <file name="sample-po-page-default-dashboard/sample-po-page-default-dashboard.service.ts"> </file>
 * </example>
 */
let PoPageDefaultComponent = class PoPageDefaultComponent extends PoPageDefaultBaseComponent {
    constructor(viewRef, renderer, router) {
        super();
        this.renderer = renderer;
        this.router = router;
        this.limitPrimaryActions = 3;
        this.maxWidthMobile = 480;
        this.parentRef = viewRef['_view']['component'];
    }
    ngAfterContentInit() {
        this.setIsMobile();
        this.setDropdownActions();
        this.renderer.listen('window', 'resize', (event) => {
            this.onResize(event);
        });
    }
    ngOnChanges(changes) {
        this.setDropdownActions();
    }
    actionIsDisabled(action) {
        return isTypeof(action.disabled, 'function') ? action.disabled(action) : action.disabled;
    }
    callAction(item) {
        if (item.url) {
            this.router.navigate([item.url]);
        }
        else if (item.action) {
            callFunction(item.action, this.parentRef);
        }
    }
    hasPageHeader() {
        return !!(this.title || (this.actions && this.actions.length) || (this.breadcrumb && this.breadcrumb.items.length));
    }
    onResize(event) {
        const width = event.target.innerWidth;
        if (width < this.maxWidthMobile) {
            this.isMobile = true;
            this.limitPrimaryActions = 2;
            this.setDropdownActions();
        }
        else {
            this.isMobile = false;
            this.limitPrimaryActions = 3;
            this.setDropdownActions();
        }
    }
    setIsMobile() {
        if (window.innerWidth < this.maxWidthMobile) {
            this.isMobile = true;
            this.limitPrimaryActions = 2;
        }
    }
    setDropdownActions() {
        if (this.actions.length > this.limitPrimaryActions) {
            this.dropdownActions = this.actions.slice(this.limitPrimaryActions - 1);
        }
    }
};
PoPageDefaultComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Renderer2 },
    { type: Router }
];
PoPageDefaultComponent = tslib_1.__decorate([
    Component({
        selector: 'po-page-default',
        template: "<po-page>\n\n  <!-- HEADER -->\n  <po-page-header\n    *ngIf=\"hasPageHeader()\"\n    [p-breadcrumb]=\"breadcrumb\"\n    [p-title]=\"title\">\n\n    <!-- OPERATIONS -->\n    <div class=\"po-page-header-actions\">\n      <po-dropdown *ngIf=\"actions.length > limitPrimaryActions\"\n        p-label=\"Outras a\u00E7\u00F5es\"\n        [p-actions]=\"dropdownActions\">\n      </po-dropdown>\n\n      <po-button *ngIf=\"actions.length === 3 && actions[2] && !isMobile\"\n        [p-disabled]=\"actionIsDisabled(actions[2])\"\n        [p-label]=\"actions[2].label\"\n        (p-click)=\"callAction(actions[2])\">\n      </po-button>\n\n      <po-button *ngIf=\"actions[1] && (actions.length === 2 || !isMobile)\"\n        [p-disabled]=\"actionIsDisabled(actions[1])\"\n        [p-label]=\"actions[1].label\"\n        (p-click)=\"callAction(actions[1])\">\n      </po-button>\n\n      <po-button *ngIf=\"actions[0]\"\n        p-type=\"primary\"\n        [p-disabled]=\"actionIsDisabled(actions[0])\"\n        [p-icon]=\"actions[0].icon\"\n        [p-label]=\"actions[0].label\"\n        (p-click)=\"callAction(actions[0])\">\n      </po-button>\n    </div>\n\n  </po-page-header>\n\n  <!-- CONTENT -->\n  <po-page-content>\n    <ng-content></ng-content>\n  </po-page-content>\n\n</po-page>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [ViewContainerRef,
        Renderer2,
        Router])
], PoPageDefaultComponent);
export { PoPageDefaultComponent };
