// @portinari/portinari-ui/esm5/lib/components/po-field/po-multiselect/po-multiselect-search/po-multiselect-search.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente de pesquisa que será criado dentro do dropdown do `po-multiselect`.
 */
var PoMultiselectSearchComponent = /** @class */ (function () {
    function PoMultiselectSearchComponent() {
        /** Evento que será disparado a cada tecla digitada no campo de busca. */
        this.change = new EventEmitter();
    }
    Object.defineProperty(PoMultiselectSearchComponent.prototype, "placeholder", {
        get: function () {
            return this._placeholder || this.literals.placeholderSearch;
        },
        /**
         * @optional
         *
         * @description
         *
         * Placeholder do campo de pesquisa.
         *
         * > Caso o mesmo não seja informado, o valor padrão será traduzido com base no idioma do navegador (pt, es e en).
         *
         * @default `Buscar`
         */
        set: function (placeholder) {
            this._placeholder = placeholder && isTypeof(placeholder, 'string') ? placeholder : this.literals.placeholderSearch;
        },
        enumerable: true,
        configurable: true
    });
    PoMultiselectSearchComponent.prototype.onChange = function (event) {
        this.change.emit({ event: event, value: this.inputElement.nativeElement.value });
    };
    PoMultiselectSearchComponent.prototype.setFocus = function () {
        this.inputElement.nativeElement.focus();
    };
    PoMultiselectSearchComponent.prototype.clean = function () {
        this.inputElement.nativeElement.value = '';
    };
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object)
    ], PoMultiselectSearchComponent.prototype, "literals", void 0);
    tslib_1.__decorate([
        Input('p-placeholder'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoMultiselectSearchComponent.prototype, "placeholder", null);
    tslib_1.__decorate([
        Output('p-change'),
        tslib_1.__metadata("design:type", Object)
    ], PoMultiselectSearchComponent.prototype, "change", void 0);
    tslib_1.__decorate([
        ViewChild('inputElement', { read: ElementRef, static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoMultiselectSearchComponent.prototype, "inputElement", void 0);
    PoMultiselectSearchComponent = tslib_1.__decorate([
        Component({
            selector: 'po-multiselect-search',
            template: "<div class=\"po-field-container-content po-multiselect-container-search\">\n\n  <div class=\"po-field-icon-container-left\">\n    <span class=\"po-icon po-icon-search po-field-icon\"></span>\n  </div>\n\n  <input #inputElement\n    class=\"po-input po-input-icon-left\"\n    type=\"text\"\n    [placeholder]=\"placeholder\"\n    (keyup)=\"onChange($event)\">\n\n</div>\n"
        })
    ], PoMultiselectSearchComponent);
    return PoMultiselectSearchComponent;
}());
export { PoMultiselectSearchComponent };
