// @portinari/portinari-ui/esm5/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area/po-upload-drag-drop-area.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
var PoUploadDragDropAreaComponent = /** @class */ (function () {
    function PoUploadDragDropAreaComponent(elementRef) {
        this.elementRef = elementRef;
        this.selectFiles = new EventEmitter();
    }
    PoUploadDragDropAreaComponent.prototype.focus = function () {
        this.selectFilesLinkElement.nativeElement.focus();
    };
    PoUploadDragDropAreaComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    tslib_1.__decorate([
        Input('p-directory-compatible'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoUploadDragDropAreaComponent.prototype, "directoryCompatible", void 0);
    tslib_1.__decorate([
        Input('p-disabled'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoUploadDragDropAreaComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        Input('p-height'),
        tslib_1.__metadata("design:type", Number)
    ], PoUploadDragDropAreaComponent.prototype, "height", void 0);
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object)
    ], PoUploadDragDropAreaComponent.prototype, "literals", void 0);
    tslib_1.__decorate([
        Input('p-overlay'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoUploadDragDropAreaComponent.prototype, "overlay", void 0);
    tslib_1.__decorate([
        Output('p-select-files'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoUploadDragDropAreaComponent.prototype, "selectFiles", void 0);
    tslib_1.__decorate([
        ViewChild('selectFilesLink', { static: false }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoUploadDragDropAreaComponent.prototype, "selectFilesLinkElement", void 0);
    PoUploadDragDropAreaComponent = tslib_1.__decorate([
        Component({
            selector: 'po-upload-drag-drop-area',
            template: "<div\r\n  class=\"po-upload-drag-drop-area\"\r\n  [ngClass]=\"{'po-upload-drag-drop-area-disabled': disabled}\"\r\n  [style.height.px]=\"height\">\r\n\r\n  <div class=\"po-upload-drag-drop-area-container\">\r\n    <ng-container *ngIf=\"overlay && !disabled; then overlayTemplate; else defaultTemplate\">\r\n    </ng-container>\r\n  </div>\r\n\r\n</div>\r\n\r\n<ng-template #overlayTemplate>\r\n  <span class=\"po-upload-drag-drop-area-overlay-icon po-icon po-icon-upload-cloud\"></span>\r\n\r\n  <div class=\"po-upload-drag-drop-area-overlay-label\">{{ directoryCompatible ? literals?.dropFoldersHere : literals?.dropFilesHere }}</div>\r\n</ng-template>\r\n\r\n<ng-template #defaultTemplate>\r\n  <span class=\"po-upload-drag-drop-area-icon po-icon po-icon-upload-cloud\"></span>\r\n\r\n  <div class=\"po-upload-drag-drop-area-label\">{{ directoryCompatible ? literals?.dragFoldersHere : literals?.dragFilesHere }}</div>\r\n\r\n  <button #selectFilesLink\r\n    class=\"po-upload-drag-drop-area-select-files\"\r\n    [disabled]=\"disabled\"\r\n    [ngClass]=\"{'po-clickable': !disabled}\"\r\n    (click)=\"selectFiles.emit()\">{{ directoryCompatible ? literals?.selectFolderOnComputer : literals?.selectFilesOnComputer }}</button>\r\n</ng-template>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef])
    ], PoUploadDragDropAreaComponent);
    return PoUploadDragDropAreaComponent;
}());
export { PoUploadDragDropAreaComponent };
