// @portinari/portinari-templates/esm5/lib/components/po-page-background/po-page-background.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, convertToBoolean, isTypeof } from './../../utils/util';
var PoPageBackgroundComponent = /** @class */ (function () {
    function PoPageBackgroundComponent() {
        this._showSelectLanguage = false;
        this.selectLanguageOptions = [
            { label: 'English', value: 'en' },
            { label: 'Español', value: 'es' },
            { label: 'Português', value: 'pt' },
            { label: 'Pусский', value: 'ru' }
        ];
        /**
         * Evento disparado ao selecionar alguma opção no seletor de idiomas.
         * Para este evento será passado como parâmetro o valor de idioma selecionado.
         */
        this.selectedLanguage = new EventEmitter();
    }
    Object.defineProperty(PoPageBackgroundComponent.prototype, "logo", {
        get: /**
         * @return {?}
         */
        function () {
            return this._logo;
        },
        /** Caminho para a logomarca localizada na parte superior. */
        set: /**
         * Caminho para a logomarca localizada na parte superior.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._logo = isTypeof(value, 'string') && value.trim() ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageBackgroundComponent.prototype, "secondaryLogo", {
        get: /**
         * @return {?}
         */
        function () {
            return this._secondaryLogo;
        },
        /**
         * @optional
         *
         * @description
         *
         * Caminho para a logomarca localizada no rodapé.
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Caminho para a logomarca localizada no rodapé.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._secondaryLogo = isTypeof(value, 'string') && value.trim() ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageBackgroundComponent.prototype, "showSelectLanguage", {
        get: /**
         * @return {?}
         */
        function () {
            return this._showSelectLanguage;
        },
        /** Define se o seletor de idiomas deve ser exibido. */
        set: /**
         * Define se o seletor de idiomas deve ser exibido.
         * @param {?} showSelectLanguage
         * @return {?}
         */
        function (showSelectLanguage) {
            this._showSelectLanguage = convertToBoolean(showSelectLanguage);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoPageBackgroundComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.selectedLanguageOption = browserLanguage();
    };
    /**
     * @return {?}
     */
    PoPageBackgroundComponent.prototype.onChangeLanguage = /**
     * @return {?}
     */
    function () {
        this.selectedLanguage.emit(this.selectedLanguageOption);
    };
    PoPageBackgroundComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-background',
                    template: "<div class=\"po-page-login-container\">\n  <div class=\"po-page-login-panel\">\n    <img *ngIf=\"logo\"\n      class=\"po-page-background-main-logo\"\n      [class.po-page-background-hide-logo]=\"hideLogo\"\n      alt=\"main-logo\"\n      [src]=\"logo\">\n\n    <div class=\"po-page-login-body\">\n      <div class=\"po-page-login-panel-content\">\n        <ng-content></ng-content>\n      </div>\n    </div>\n    <div class=\"po-page-background-footer po-sm-12\">\n\n      <po-divider class=\"po-page-background-footer-mobile-only\"></po-divider>\n\n      <div class=\"po-page-background-footer-content\">\n        <div *ngIf=\"showSelectLanguage\" class=\"po-page-background-footer-select\">\n          <po-select\n            name=\"selectedLanguageOption\"\n            [(ngModel)]=\"selectedLanguageOption\"\n            [p-options]=\"selectLanguageOptions\"\n            (p-change) = \"onChangeLanguage()\">\n          </po-select>\n        </div>\n\n        <div class=\"po-page-background-secondary-logo\"\n          [ngClass]=\"showSelectLanguage ? 'po-page-background-secondary-logo-right' : 'po-page-background-secondary-logo-centered'\">\n\n          <img *ngIf=\"secondaryLogo\"\n            class=\"po-page-background-secondary-logo-image\"\n            alt=\"secondary-logo\"\n            [src]=\"secondaryLogo\">\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div *ngIf=\"background\"\n    [ngClass]=\"background ? 'po-page-login-highlight-image' : 'po-page-login-highlight-image-off'\"\n    [style.background-image]=\"'url(' + background + ')'\">\n    <div class=\"po-page-login-highlight-text\">\n      <div class=\"po-font-display\">{{ highlightInfo }}</div>\n    </div>\n  </div>\n</div>\n"
                }] }
    ];
    PoPageBackgroundComponent.propDecorators = {
        background: [{ type: Input, args: ['p-background',] }],
        hideLogo: [{ type: Input, args: ['p-hide-logo',] }],
        highlightInfo: [{ type: Input, args: ['p-highlight-info',] }],
        logo: [{ type: Input, args: ['p-logo',] }],
        secondaryLogo: [{ type: Input, args: ['p-secondary-logo',] }],
        showSelectLanguage: [{ type: Input, args: ['p-show-select-language',] }],
        selectedLanguage: [{ type: Output, args: ['p-selected-language',] }]
    };
    return PoPageBackgroundComponent;
}());
export { PoPageBackgroundComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageBackgroundComponent.prototype._logo;
    /**
     * @type {?}
     * @private
     */
    PoPageBackgroundComponent.prototype._secondaryLogo;
    /**
     * @type {?}
     * @private
     */
    PoPageBackgroundComponent.prototype._showSelectLanguage;
    /** @type {?} */
    PoPageBackgroundComponent.prototype.selectedLanguageOption;
    /** @type {?} */
    PoPageBackgroundComponent.prototype.selectLanguageOptions;
    /**
     * Insere uma imagem de destaque ao lado direito do container.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.background;
    /**
     * Designa se o logotipo deve desaparecer em resoluções menores.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.hideLogo;
    /**
     * Texto de destaque sobreposto à imagem de destaque. Essa opção é utilizada em conjunto com o atributo `p-background`.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.highlightInfo;
    /**
     * Evento disparado ao selecionar alguma opção no seletor de idiomas.
     * Para este evento será passado como parâmetro o valor de idioma selecionado.
     * @type {?}
     */
    PoPageBackgroundComponent.prototype.selectedLanguage;
}
