// @portinari/portinari-ui/esm5/lib/components/po-field/po-number/po-number-base.component.js
import * as tslib_1 from "tslib";
import { PoInputGeneric } from '../po-input-generic/po-input-generic';
var PoNumberBaseComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoNumberBaseComponent, _super);
    function PoNumberBaseComponent(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.type = 'number';
        return _this;
    }
    PoNumberBaseComponent.prototype.eventOnInput = function (e) {
        if (!this.mask) {
            var value = e.target.value;
            var valueMaxlength = this.validMaxLength(this.maxlength, value);
            if (value !== valueMaxlength) {
                value = valueMaxlength;
                this.inputEl.nativeElement.value = value;
            }
            this.callOnChange(this.formatNumber(value));
        }
    };
    PoNumberBaseComponent.prototype.validMaxLength = function (maxlength, value) {
        if (maxlength && value.length > maxlength) {
            var substringValue = value.toString().substring(0, maxlength);
            if (substringValue && this.isEndWithDot(substringValue)) {
                return substringValue.toString().substring(0, maxlength - 1);
            }
            return substringValue;
        }
        return value;
    };
    PoNumberBaseComponent.prototype.writeValueModel = function (value) {
        if (this.inputEl) {
            if (value || value === 0) {
                if (this.mask) {
                    this.inputEl.nativeElement.value = this.objMask.controlFormatting(String(value));
                    // Se o model for definido como formatado, então precisa atualizá-lo no primeiro acesso
                    if (this.objMask.formatModel) {
                        this.onChangePropagate(this.objMask.valueToModel);
                    }
                }
                else {
                    this.inputEl.nativeElement.value = value;
                }
            }
            else { // Se for o valor for undefined, deve limpar o campo
                this.inputEl.nativeElement.value = '';
            }
        }
        // Emite evento quando o model é atualizado, inclusive a primeira vez
        this.changeModel.emit(value);
    };
    PoNumberBaseComponent.prototype.isEndWithDot = function (value) {
        return value && value.lastIndexOf('.') === value.length - 1;
    };
    PoNumberBaseComponent.prototype.formatNumber = function (value) {
        return value ? Number(value) : null;
    };
    return PoNumberBaseComponent;
}(PoInputGeneric));
export { PoNumberBaseComponent };
