// @portinari/portinari-ui/esm5/lib/components/po-navbar/po-navbar-actions/po-navbar-action/po-navbar-action.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { callFunction, isExternalLink, openExternalLink } from '../../../../utils/util';
var PoNavbarActionComponent = /** @class */ (function () {
    function PoNavbarActionComponent(viewContainerRef, router) {
        this.router = router;
        this.parentRef = viewContainerRef['_view']['component'];
    }
    PoNavbarActionComponent.prototype.click = function () {
        if (this.action) {
            callFunction(this.action, this.parentRef || this);
            return;
        }
        if (this.link) {
            return this.openUrl(this.link);
        }
    };
    PoNavbarActionComponent.prototype.openUrl = function (url) {
        if (isExternalLink(url)) {
            return openExternalLink(url);
        }
        if (url) {
            return this.router.navigate([url]);
        }
    };
    PoNavbarActionComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: Router }
    ]; };
    tslib_1.__decorate([
        Input('p-action'),
        tslib_1.__metadata("design:type", Function)
    ], PoNavbarActionComponent.prototype, "action", void 0);
    tslib_1.__decorate([
        Input('p-icon'),
        tslib_1.__metadata("design:type", String)
    ], PoNavbarActionComponent.prototype, "icon", void 0);
    tslib_1.__decorate([
        Input('p-label'),
        tslib_1.__metadata("design:type", String)
    ], PoNavbarActionComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input('p-link'),
        tslib_1.__metadata("design:type", String)
    ], PoNavbarActionComponent.prototype, "link", void 0);
    tslib_1.__decorate([
        Input('p-tooltip-text'),
        tslib_1.__metadata("design:type", String)
    ], PoNavbarActionComponent.prototype, "tooltip", void 0);
    PoNavbarActionComponent = tslib_1.__decorate([
        Component({
            selector: 'po-navbar-action',
            template: "<div \n  tabindex=\"0\"\n  class=\"po-navbar-action-content po-clickable\"\n  (click)=\"click()\">\n\n  <span\n    class=\"po-icon {{ icon }}\"\n    [p-tooltip]=\"tooltip\">\n  </span>\n\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [ViewContainerRef, Router])
    ], PoNavbarActionComponent);
    return PoNavbarActionComponent;
}());
export { PoNavbarActionComponent };
