// @portinari/portinari-ui/esm2015/lib/components/po-grid/po-grid-head/po-grid-head.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ElementRef, ChangeDetectorRef } from '@angular/core';
let PoGridHeadComponent = class PoGridHeadComponent {
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.align = 'left';
        this.boolean = false;
        this.position = '';
        this.tabIndex = -1;
    }
    set width(value) {
        this._width = value;
    }
    get width() {
        return this._width ? `${this._width}px` : '100%';
    }
};
PoGridHeadComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
tslib_1.__decorate([
    Input('p-align'),
    tslib_1.__metadata("design:type", String)
], PoGridHeadComponent.prototype, "align", void 0);
tslib_1.__decorate([
    Input('p-freeze'),
    tslib_1.__metadata("design:type", Boolean)
], PoGridHeadComponent.prototype, "boolean", void 0);
tslib_1.__decorate([
    Input('p-position'),
    tslib_1.__metadata("design:type", String)
], PoGridHeadComponent.prototype, "position", void 0);
tslib_1.__decorate([
    Input('p-tab-index'),
    tslib_1.__metadata("design:type", Number)
], PoGridHeadComponent.prototype, "tabIndex", void 0);
tslib_1.__decorate([
    Input('p-width'),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], PoGridHeadComponent.prototype, "width", null);
tslib_1.__decorate([
    Input('p-title'),
    tslib_1.__metadata("design:type", String)
], PoGridHeadComponent.prototype, "title", void 0);
PoGridHeadComponent = tslib_1.__decorate([
    Component({
        selector: 'po-grid-head',
        template: "<div>\n  <div [id]=\"position\"\n    class=\"po-grid-head-content po-grid-head-{{ align }}\"\n    tabindex=\"{{ title ? 0 : -1 }}\">\n    <span>{{ title || '&nbsp;' }}</span>\n  </div>\n</div>\n\n"
    }),
    tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
], PoGridHeadComponent);
export { PoGridHeadComponent };
