// @portinari/portinari-ui/esm2015/lib/components/po-navbar/po-navbar-items/po-navbar-item/po-navbar-item.component.js
import * as tslib_1 from "tslib";
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { isExternalLink } from '../../../../utils/util';
let PoNavbarItemComponent = class PoNavbarItemComponent {
    constructor() {
        this.click = new EventEmitter();
    }
    get type() {
        return isExternalLink(this.link) ? 'externalLink' : 'internalLink';
    }
    itemClick(label, link) {
        if (this.action) {
            this.action({ label, link });
        }
        this.click.emit();
    }
};
tslib_1.__decorate([
    Input('p-action'),
    tslib_1.__metadata("design:type", Function)
], PoNavbarItemComponent.prototype, "action", void 0);
tslib_1.__decorate([
    Input('p-clickable'),
    tslib_1.__metadata("design:type", Boolean)
], PoNavbarItemComponent.prototype, "clickable", void 0);
tslib_1.__decorate([
    Input('p-label'),
    tslib_1.__metadata("design:type", String)
], PoNavbarItemComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Input('p-link'),
    tslib_1.__metadata("design:type", String)
], PoNavbarItemComponent.prototype, "link", void 0);
tslib_1.__decorate([
    Output('p-click'),
    tslib_1.__metadata("design:type", EventEmitter)
], PoNavbarItemComponent.prototype, "click", void 0);
PoNavbarItemComponent = tslib_1.__decorate([
    Component({
        selector: 'po-navbar-item',
        template: "<ng-container [ngSwitch]=\"type\">\n\n  <a *ngSwitchCase=\"'externalLink'\"\n    class=\"po-navbar-item-link\"\n    [class.po-clickable]=\"clickable\"\n    [href]=\"link\" \n    target=\"_blank\"\n    (click)=\"itemClick()\">{{ label }}</a>\n\n  <a *ngSwitchCase=\"'internalLink'\"\n    class=\"po-navbar-item-link\"\n    [class.po-clickable]=\"clickable\"\n    [routerLink]=\"link\"\n    (click)=\"itemClick(label, link)\">{{ label }}</a>\n\n</ng-container>\n"
    })
], PoNavbarItemComponent);
export { PoNavbarItemComponent };
