// @portinari/portinari-ui/esm2015/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { convertToInt } from '../../../../utils/util';
import { PoUploadDragDropAreaComponent } from './po-upload-drag-drop-area/po-upload-drag-drop-area.component';
const PoUploadDragDropHeightDefault = 320;
const PoUploadDragDropHeightMin = 160;
let PoUploadDragDropComponent = class PoUploadDragDropComponent {
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this._dragDropHeight = PoUploadDragDropHeightDefault;
        this.isDragOver = false;
        this.fileChange = new EventEmitter();
        this.selectFiles = new EventEmitter();
    }
    set dragDropHeight(value) {
        const dragDropHeight = convertToInt(value, PoUploadDragDropHeightDefault);
        this._dragDropHeight = (dragDropHeight < PoUploadDragDropHeightMin) ? PoUploadDragDropHeightMin : dragDropHeight;
    }
    get dragDropHeight() {
        return this._dragDropHeight;
    }
    focus() {
        this.dragDropAreaComponent.focus();
    }
    onAreaElement(element) {
        this.areaElement = element;
        // necessário para não ocorrer o erro HasBeenChecked
        this.changeDetector.detectChanges();
    }
    onDragLeave() {
        this.isDragOver = false;
    }
    onDragOver() {
        this.isDragOver = true;
    }
    onFileChange(files) {
        this.isDragOver = false;
        this.fileChange.emit(files);
    }
};
PoUploadDragDropComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
tslib_1.__decorate([
    ViewChild('dragDropOverlay', { read: ElementRef, static: false }),
    tslib_1.__metadata("design:type", ElementRef)
], PoUploadDragDropComponent.prototype, "dragDropOverlayElement", void 0);
tslib_1.__decorate([
    ViewChild(PoUploadDragDropAreaComponent, { static: true }),
    tslib_1.__metadata("design:type", PoUploadDragDropAreaComponent)
], PoUploadDragDropComponent.prototype, "dragDropAreaComponent", void 0);
tslib_1.__decorate([
    Input('p-directory-compatible'),
    tslib_1.__metadata("design:type", Boolean)
], PoUploadDragDropComponent.prototype, "directoryCompatible", void 0);
tslib_1.__decorate([
    Input('p-disabled'),
    tslib_1.__metadata("design:type", Boolean)
], PoUploadDragDropComponent.prototype, "disabled", void 0);
tslib_1.__decorate([
    Input('p-drag-drop-height'),
    tslib_1.__metadata("design:type", Number),
    tslib_1.__metadata("design:paramtypes", [Number])
], PoUploadDragDropComponent.prototype, "dragDropHeight", null);
tslib_1.__decorate([
    Input('p-literals'),
    tslib_1.__metadata("design:type", Object)
], PoUploadDragDropComponent.prototype, "literals", void 0);
tslib_1.__decorate([
    Output('p-file-change'),
    tslib_1.__metadata("design:type", EventEmitter)
], PoUploadDragDropComponent.prototype, "fileChange", void 0);
tslib_1.__decorate([
    Output('p-select-files'),
    tslib_1.__metadata("design:type", EventEmitter)
], PoUploadDragDropComponent.prototype, "selectFiles", void 0);
PoUploadDragDropComponent = tslib_1.__decorate([
    Component({
        selector: 'po-upload-drag-drop',
        template: "<po-upload-drag-drop-area-overlay #dragDropOverlay *ngIf=\"isDragOver\"\r\n  [p-directory-compatible]=\"directoryCompatible\"\r\n  [p-disabled]=\"disabled\"\r\n  [p-literals]=\"literals\"\r\n  [p-target]=\"dragDropAreaComponent.elementRef\"\r\n  (p-area-element)=\"onAreaElement($event)\">\r\n</po-upload-drag-drop-area-overlay>\r\n\r\n<po-upload-drag-drop-area\r\n  p-upload-drag-drop\r\n  [p-area-element]=\"areaElement\"\r\n  [p-directory-compatible]=\"directoryCompatible\"\r\n  [p-disabled]=\"disabled\"\r\n  [p-height]=\"dragDropHeight\"\r\n  [p-literals]=\"literals\"\r\n  [p-overlay]=\"false\"\r\n  (p-drag-leave)=\"onDragLeave()\"\r\n  (p-drag-over)=\"onDragOver()\"\r\n  (p-file-change)=\"onFileChange($event)\"\r\n  (p-select-files)=\"selectFiles.emit()\">\r\n</po-upload-drag-drop-area>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
], PoUploadDragDropComponent);
export { PoUploadDragDropComponent };
