// @portinari/portinari-ui/esm5/lib/components/po-field/po-clean/po-clean-base.component.js
import * as tslib_1 from "tslib";
import { ElementRef, EventEmitter, Input, Output } from '@angular/core';
/**
 * @docsPrivate
 *
 * Este componente é de uso interno utilizado por componentes de entrada de dados com o objetivo de resetar as informações do model.
 *
 * Por padrão limpa o valor do campo e executa o método onChangePropagate, caso tenha a necessidade de tratar a função de limpar o campo,
 * deve implementar a interface PoClean.
 */
var PoCleanBaseComponent = /** @class */ (function () {
    function PoCleanBaseComponent() {
        /** Valor que será atribuído ao campo quando for clicado no po-clean. */
        this.defaultValue = '';
        /**
         * Evento disparado quando executada ação do po-clean.
         * Este evento deve ser usado para avisar para o componente que está usando o po-clean, que o botão foi disparado,
         * e provavelmente será preciso emitir o evento para atualizar o model.
         */
        this.changeEvent = new EventEmitter();
    }
    PoCleanBaseComponent.prototype.clear = function () {
        this.setInputValue(this.defaultValue);
        this.changeEvent.emit(this.defaultValue);
    };
    PoCleanBaseComponent.prototype.showIcon = function () {
        return this.defaultValue !== this.getInputValue() && this.hasCleanAttr() && !this.isDisabled() && !this.isReadonly();
    };
    // Este método verifica se o componente pai possui a propriedade clean diferente de vazio,
    // ou seja, se o po-clean deve ser usado.
    PoCleanBaseComponent.prototype.hasCleanAttr = function () {
        return this.parentComponent.clean;
    };
    // Este método verifica se o componente pai está desabilitado.
    PoCleanBaseComponent.prototype.isDisabled = function () {
        return this.parentComponent.disabled;
    };
    // Este método verifica se o componente pai está somente leitura.
    PoCleanBaseComponent.prototype.isReadonly = function () {
        return this.parentComponent.readonly;
    };
    tslib_1.__decorate([
        Input('p-element-ref'),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoCleanBaseComponent.prototype, "inputRef", void 0);
    tslib_1.__decorate([
        Input('p-default-value'),
        tslib_1.__metadata("design:type", String)
    ], PoCleanBaseComponent.prototype, "defaultValue", void 0);
    tslib_1.__decorate([
        Output('p-change-event'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoCleanBaseComponent.prototype, "changeEvent", void 0);
    return PoCleanBaseComponent;
}());
export { PoCleanBaseComponent };
