// @portinari/portinari-templates/esm2015/lib/components/po-page-change-password/po-page-change-password.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, ViewContainerRef } from '@angular/core';
import { NgForm } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { PoComponentInjectorService, PoModalComponent } from '@portinari/portinari-ui';
import { browserLanguage, isExternalLink, isTypeof, poLocaleDefault } from '../../utils/util';
import { PoModalPasswordRecoveryComponent } from '../po-modal-password-recovery/po-modal-password-recovery.component';
import { PoModalPasswordRecoveryType } from '../po-modal-password-recovery/enums/po-modal-password-recovery-type.enum';
import { PoPageChangePasswordBaseComponent } from './po-page-change-password-base.component';
import { poPageChangePasswordLiterals } from './literals/i18n/po-page-change-password-literals';
import { PoPageChangePasswordService } from './po-page-change-password.service';
/**
 * \@docsExtends PoPageChangePasswordBaseComponent
 *
 * \@example
 *
 * <example name="po-page-change-password-basic" title="Portinari Page Change Password Basic">
 *  <file name="sample-po-page-change-password-basic/sample-po-page-change-password-basic.component.html"> </file>
 *  <file name="sample-po-page-change-password-basic/sample-po-page-change-password-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-labs" title="Portinari Page Change Password Labs">
 *  <file name="sample-po-page-change-password-labs/sample-po-page-change-password-labs.component.html"> </file>
 *  <file name="sample-po-page-change-password-labs/sample-po-page-change-password-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-modify" title="Portinari Page Change Password Modify">
 *  <file name="sample-po-page-change-password-modify/sample-po-page-change-password-modify.component.html"> </file>
 *  <file name="sample-po-page-change-password-modify/sample-po-page-change-password-modify.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-create" title="Portinari Page Change Password Create">
 *  <file name="sample-po-page-change-password-create/sample-po-page-change-password-create.component.html"> </file>
 *  <file name="sample-po-page-change-password-create/sample-po-page-change-password-create.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-request" title="Portinari Page Change Password Request">
 *  <file name="sample-po-page-change-password-request/sample-po-page-change-password-request.component.html"> </file>
 *  <file name="sample-po-page-change-password-request/sample-po-page-change-password-request.component.ts"> </file>
 * </example>
 */
export class PoPageChangePasswordComponent extends PoPageChangePasswordBaseComponent {
    /**
     * @param {?} activatedRoute
     * @param {?} route
     * @param {?} router
     * @param {?} service
     * @param {?} poComponentInjector
     * @param {?} viewRef
     */
    constructor(activatedRoute, route, router, service, poComponentInjector, viewRef) {
        super();
        this.activatedRoute = activatedRoute;
        this.route = route;
        this.router = router;
        this.service = service;
        this.poComponentInjector = poComponentInjector;
        this.literals = Object.assign({}, poPageChangePasswordLiterals[poLocaleDefault], poPageChangePasswordLiterals[browserLanguage()]);
        this.componentRef = null;
        this.modalAction = {
            action: this.navigateTo.bind(this, this.urlHome),
            label: this.literals.enterSystemButton
        };
        this.parentRef = viewRef['_view']['component'];
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.urlNewPassword) {
            this.subscribeToTokenParameter();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.newPasswordSubscription) {
            this.newPasswordSubscription.unsubscribe();
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    navigateTo(url) {
        isExternalLink(url) ? window.open(url) : this.router.navigate([url || '/']);
    }
    /**
     * @param {?} recovery
     * @return {?}
     */
    onForgotPasswordClick(recovery) {
        if (isTypeof(recovery, 'function')) {
            recovery();
        }
        else {
            this.createModalPasswordRecoveryComponent(recovery);
        }
    }
    /**
     * @return {?}
     */
    onLoginSubmit() {
        /** @type {?} */
        const form = this.getLoginForm();
        if (this.urlNewPassword) {
            this.postUrlNewPassword(form);
        }
        else if (this.submit && this.submit.observers.length) {
            this.emitSubmit(form);
        }
    }
    /**
     * Abre uma modal de confirmação com texto, imagem e botão que redireciona para o link definido na propriedade `p-url-home`
     * @return {?}
     */
    openConfirmation() {
        this.modal.open();
    }
    /**
     * @return {?}
     */
    validatePassword() {
        /** @type {?} */
        const controls = this.passwordForm.form.controls;
        /** @type {?} */
        const controlConfirmPassword = controls['confirmPassword'];
        /** @type {?} */
        const controlNewPassword = controls['newPassword'];
        if (!this.newPassword) {
            this.setFormErrors({ 'required': true }, [controlNewPassword]);
        }
        else if (!this.confirmPassword) {
            this.setFormErrors({ 'required': true }, [controlConfirmPassword]);
        }
        else if (this.newPassword && this.confirmPassword && this.newPassword !== this.confirmPassword) {
            this.setFormErrors({ 'equalPassword': true }, [controlNewPassword, controlConfirmPassword]);
        }
        else {
            this.setFormErrors(null, [controlConfirmPassword, controlNewPassword]);
        }
        if (this.requirements.length && this.requirements.find((/**
         * @param {?} requirement
         * @return {?}
         */
        requirement => this.validateRequirement(requirement) === false))) {
            this.setFormErrors({ 'requirement': true }, [controlNewPassword]);
        }
    }
    /**
     * @param {?} requirement
     * @return {?}
     */
    validateRequirement(requirement) {
        return typeof requirement.status === 'function' ? requirement.status.call(this.parentRef, this.newPassword) : requirement.status;
    }
    /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    checkingForMetadataProperty(object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    }
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    checkingForRouteMetadata(data) {
        if (Object.keys(data).length !== 0) {
            this.urlNewPassword = this.checkingForMetadataProperty(data, 'serviceApi') || this.urlNewPassword;
            this.recovery = this.checkingForMetadataProperty(data, 'recovery') || this.recovery;
            this.hideCurrentPassword = this.checkingForMetadataProperty(data, 'hideCurrentPassword') || this.hideCurrentPassword;
        }
    }
    /**
     * @private
     * @param {?} recovery
     * @return {?}
     */
    createModalPasswordRecoveryComponent(recovery) {
        if (this.componentRef) {
            this.poComponentInjector.destroyComponentInApplication(this.componentRef);
        }
        this.componentRef = this.poComponentInjector.createComponentInApplication(PoModalPasswordRecoveryComponent);
        this.componentRef.instance.recovery = recovery.url;
        this.componentRef.instance.contactEmail = recovery.contactMail;
        this.componentRef.instance.phoneMask = recovery.phoneMask;
        this.componentRef.instance.type = recovery.type || PoModalPasswordRecoveryType.Email;
        this.componentRef.changeDetectorRef.detectChanges();
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.componentRef.instance.open();
        }));
    }
    /**
     * @private
     * @param {?} form
     * @return {?}
     */
    emitSubmit(form) {
        this.submit.emit(form);
    }
    /**
     * @private
     * @return {?}
     */
    getLoginForm() {
        return {
            currentPassword: this.currentPassword,
            newPassword: this.newPassword
        };
    }
    /**
     * @private
     * @param {?} form
     * @return {?}
     */
    postUrlNewPassword(form) {
        form['token'] = this.token;
        this.service.post(this.urlNewPassword, form).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            if (response.status === 204) {
                this.openConfirmation();
            }
        }));
    }
    /**
     * @private
     * @param {?} error
     * @param {?=} controls
     * @return {?}
     */
    setFormErrors(error, controls) {
        controls.forEach((/**
         * @param {?} control
         * @return {?}
         */
        control => {
            control.setErrors(error);
        }));
    }
    /**
     * @private
     * @return {?}
     */
    subscribeToTokenParameter() {
        this.route.queryParams.subscribe((/**
         * @param {?} params
         * @return {?}
         */
        params => {
            /** @type {?} */
            const token = params['token'];
            if (token) {
                this.token = token;
            }
        }));
    }
}
PoPageChangePasswordComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-page-change-password',
                template: "<div class=\"po-page-change-password-wrapper\" #pageChangePassword>\n  <div class=\"po-page-change-password-container\">\n    <po-page-background class=\"po-page-change-password\"\n      [p-logo]=\"logo\"\n      [p-secondary-logo]=\"secondaryLogo\">\n\n      <div class=\"po-page-blocked-user-header po-row\">\n        <div class=\"po-md-12\">{{literals.createNewPassword}}</div>\n        <div class=\"po-md-12 po-text-center po-font-text\">\n          {{literals.createNewPasswordPhrase}}\n        </div>\n      </div>\n\n      <form #passwordForm=\"ngForm\">\n\n        <div class=\"po-row po-mt-3 po-mt-sm-1\">\n          <div class=\"po-md-6\"\n            [ngClass]=\"{'po-offset-md-3 po-offset-lg-3 po-offset-xl-3': !showRequirements}\">\n            <div *ngIf=\"!hideCurrentPassword\">\n              <po-password\n                name=\"currentPassword\"\n                [(ngModel)]=\"currentPassword\"\n                p-required\n                [p-label]=\"literals.currentPassword\">\n              </po-password>\n            </div>\n            <div *ngIf=\"!hideCurrentPassword && recovery\" class=\"po-text-center po-mb-sm-1 po-mb-2\">\n              <a *ngIf=\"recoveryUrlType === 'internalLink'\"\n                class=\"po-font-text-bold po-page-change-password-forgot-password\"\n                [routerLink]=\"recovery\">\n                {{literals.forgotPassword}}\n              </a>\n              <a *ngIf=\"recoveryUrlType === 'externalLink'\"\n                class=\"po-font-text-bold po-page-change-password-forgot-password\"\n                [href]=\"recovery\">\n                {{literals.forgotPassword}}\n              </a>\n              <a *ngIf=\"!recoveryUrlType\"\n                class=\"po-font-text-bold po-page-change-password-forgot-password po-clickable\"\n                (click)=\"onForgotPasswordClick(recovery)\">\n                {{literals.forgotPassword}}\n              </a>\n            </div>\n            <div>\n              <po-password\n                name=\"newPassword\"\n                [(ngModel)]=\"newPassword\"\n                p-required\n                [p-label]=\"literals.newPassword\"\n                (p-change-model)=\"validatePassword()\">\n              </po-password>\n            </div>\n            <div>\n              <po-password\n                name=\"confirmPassword\"\n                [(ngModel)]=\"confirmPassword\"\n                p-required\n                [p-label]=\"literals.confirmPassword\"\n                (p-change-model)=\"validatePassword()\">\n              </po-password>\n            </div>\n          </div>\n          <div *ngIf=\"showRequirements\"\n            class=\"po-md-6 po-page-change-password-required-container\">\n            <po-container>\n              <ul class=\"po-page-change-password-required-list\">\n                <li class=\"po-page-change-password-required-item\">\n                  <p class=\"po-mb-1 po-font-text-bold po-page-change-password-required-title\">\n                    {{literals?.requirements}}\n                  </p>\n                </li>\n                <li *ngFor=\"let requirement of requirements\"\n                  class=\"po-page-change-password-required-item\">\n                  <span class=\"po-icon po-page-change-password-required-icon\"\n                    [ngClass]=\"validateRequirement(requirement) ? 'po-icon-ok po-page-change-password-required-ok' : 'po-icon-minus po-page-change-password-required-minus'\">\n                  </span>\n                  <p class=\"po-page-change-password-required-text po-font-text\">\n                    {{requirement.requirement}}\n                  </p>\n                </li>\n              </ul>\n            </po-container>\n          </div>\n        </div>\n      </form>\n\n      <po-divider class=\"po-sm-12\"></po-divider>\n\n      <div class=\"po-page-change-password-tips po-row\">\n        <div class=\"po-md-2 po-mr-3\">\n          <img\n            class=\"po-page-change-password-lock-image po-mb-3\"\n            alt=\"Blocked user image\"\n            src=\"./assets/images/big-lock.svg\"\n          />\n        </div>\n        <div class=\"po-md-9 po-page-change-password-tips-text\">\n          <div class=\"po-font-text-bold po-pb-1\">{{literals.safetyTips}}</div>\n          <div class=\"po-font-text-small\">\n            <div class=\"po-row po-pb-1\">{{literals.safetyTipsPhrase}}</div>\n            <div class=\"po-row po-pb-1\">\u2022 {{literals.safetyTipsFirst}}</div>\n            <div class=\"po-row po-pb-1\">\u2022 {{literals.safetyTipsSecond}}</div>\n            <div class=\"po-row po-pb-1\">\u2022 {{literals.safetyTipsThird}}</div>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"po-page-change-password-buttons po-row\">\n\n        <po-divider class=\"po-sm-12 po-mobile-only\"></po-divider>\n\n        <po-button *ngIf=\"!hideCurrentPassword\"\n          class=\"po-sm-6 po-mb-sm-1\"\n          [p-label]=\"literals.backButton\"\n          (p-click)=\"navigateTo(urlBack)\">\n        </po-button>\n\n        <po-button\n          [ngClass]=\"hideCurrentPassword ? 'po-md-12':'po-sm-6'\"\n          p-type=\"primary\"\n          [p-disabled]=\"passwordForm.invalid\"\n          [p-label]=\"literals.saveButton\"\n          (p-click)=\"onLoginSubmit()\">\n        </po-button>\n\n      </div>\n\n    </po-page-background>\n\n  </div>\n</div>\n\n<po-modal\n  [p-primary-action]=\"modalAction\"\n  p-hide-close\n  p-size=\"auto\"\n  [p-title]=\"hideCurrentPassword ? literals.passwordSuccessfullyCreated : literals.passwordSuccessfullyUpdated\">\n\n  <div class=\"po-text-center\">\n    <img\n      alt=\"Blocked user image\"\n      class=\"po-page-change-password-lock-image po-mb-3\"\n      src=\"./assets/images/success.svg\"\n    />\n  </div>\n</po-modal>\n"
            }] }
];
/** @nocollapse */
PoPageChangePasswordComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: ActivatedRoute },
    { type: Router },
    { type: PoPageChangePasswordService },
    { type: PoComponentInjectorService },
    { type: ViewContainerRef }
];
PoPageChangePasswordComponent.propDecorators = {
    modal: [{ type: ViewChild, args: [PoModalComponent, { static: true },] }],
    pageChangePassword: [{ type: ViewChild, args: ['pageChangePassword', { read: ViewContainerRef, static: true },] }],
    passwordForm: [{ type: ViewChild, args: ['passwordForm', { read: NgForm, static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.newPasswordSubscription;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.literals;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.componentRef;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.modalAction;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.parentRef;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.modal;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.pageChangePassword;
    /** @type {?} */
    PoPageChangePasswordComponent.prototype.passwordForm;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.service;
    /**
     * @type {?}
     * @private
     */
    PoPageChangePasswordComponent.prototype.poComponentInjector;
}
