// @portinari/portinari-ui/esm5/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area-overlay/po-upload-drag-drop-area-overlay.component.js
import * as tslib_1 from "tslib";
import { AfterViewInit, ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
var PoUploadDragDropAreaOverlayComponent = /** @class */ (function () {
    function PoUploadDragDropAreaOverlayComponent(renderer) {
        this.renderer = renderer;
        this.areaElement = new EventEmitter();
    }
    PoUploadDragDropAreaOverlayComponent.prototype.ngAfterViewInit = function () {
        if (this.target) {
            this.setPosition(this.target);
        }
        this.areaElement.emit(this.DragDropAreaFixed.nativeElement);
    };
    PoUploadDragDropAreaOverlayComponent.prototype.setPosition = function (targetElement) {
        var _this = this;
        var boundingClientRect = targetElement.nativeElement.getBoundingClientRect();
        var clientRectKeys = ['bottom', 'left', 'height', 'right', 'top', 'width'];
        clientRectKeys.forEach(function (clientRectKey) {
            var clientRectValue = boundingClientRect[clientRectKey];
            _this.renderer.setStyle(_this.DragDropAreaFixed.nativeElement, clientRectKey, clientRectValue + "px");
        });
    };
    PoUploadDragDropAreaOverlayComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    tslib_1.__decorate([
        ViewChild('DragDropAreaFixed', { read: ElementRef, static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoUploadDragDropAreaOverlayComponent.prototype, "DragDropAreaFixed", void 0);
    tslib_1.__decorate([
        Input('p-directory-compatible'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoUploadDragDropAreaOverlayComponent.prototype, "directoryCompatible", void 0);
    tslib_1.__decorate([
        Input('p-disabled'),
        tslib_1.__metadata("design:type", Boolean)
    ], PoUploadDragDropAreaOverlayComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object)
    ], PoUploadDragDropAreaOverlayComponent.prototype, "literals", void 0);
    tslib_1.__decorate([
        Input('p-target'),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoUploadDragDropAreaOverlayComponent.prototype, "target", void 0);
    tslib_1.__decorate([
        Output('p-area-element'),
        tslib_1.__metadata("design:type", EventEmitter)
    ], PoUploadDragDropAreaOverlayComponent.prototype, "areaElement", void 0);
    PoUploadDragDropAreaOverlayComponent = tslib_1.__decorate([
        Component({
            selector: 'po-upload-drag-drop-area-overlay',
            template: "<div class=\"po-overlay-fixed\">\n\n  <po-upload-drag-drop-area #DragDropAreaFixed\n    class=\"po-upload-drag-drop-area-overlay\"\n    [p-directory-compatible]=\"directoryCompatible\"\n    [p-disabled]=\"disabled\"\n    [p-literals]=\"literals\"\n    [p-overlay]=\"true\">\n  </po-upload-drag-drop-area>\n\n</div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2])
    ], PoUploadDragDropAreaOverlayComponent);
    return PoUploadDragDropAreaOverlayComponent;
}());
export { PoUploadDragDropAreaOverlayComponent };
