// @portinari/portinari-ui/esm5/lib/components/po-field/po-rich-text/po-rich-text-body/po-rich-text-body.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { isFirefox, isIE, isIEOrEdge, openExternalLink } from './../../../../utils/util';
import { PoKeyCodeEnum } from './../../../../enums/po-key-code.enum';
var poRichTextBodyCommands = [
    'bold', 'italic', 'underline', 'justifyleft', 'justifycenter', 'justifyright', 'justifyfull', 'insertUnorderedList', 'Createlink'
];
var PoRichTextBodyComponent = /** @class */ (function () {
    function PoRichTextBodyComponent() {
        this.change = new EventEmitter();
        this.commands = new EventEmitter();
        this.selectedLink = new EventEmitter();
        this.shortcutCommand = new EventEmitter();
        this.value = new EventEmitter();
        this.onAnchorClick = function (event) {
            var target = event.target, ctrlKey = event.ctrlKey, metaKey = event.metaKey;
            var url;
            var elementLink;
            if (ctrlKey || metaKey) {
                if (event.path) {
                    event.path.forEach(function (element) {
                        if (element.nodeName === 'A') {
                            url = element.href;
                            elementLink = element;
                        }
                    });
                }
                else {
                    url = target.attributes.href.value;
                    elementLink = target;
                }
                openExternalLink(url);
                elementLink.classList.remove('po-clickable');
            }
        };
    }
    PoRichTextBodyComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.bodyElement.nativeElement.designMode = 'on';
        // timeout necessário para setar o valor vindo do writeValue do componente principal.
        setTimeout(function () { return _this.updateValueWithModelValue(); });
    };
    PoRichTextBodyComponent.prototype.executeCommand = function (command) {
        this.bodyElement.nativeElement.focus();
        if (typeof (command) === 'object') {
            if (command.command === 'InsertHTML') {
                var linkCommand = command.command, urlLink = command.value.urlLink, urlLinkText = command.value.urlLinkText;
                this.handleCommandLink(linkCommand, urlLink, urlLinkText);
            }
            else {
                document.execCommand(command.command, false, command.value);
            }
        }
        else {
            document.execCommand(command, false, null);
        }
        this.updateModel();
        this.value.emit(this.modelValue);
    };
    PoRichTextBodyComponent.prototype.linkEditing = function (event) {
        this.isLinkEditing = !!event;
    };
    PoRichTextBodyComponent.prototype.onBlur = function () {
        var _this = this;
        if (this.modelValue !== this.valueBeforeChange) {
            clearTimeout(this.timeoutChange);
            this.timeoutChange = setTimeout(function () {
                _this.change.emit(_this.modelValue);
            }, 200);
        }
    };
    PoRichTextBodyComponent.prototype.focus = function () {
        this.bodyElement.nativeElement.focus();
    };
    PoRichTextBodyComponent.prototype.onClick = function () {
        this.emitSelectionCommands();
    };
    PoRichTextBodyComponent.prototype.onFocus = function () {
        this.valueBeforeChange = this.modelValue;
    };
    PoRichTextBodyComponent.prototype.onKeyDown = function (event) {
        var keyK = event.keyCode === PoKeyCodeEnum.keyK;
        var isLinkShortcut = keyK && event.ctrlKey || keyK && event.metaKey;
        if (isLinkShortcut) {
            event.preventDefault();
            this.shortcutCommand.emit();
        }
        this.toggleCursorOnLink(event, 'add');
    };
    PoRichTextBodyComponent.prototype.onKeyUp = function (event) {
        this.toggleCursorOnLink(event, 'remove');
        this.removeBrElement();
        this.updateModel();
        this.emitSelectionCommands();
    };
    PoRichTextBodyComponent.prototype.onPaste = function () {
        this.addClickListenerOnAnchorElements();
        this.update();
    };
    PoRichTextBodyComponent.prototype.update = function () {
        var _this = this;
        setTimeout(function () { return _this.updateModel(); });
        setTimeout(function () {
            _this.removeBrElement();
            _this.updateModel();
            _this.emitSelectionCommands();
        });
    };
    PoRichTextBodyComponent.prototype.addClickListenerOnAnchorElements = function () {
        var _this = this;
        this.bodyElement.nativeElement.querySelectorAll('a').forEach(function (element) {
            element.addEventListener('click', _this.onAnchorClick);
        });
    };
    PoRichTextBodyComponent.prototype.emitSelectionCommands = function () {
        var commands = poRichTextBodyCommands.filter(function (command) { return document.queryCommandState(command); });
        var rgbColor = document.queryCommandValue('ForeColor');
        var hexColor;
        if (!isIE()) {
            hexColor = this.rgbToHex(rgbColor);
        }
        if (this.isCursorPositionedInALink()) {
            commands.push('Createlink');
        }
        this.selectedLink.emit(this.linkElement); // importante ficar fora do if para emitir mesmo undefined.
        this.commands.emit({ commands: commands, hexColor: hexColor });
    };
    PoRichTextBodyComponent.prototype.getTextSelection = function () {
        var textSelection = document.getSelection();
        if (!textSelection) {
            return;
        }
        var focusNode = textSelection.focusNode ? textSelection.focusNode.parentElement : undefined;
        var anchorNode = textSelection.anchorNode ? textSelection.anchorNode.parentNode : undefined;
        var node = focusNode || anchorNode;
        var tagName;
        if (node) {
            tagName = node['tagName'] || node['nodeName'];
            return {
                node: node,
                tagName: tagName
            };
        }
    };
    PoRichTextBodyComponent.prototype.handleCommandLink = function (linkCommand, urlLink, urlLinkText) {
        if (isIE()) {
            this.insertHtmlLinkElement(urlLink, urlLinkText);
        }
        else {
            // '&nbsp;' necessário para o cursor não ficar preso dentro do link no Firefox.
            var linkValue = isFirefox() && !this.isLinkEditing ?
                "&nbsp;" + this.makeLinkTag(urlLink, urlLinkText) + "&nbsp;" :
                this.makeLinkTag(urlLink, urlLinkText);
            document.execCommand(linkCommand, false, linkValue);
        }
        this.addClickListenerOnAnchorElements();
    };
    // tratamento específico para IE pois não suporta o comando 'insertHTML'.
    PoRichTextBodyComponent.prototype.insertHtmlLinkElement = function (urlLink, urlLinkText) {
        var selection = document.getSelection();
        var selectionRange = selection.getRangeAt(0);
        var elementLink = document.createElement('a');
        var elementlinkText = document.createTextNode(urlLinkText);
        elementLink.appendChild(elementlinkText);
        elementLink.href = urlLink;
        elementLink.setAttribute('target', '_blank');
        elementLink.classList.add('po-rich-text-link');
        selectionRange.deleteContents();
        selectionRange.insertNode(elementLink);
    };
    PoRichTextBodyComponent.prototype.isCursorPositionedInALink = function () {
        var textSelection = this.getTextSelection();
        this.linkElement = undefined;
        var isLink = false;
        if (textSelection && textSelection.node && textSelection.tagName === 'A') {
            this.linkElement = textSelection.node;
            isLink = true;
        }
        else if ((isFirefox() || isIEOrEdge()) && this.verifyCursorPositionInFirefoxIEEdge()) {
            isLink = true;
        }
        else {
            isLink = textSelection ? this.isParentNodeAnchor(textSelection) : false;
        }
        return isLink;
    };
    PoRichTextBodyComponent.prototype.isParentNodeAnchor = function (textSelection) {
        var element = textSelection.node;
        var isLink = false;
        while (element && (element.tagName !== null || element.nodeName !== null)) {
            if (element.tagName === 'A' || element.nodeName === 'A') {
                this.linkElement = element;
                isLink = true;
                return isLink;
            }
            element = element.parentElement || element.parentNode;
        }
        this.linkElement = undefined;
        return isLink;
    };
    PoRichTextBodyComponent.prototype.makeLinkTag = function (urlLink, urlLinkText) {
        return "<a class=\"po-rich-text-link\" href=\"" + urlLink + "\" target=\"_blank\">" + (urlLinkText || urlLink) + "</a>";
    };
    // Tratamento necessário para eliminar a tag <br> criada no firefox quando o body for limpo.
    PoRichTextBodyComponent.prototype.removeBrElement = function () {
        var bodyElement = this.bodyElement.nativeElement;
        if (!bodyElement.innerText.trim() && bodyElement.childNodes.length === 1 && bodyElement.querySelector('br')) {
            bodyElement.querySelector('br').remove();
        }
    };
    PoRichTextBodyComponent.prototype.rgbToHex = function (rgb) {
        // Tratamento necessário para converter o código rgb para hexadecimal.
        var sep = rgb.indexOf(',') > -1 ? ',' : ' ';
        rgb = rgb.substr(4).split(')')[0].split(sep);
        var r = (+rgb[0]).toString(16);
        var g = (+rgb[1]).toString(16);
        var b = (+rgb[2]).toString(16);
        if (r.length === 1) {
            r = '0' + r;
        }
        if (g.length === 1) {
            g = '0' + g;
        }
        if (b.length === 1) {
            b = '0' + b;
        }
        return '#' + r + g + b;
    };
    PoRichTextBodyComponent.prototype.toggleCursorOnLink = function (event, action) {
        var selection = document.getSelection();
        var element = selection.focusNode ? selection.focusNode.parentNode : undefined;
        var isCtrl = event.key === 'Control';
        var isCommand = event.key === 'Meta';
        var isOnCtrlLink = this.isCursorPositionedInALink() && (isCtrl || isCommand);
        if (element) {
            if (isOnCtrlLink) {
                element['classList'][action]('po-clickable');
            }
            else {
                var isClickable = element['classList'] && element['classList'].contains('po-clickable');
                if (isClickable) {
                    element['classList'].remove('po-clickable');
                }
            }
            this.updateModel();
        }
    };
    PoRichTextBodyComponent.prototype.updateModel = function () {
        this.modelValue = this.bodyElement.nativeElement.innerHTML;
        this.value.emit(this.modelValue);
    };
    PoRichTextBodyComponent.prototype.updateValueWithModelValue = function () {
        if (this.modelValue) {
            this.bodyElement.nativeElement.insertAdjacentHTML('afterbegin', this.modelValue);
        }
    };
    PoRichTextBodyComponent.prototype.verifyCursorPositionInFirefoxIEEdge = function () {
        var textSelection = document.getSelection();
        var nodeLink = textSelection.focusNode;
        var isLink = false;
        if (nodeLink && nodeLink.nodeName === 'A') {
            this.linkElement = nodeLink;
            isLink = true;
        }
        else {
            var range = textSelection.getRangeAt(0);
            var fragmentDocument = range.cloneContents();
            var element = fragmentDocument.childNodes[0] || fragmentDocument.firstElementChild;
            this.linkElement = (element && element.nodeName === 'A') ? element : undefined;
            isLink = !!this.linkElement;
        }
        return isLink;
    };
    tslib_1.__decorate([
        ViewChild('bodyElement', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoRichTextBodyComponent.prototype, "bodyElement", void 0);
    tslib_1.__decorate([
        Input('p-height'),
        tslib_1.__metadata("design:type", String)
    ], PoRichTextBodyComponent.prototype, "height", void 0);
    tslib_1.__decorate([
        Input('p-model-value'),
        tslib_1.__metadata("design:type", String)
    ], PoRichTextBodyComponent.prototype, "modelValue", void 0);
    tslib_1.__decorate([
        Input('p-placeholder'),
        tslib_1.__metadata("design:type", String)
    ], PoRichTextBodyComponent.prototype, "placeholder", void 0);
    tslib_1.__decorate([
        Input('p-readonly'),
        tslib_1.__metadata("design:type", String)
    ], PoRichTextBodyComponent.prototype, "readonly", void 0);
    tslib_1.__decorate([
        Output('p-change'),
        tslib_1.__metadata("design:type", Object)
    ], PoRichTextBodyComponent.prototype, "change", void 0);
    tslib_1.__decorate([
        Output('p-commands'),
        tslib_1.__metadata("design:type", Object)
    ], PoRichTextBodyComponent.prototype, "commands", void 0);
    tslib_1.__decorate([
        Output('p-selected-link'),
        tslib_1.__metadata("design:type", Object)
    ], PoRichTextBodyComponent.prototype, "selectedLink", void 0);
    tslib_1.__decorate([
        Output('p-shortcut-command'),
        tslib_1.__metadata("design:type", Object)
    ], PoRichTextBodyComponent.prototype, "shortcutCommand", void 0);
    tslib_1.__decorate([
        Output('p-value'),
        tslib_1.__metadata("design:type", Object)
    ], PoRichTextBodyComponent.prototype, "value", void 0);
    PoRichTextBodyComponent = tslib_1.__decorate([
        Component({
            selector: 'po-rich-text-body',
            template: "<div #bodyElement\n  class=\"po-rich-text-body\"\n  tabindex=\"0\"\n  [attr.contenteditable]=\"!readonly\"\n  [attr.data-placeholder]=\"placeholder\"\n  [style.height.px]=\"height\"\n  (blur)=\"onBlur()\"\n  (click)=\"onClick()\"\n  (cut)=\"update()\"\n  (focus)=\"onFocus()\"\n  (keydown)=\"onKeyDown($event)\"\n  (keyup)=\"onKeyUp($event)\"\n  (paste)=\"onPaste()\">\n</div>\n"
        })
    ], PoRichTextBodyComponent);
    return PoRichTextBodyComponent;
}());
export { PoRichTextBodyComponent };
