// @portinari/portinari-ui/esm5/lib/components/po-accordion/po-accordion-item/po-accordion-item.component.js
import * as tslib_1 from "tslib";
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { PoAccordionService } from '../services/po-accordion.service';
/**
 * @description
 *
 * Componente utilizado para renderizar os itens do `po-accordion`.
 *
 * O componente `po-accordion` já faz o controle de abertura e fechamento dos itens automaticamente,
 * mas caso houver a necessidade de abrir algum dos `po-accordion-item` via Typescript, pode ser feita da seguinte forma:
 *
 * ```
 * <po-accordion>
 *   <po-accordion-item p-label="Portinari Accordion 1" #item1>
 *      Accordion 1
 *   </po-accordion-item>
 *
 *   <po-accordion-item p-label="Portinari Accordion 2">
 *      Accordion 2
 *   </po-accordion-item>
 * </po-accordion>
 * ```
 *
 * e no typescript pode-se utilizar o `@ViewChild`:
 *
 * ```
 *  @ViewChild(PoAccordionItemComponent, { static: true }) item1: PoAccordionItemComponent;
 *
 *  ngAfterContentInit() {
 *    // ou utilizar o método collapse()
 *    this.item1.expand();
 *  }
 * ```
 */
var PoAccordionItemComponent = /** @class */ (function () {
    function PoAccordionItemComponent(accordionService) {
        this.accordionService = accordionService;
    }
    /**
     * Método para colapsar o `po-accordion-item`.
     */
    PoAccordionItemComponent.prototype.collapse = function () {
        this.expanded = false;
        this.accordionService.sendToParentAccordionItemClicked(this);
    };
    /**
     * Método para expandir o `po-accordion-item`.
     */
    PoAccordionItemComponent.prototype.expand = function () {
        this.expanded = true;
        this.accordionService.sendToParentAccordionItemClicked(this);
    };
    PoAccordionItemComponent.ctorParameters = function () { return [
        { type: PoAccordionService }
    ]; };
    tslib_1.__decorate([
        Input('p-label'),
        tslib_1.__metadata("design:type", String)
    ], PoAccordionItemComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        ViewChild(TemplateRef, { static: true }),
        tslib_1.__metadata("design:type", TemplateRef)
    ], PoAccordionItemComponent.prototype, "templateRef", void 0);
    PoAccordionItemComponent = tslib_1.__decorate([
        Component({
            selector: 'po-accordion-item',
            template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [PoAccordionService])
    ], PoAccordionItemComponent);
    return PoAccordionItemComponent;
}());
export { PoAccordionItemComponent };
