// @portinari/portinari-ui/esm2015/lib/components/po-breadcrumb/po-breadcrumb-favorite/po-breadcrumb-favorite.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { PoBreadcrumbFavoriteService } from './po-breadcrumb-favorite.service';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que renderiza o serviço de favoritar do po-breadcrumb.
 */
let PoBreadcrumbFavoriteComponent = class PoBreadcrumbFavoriteComponent {
    constructor(service) {
        this.service = service;
        this.favorite = false;
    }
    ngOnInit() {
        this.service.configService(this.favoriteService, this.paramsService, this.itemActive);
        this.getStatusFavorite();
    }
    ngOnDestroy() {
        if (this.getSubscription) {
            this.getSubscription.unsubscribe();
        }
        if (this.setSubscription) {
            this.setSubscription.unsubscribe();
        }
    }
    toggleFavoriteAction() {
        this.favorite ? this.setStatusFavorite(false) : this.setStatusFavorite(true);
    }
    getStatusFavorite() {
        this.getSubscription = this.service.getFavorite().subscribe(result => this.favorite = result.isFavorite);
    }
    setStatusFavorite(status) {
        this.setSubscription = this.service.sendStatusFavorite(status).subscribe(result => this.favorite = result.isFavorite);
    }
};
PoBreadcrumbFavoriteComponent.ctorParameters = () => [
    { type: PoBreadcrumbFavoriteService }
];
tslib_1.__decorate([
    Input('p-favorite-service'),
    tslib_1.__metadata("design:type", String)
], PoBreadcrumbFavoriteComponent.prototype, "favoriteService", void 0);
tslib_1.__decorate([
    Input('p-item-active'),
    tslib_1.__metadata("design:type", Object)
], PoBreadcrumbFavoriteComponent.prototype, "itemActive", void 0);
tslib_1.__decorate([
    Input('p-params-service'),
    tslib_1.__metadata("design:type", Object)
], PoBreadcrumbFavoriteComponent.prototype, "paramsService", void 0);
PoBreadcrumbFavoriteComponent = tslib_1.__decorate([
    Component({
        selector: 'po-breadcrumb-favorite',
        template: "<div class=\"po-breadcrumb-favorite po-clickable\" (click)=\"toggleFavoriteAction()\">\n\n  <span\n    class=\"po-icon po-icon-star po-breadcrumb-favorite-star po-clickable\"\n    [class.po-breadcrumb-favorite-star-active]=\"favorite\">\n  </span>\n\n  <span class=\"po-hidden-sm po-breadcrumb-favorite-label\" *ngIf=\"favorite\">Desfavoritar</span>\n  <span class=\"po-hidden-sm po-breadcrumb-favorite-label\" *ngIf=\"!favorite\">Favoritar</span>\n</div>\n",
        providers: [PoBreadcrumbFavoriteService]
    }),
    tslib_1.__metadata("design:paramtypes", [PoBreadcrumbFavoriteService])
], PoBreadcrumbFavoriteComponent);
export { PoBreadcrumbFavoriteComponent };
