// @portinari/portinari-ui/esm5/lib/components/po-chart/po-chart-types/po-chart-gauge/po-chart-gauge-text-content/po-chart-gauge-text-content.component.js
import * as tslib_1 from "tslib";
import { AfterViewInit, ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, Input, ViewChild } from '@angular/core';
import { convertNumberToDecimal } from '../../../../../utils/util';
import { poChartGaugeSerieWidth } from '../../po-chart-circular/po-chart-circular.constant';
var PoChartGaugeTextContentComponent = /** @class */ (function () {
    function PoChartGaugeTextContentComponent(changeDetection) {
        this.changeDetection = changeDetection;
    }
    Object.defineProperty(PoChartGaugeTextContentComponent.prototype, "gaugeWidth", {
        get: function () {
            return this._gaugeWidth;
        },
        set: function (value) {
            this._gaugeWidth = value;
            this.checkTextDescriptionSize();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoChartGaugeTextContentComponent.prototype, "serie", {
        get: function () {
            return this._serie;
        },
        set: function (serie) {
            if (serie) {
                this.serieValueConverted = this.convertValueInPercentFormat(serie.value);
                this._serie = tslib_1.__assign({}, serie);
            }
            else {
                this._serie = undefined;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoChartGaugeTextContentComponent.prototype, "hasSerieDescription", {
        get: function () {
            return this.serie && this.serie.description;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoChartGaugeTextContentComponent.prototype, "hasSerieValue", {
        get: function () {
            return this.serie && this.serie.value >= 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoChartGaugeTextContentComponent.prototype, "maxDescriptionWidth", {
        get: function () {
            // Diferença contemplando a largura do path base e padding interno.
            var subtractionArea = (this.gaugeWidth * poChartGaugeSerieWidth * 2) * 2;
            var descriptionWidth = this.gaugeWidth - subtractionArea;
            return { 'max-width': descriptionWidth + "px" };
        },
        enumerable: true,
        configurable: true
    });
    PoChartGaugeTextContentComponent.prototype.ngAfterViewInit = function () {
        this.checkTextDescriptionSize();
    };
    PoChartGaugeTextContentComponent.prototype.checkTextDescriptionSize = function () {
        var _this = this;
        if (this.descriptionElement) {
            // tooltip necessário para contornar comportamento inesperado do tooltip.
            setTimeout(function () {
                _this.tooltip = _this.isEllipsisActive();
            });
            this.changeDetection.detectChanges();
        }
    };
    PoChartGaugeTextContentComponent.prototype.convertValueInPercentFormat = function (value) {
        var decimalValue = String(convertNumberToDecimal(value, 1)).replace('.', ',');
        return decimalValue + "%";
    };
    PoChartGaugeTextContentComponent.prototype.isEllipsisActive = function () {
        var isExceededWidth = this.descriptionElement.nativeElement.offsetWidth < this.descriptionElement.nativeElement.scrollWidth;
        return isExceededWidth ? this.serie.description : undefined;
    };
    PoChartGaugeTextContentComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    tslib_1.__decorate([
        Input('p-gauge-width'),
        tslib_1.__metadata("design:type", Number),
        tslib_1.__metadata("design:paramtypes", [Number])
    ], PoChartGaugeTextContentComponent.prototype, "gaugeWidth", null);
    tslib_1.__decorate([
        Input('p-serie'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoChartGaugeTextContentComponent.prototype, "serie", null);
    tslib_1.__decorate([
        ViewChild('description', { static: false, read: ElementRef }),
        tslib_1.__metadata("design:type", ElementRef)
    ], PoChartGaugeTextContentComponent.prototype, "descriptionElement", void 0);
    PoChartGaugeTextContentComponent = tslib_1.__decorate([
        Component({
            selector: 'po-chart-gauge-text-content',
            template: "<div *ngIf=\"hasSerieValue\" class=\"po-chart-gauge-text-content\" tabindex=\"0\" [ngStyle]=\"maxDescriptionWidth\">\n  <p class=\"po-chart-gauge-text-value\">{{ serieValueConverted }}</p>\n  <p *ngIf=\"hasSerieDescription\"\n    #description\n    class=\"po-chart-gauge-text-description\"\n    [p-tooltip]=\"tooltip\">{{ serie?.description }}</p>\n</div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
    ], PoChartGaugeTextContentComponent);
    return PoChartGaugeTextContentComponent;
}());
export { PoChartGaugeTextContentComponent };
