// @portinari/portinari-templates/esm5/lib/components/po-page-blocked-user/po-page-blocked-user-reason/po-page-blocked-user-reason.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input } from '@angular/core';
import { browserLanguage, poLocaleDefault } from '../../../utils/util';
import { poPageBlockedUserLiterals } from './../literals/i18n/po-page-blocked-user-literals';
import { PoPageBlockedUserReason } from '../enums/po-page-blocked-user-reason.enum';
var PoPageBlockedUserReasonComponent = /** @class */ (function () {
    function PoPageBlockedUserReasonComponent(changeDetector) {
        this.changeDetector = changeDetector;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    PoPageBlockedUserReasonComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.reason || changes.params) {
            this.getLiterals();
        }
    };
    /**
     * @return {?}
     */
    PoPageBlockedUserReasonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getLiterals();
    };
    /**
     * @return {?}
     */
    PoPageBlockedUserReasonComponent.prototype.getImageByReasonType = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var reasonImage;
        switch (this.reason) {
            case 'none': {
                reasonImage = 'big-lock';
                break;
            }
            case 'exceededAttempts': {
                reasonImage = 'blocked-user';
                break;
            }
            case 'expiredPassword': {
                reasonImage = 'expired';
                break;
            }
        }
        return "./assets/images/" + reasonImage + ".svg";
    };
    /**
     * @return {?}
     */
    PoPageBlockedUserReasonComponent.prototype.getParams = /**
     * @return {?}
     */
    function () {
        this.literalParams =
            this.reason === 'expiredPassword' ? [this.params.days, this.params.days] : [this.params.attempts, this.params.hours];
    };
    /**
     * @private
     * @return {?}
     */
    PoPageBlockedUserReasonComponent.prototype.getLiterals = /**
     * @private
     * @return {?}
     */
    function () {
        this.getParams();
        this.literals = tslib_1.__assign({}, poPageBlockedUserLiterals[this.reason][poLocaleDefault], poPageBlockedUserLiterals[this.reason][browserLanguage()]);
        this.changeDetector.detectChanges();
    };
    PoPageBlockedUserReasonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-page-blocked-user-reason',
                    template: "<div class=\"po-page-blocked-user-reason-content\">\r\n  <img \r\n    class=\"po-page-blocked-user-image po-mb-3\" \r\n    [src]=\"getImageByReasonType()\"> \r\n  <div class=\"po-page-blocked-user-header po-mb-md-3 po-row\">\r\n    <div class=\"po-md-12\">{{ literals?.title }}</div>\r\n  </div>\r\n  <div class=\"po-page-blocked-user-text po-font-text po-row\">\r\n    <div class=\"po-mb-sm-2 po-mb-md-3 po-md-12 po-page-blocked-user-text-bold\">{{ literals?.firstPhrase | poI18n:literalParams }}</div>\r\n    <div class=\"po-mb-sm-2 po-mb-md-3 po-md-12\">{{ literals?.secondPhrase }}</div>\r\n    <div class=\"po-mb-sm-2 po-mb-md-3 po-md-12\" *ngIf=\"literals?.thirdPhrase\">{{ literals?.thirdPhrase }}</div>\r\n  </div>\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    PoPageBlockedUserReasonComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    PoPageBlockedUserReasonComponent.propDecorators = {
        params: [{ type: Input, args: ['p-params',] }],
        reason: [{ type: Input, args: ['p-reason',] }]
    };
    return PoPageBlockedUserReasonComponent;
}());
export { PoPageBlockedUserReasonComponent };
if (false) {
    /** @type {?} */
    PoPageBlockedUserReasonComponent.prototype.literalParams;
    /** @type {?} */
    PoPageBlockedUserReasonComponent.prototype.literals;
    /** @type {?} */
    PoPageBlockedUserReasonComponent.prototype.params;
    /** @type {?} */
    PoPageBlockedUserReasonComponent.prototype.reason;
    /**
     * @type {?}
     * @private
     */
    PoPageBlockedUserReasonComponent.prototype.changeDetector;
}
