// @portinari/portinari-ui/esm5/lib/components/po-field/po-lookup/services/po-lookup-filter.service.js
import * as tslib_1 from "tslib";
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço padrão utilizado para filtrar os dados do componente po-lookup.
 */
var PoLookupFilterService = /** @class */ (function () {
    function PoLookupFilterService(httpClient) {
        this.httpClient = httpClient;
    }
    PoLookupFilterService.prototype.getFilteredItems = function (filteredItemsParams) {
        var filterParams = filteredItemsParams.filterParams, restFilteredItemsParams = tslib_1.__rest(filteredItemsParams, ["filterParams"]);
        var validatedFilterParams = this.validateParams(filterParams);
        var params = tslib_1.__assign({}, restFilteredItemsParams, validatedFilterParams);
        return this.httpClient.get(this.url, { params: params });
    };
    PoLookupFilterService.prototype.getObjectByValue = function (value, filterParams) {
        var validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(this.url + "/" + value, { params: validatedFilterParams });
    };
    PoLookupFilterService.prototype.setUrl = function (url) {
        this.url = url;
    };
    PoLookupFilterService.prototype.validateParams = function (params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    };
    PoLookupFilterService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoLookupFilterService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [HttpClient])
    ], PoLookupFilterService);
    return PoLookupFilterService;
}());
export { PoLookupFilterService };
