// @portinari/portinari-templates/esm2015/lib/services/po-page-customization/po-page-customization.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { from } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class PoPageCustomizationService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} origin
     * @param {?} originalOption
     * @return {?}
     */
    getCustomOptions(origin, originalOption) {
        return this.createObservable(origin).pipe(map((/**
         * @param {?} newPageOptions
         * @return {?}
         */
        newPageOptions => this.mergePageOptions(originalOption, newPageOptions))));
    }
    /**
     * @template T, K
     * @param {?} objectToChange
     * @param {?} newOptions
     * @return {?}
     */
    changeOriginalOptionsToNewOptions(objectToChange, newOptions) {
        Object.keys(newOptions).forEach((/**
         * @param {?} key
         * @return {?}
         */
        key => {
            /** @type {?} */
            const value = newOptions[key];
            if (objectToChange[key]) {
                if (Array.isArray(value)) {
                    objectToChange[key] = [...value];
                    return;
                }
                if ((typeof (value) === 'number' || typeof (value) === 'string')) {
                    objectToChange[key] = value;
                    return;
                }
                if (typeof (value) === 'object') {
                    objectToChange[key] = Object.assign({}, value);
                }
            }
        }));
    }
    /**
     * @private
     * @param {?} origin
     * @return {?}
     */
    createObservable(origin) {
        if (typeof origin === 'string') {
            return this.http.post(origin, {});
        }
        return from(Promise.resolve(origin()));
    }
    /**
     * @private
     * @param {?} originalOption
     * @param {?} newPageOptions
     * @return {?}
     */
    mergePageOptions(originalOption, newPageOptions) {
        /** @type {?} */
        const mergePageOptions = {
            filters: this.mergeOptions(originalOption.filters, newPageOptions.filters, 'property'),
            actions: this.mergeOptions(originalOption.actions, newPageOptions.actions, 'label'),
            breadcrumb: newPageOptions.breadcrumb ? newPageOptions.breadcrumb : originalOption.breadcrumb,
            title: newPageOptions.title ? newPageOptions.title : originalOption.title
        };
        Object.keys(mergePageOptions).forEach((/**
         * @param {?} key
         * @return {?}
         */
        key => !mergePageOptions[key] && delete mergePageOptions[key]));
        return mergePageOptions;
    }
    /**
     * @private
     * @template T
     * @param {?} originalOptions
     * @param {?} newOptions
     * @param {?} filterProp
     * @return {?}
     */
    mergeOptions(originalOptions, newOptions, filterProp) {
        if (!originalOptions && !newOptions) {
            return;
        }
        if (!newOptions) {
            return originalOptions;
        }
        if (!originalOptions) {
            return newOptions;
        }
        /** @type {?} */
        const deduplicateNewOptions = newOptions.filter((/**
         * @param {?} newItem
         * @return {?}
         */
        newItem => !originalOptions.find((/**
         * @param {?} originalItem
         * @return {?}
         */
        originalItem => originalItem[filterProp] === newItem[filterProp]))));
        /** @type {?} */
        const mergedOriginalOptions = originalOptions.map((/**
         * @param {?} originalItem
         * @return {?}
         */
        originalItem => {
            /** @type {?} */
            const newItem = newOptions.find((/**
             * @param {?} newOptionsItem
             * @return {?}
             */
            newOptionsItem => originalItem[filterProp] === newOptionsItem[filterProp])) || originalItem;
            return Object.assign({}, originalItem, newItem);
        }));
        return [...mergedOriginalOptions, ...deduplicateNewOptions];
    }
}
PoPageCustomizationService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PoPageCustomizationService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ PoPageCustomizationService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function PoPageCustomizationService_Factory() { return new PoPageCustomizationService(i0.ɵɵinject(i1.HttpClient)); }, token: PoPageCustomizationService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageCustomizationService.prototype.http;
}
