// @portinari/portinari-ui/esm5/lib/components/po-page/po-page-list/po-page-list-base.component.js
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { poLocaleDefault } from './../../../utils/util';
import { PoPageDefaultBaseComponent } from '../po-page-default/po-page-default-base.component';
export var poPageListLiteralsDefault = {
    en: {
        otherActions: 'Other actions'
    },
    es: {
        otherActions: 'Otras acciones'
    },
    pt: {
        otherActions: 'Outras ações'
    },
    ru: {
        otherActions: 'Другие действия'
    }
};
/**
 * @description
 *
 * O componente `po-page-list` é utilizado como o container principal para as telas de listagem de dados,
 * podendo ser apresentado como lista ou tabela.
 *
 * Este componente possibilita realizar filtro dos dados, no qual permite que seja atribuido uma função que será executada no momento
 * da filtragem. Este comportamento pode ser acionado tanto ao *click* do ícone [po-icon-search](/guides/icons)
 * quanto ao pressionar da tecla *ENTER* quando o foco estiver no campo de pesquisa.
 *
 * Para facilitar a manipulação e visualização dos filtros aplicados, é possível também utilizar o componente
 * [`po-disclaimer-group`](/documentation/po-disclaimer-group).
 */
var PoPageListBaseComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoPageListBaseComponent, _super);
    function PoPageListBaseComponent(languageService) {
        var _this = _super.call(this) || this;
        _this.language = languageService.getShortLanguage();
        return _this;
    }
    Object.defineProperty(PoPageListBaseComponent.prototype, "disclaimerGroup", {
        get: function () {
            return this._disclaimerGroup;
        },
        /**
         * @optional
         *
         * @description
         *
         * Objeto que implementa as propriedades da interface `PoDisclaimerGroup`.
         */
        set: function (value) {
            if (!value) {
                value = {};
            }
            this._disclaimerGroup = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageListBaseComponent.prototype, "literals", {
        get: function () {
            return this._literals || poPageListLiteralsDefault[this.language];
        },
        /**
         * @optional
         *
         * @description
         *
         * Objeto com as literais usadas no `po-page-list`.
         *
         * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
         *
         * ```
         *  const customLiterals: PoPageListLiterals = {
         *    otherActions: 'Mais ações'
         *  };
         * ```
         *
         * Ou passando apenas as literais que deseja customizar:
         *
         * ```
         *  const customLiterals: PoPageListLiterals = {
         *    otherActions: 'Ações da página'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <po-page-list
         *   [p-literals]="customLiterals">
         * </po-page-list>
         * ```
         *
         * > O valor padrão será traduzido de acordo com o idioma configurado no [`PoI18nService`](/documentation/po-i18n) ou *browser*.
         */
        set: function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, poPageListLiteralsDefault[poLocaleDefault], poPageListLiteralsDefault[this.language], value);
            }
            else {
                this._literals = poPageListLiteralsDefault[this.language];
            }
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input('p-breadcrumb'),
        tslib_1.__metadata("design:type", Object)
    ], PoPageListBaseComponent.prototype, "breadcrumb", void 0);
    tslib_1.__decorate([
        Input('p-disclaimer-group'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoPageListBaseComponent.prototype, "disclaimerGroup", null);
    tslib_1.__decorate([
        Input('p-filter'),
        tslib_1.__metadata("design:type", Object)
    ], PoPageListBaseComponent.prototype, "filter", void 0);
    tslib_1.__decorate([
        Input('p-literals'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PoPageListBaseComponent.prototype, "literals", null);
    return PoPageListBaseComponent;
}(PoPageDefaultBaseComponent));
export { PoPageListBaseComponent };
