// @portinari/portinari-templates/esm2015/lib/components/po-page-login/po-page-login.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgForm } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { ChangeDetectorRef, Component, IterableDiffers, ViewChild, ViewContainerRef } from '@angular/core';
import { isExternalLink } from '../../utils/util';
import { PoComponentInjectorService } from '@portinari/portinari-ui';
import { PoModalPasswordRecoveryComponent } from '../po-modal-password-recovery/po-modal-password-recovery.component';
import { PoModalPasswordRecoveryType } from '../po-modal-password-recovery/enums/po-modal-password-recovery-type.enum';
import { PoPageLoginBaseComponent, poPageLoginLiteralIn, poPageLoginLiteralTo, poPageLoginLiteralsDefault } from './po-page-login-base.component';
import { PoPageLoginService } from './po-page-login.service';
/**
 * \@docsExtends PoPageLoginBaseComponent
 *
 * \@example
 *
 * <example name="po-page-login-basic" title="Portinari Page Login Basic">
 *  <file name="sample-po-page-login-basic/sample-po-page-login-basic.component.html"> </file>
 *  <file name="sample-po-page-login-basic/sample-po-page-login-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-login-labs" title="Portinari Page Login Labs">
 *  <file name="sample-po-page-login-labs/sample-po-page-login-labs.component.html"> </file>
 *  <file name="sample-po-page-login-labs/sample-po-page-login-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-login-human-resources" title="Portinari Page Login - Human Resources">
 *  <file name="sample-po-page-login-human-resources/sample-po-page-login-human-resources.component.html"> </file>
 *  <file name="sample-po-page-login-human-resources/sample-po-page-login-human-resources.component.ts"> </file>
 *  <file name="sample-po-page-login-human-resources/sample-po-page-login-human-resources.module.ts"> </file>
 * </example>
 *
 * <example name="po-page-login-automatic-service" title="Portinari Page Login - Automatic Service">
 *  <file name="sample-po-page-login-automatic-service/sample-po-page-login-automatic-service.component.html"> </file>
 *  <file name="sample-po-page-login-automatic-service/sample-po-page-login-automatic-service.component.ts"> </file>
 * </example>
 */
export class PoPageLoginComponent extends PoPageLoginBaseComponent {
    /**
     * @param {?} changeDetector
     * @param {?} activatedRoute
     * @param {?} poComponentInjector
     * @param {?} differs
     * @param {?} loginService
     * @param {?} router
     */
    constructor(changeDetector, activatedRoute, poComponentInjector, differs, loginService, router) {
        super(loginService, router);
        this.changeDetector = changeDetector;
        this.activatedRoute = activatedRoute;
        this.poComponentInjector = poComponentInjector;
        this.componentRef = null;
        this.customPasswordError = { custom: false };
        this.differ = differs.find([]).create(null);
    }
    /**
     * @return {?}
     */
    ngAfterViewChecked() {
        if (this.differ) {
            this.validateArrayChanges(this.differ, [
                { array: this.loginErrors, callback: this.generateLoginError.bind(this) },
                { array: this.passwordErrors, callback: this.generatePasswordError.bind(this) }
            ]);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    }
    /**
     * @return {?}
     */
    changeLoginModel() {
        if (this.authenticationUrl) {
            this.loginErrors = [];
        }
        else {
            this.setLoginErrors(this.loginErrors);
            this.loginChange.emit(this.login);
        }
    }
    /**
     * @return {?}
     */
    changePasswordModel() {
        if (this.authenticationUrl) {
            this.passwordErrors = [];
        }
        else {
            this.setPasswordErrors(this.passwordErrors);
            this.passwordChange.emit(this.password);
        }
    }
    /**
     * @param {?} language
     * @return {?}
     */
    onSelectedLanguage(language) {
        this.selectedLanguage = language;
    }
    /**
     * @param {?} recovery
     * @return {?}
     */
    openUrl(recovery) {
        switch (typeof recovery) {
            case 'string': {
                this.setUrlRedirect(recovery);
                break;
            }
            case 'function': {
                recovery();
                break;
            }
            case 'object': {
                this.createModalPasswordRecoveryComponent(recovery);
                break;
            }
        }
    }
    /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    checkingForMetadataProperty(object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    }
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    checkingForRouteMetadata(data) {
        if (Object.keys(data).length !== 0) {
            this.authenticationUrl = this.checkingForMetadataProperty(data, 'serviceApi') || this.authenticationUrl;
            this.authenticationType = this.checkingForMetadataProperty(data, 'authenticationType') || this.authenticationType;
            this.environment = this.checkingForMetadataProperty(data, 'environment') || this.environment;
            this.recovery = this.checkingForMetadataProperty(data, 'recovery') || this.recovery;
            this.registerUrl = this.checkingForMetadataProperty(data, 'registerUrl') || this.registerUrl;
        }
    }
    /**
     * @private
     * @param {?} defaultLiteral
     * @param {?} prefixLiteral
     * @param {?} value
     * @return {?}
     */
    concatenate(defaultLiteral, prefixLiteral, value) {
        return `${defaultLiteral} ${prefixLiteral} ${value}`;
    }
    /**
     * @private
     * @param {?} value
     * @param {?} literal
     * @param {?} defaultLiteral
     * @param {?} prepositionLiteral
     * @return {?}
     */
    concatenateLiteral(value, literal, defaultLiteral, prepositionLiteral) {
        return { [literal]: this.concatenate(defaultLiteral, prepositionLiteral, value) };
    }
    /**
     * @private
     * @param {?} poPageLoginRecovery
     * @return {?}
     */
    createModalPasswordRecoveryComponent(poPageLoginRecovery) {
        if (this.componentRef) {
            this.poComponentInjector.destroyComponentInApplication(this.componentRef);
        }
        this.componentRef = this.poComponentInjector.createComponentInApplication(PoModalPasswordRecoveryComponent);
        this.componentRef.instance.urlRecovery = poPageLoginRecovery.url;
        this.componentRef.instance.contactEmail = poPageLoginRecovery.contactMail;
        this.componentRef.instance.phoneMask = poPageLoginRecovery.phoneMask;
        this.componentRef.instance.type = poPageLoginRecovery.type || PoModalPasswordRecoveryType.Email;
        this.componentRef.changeDetectorRef.detectChanges();
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.componentRef.instance.open();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    generateLoginError() {
        if (this.loginErrors && this.loginErrors.length) {
            this.setLoginErrors(this.loginErrors);
        }
        else {
            /** @type {?} */
            const control = this.loginForm.form.controls['login'];
            if (control) {
                this.resetControl(control);
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    generatePasswordError() {
        if (this.passwordErrors && this.passwordErrors.length) {
            this.setPasswordErrors(this.passwordErrors);
        }
        else {
            /** @type {?} */
            const control = this.loginForm.form.controls['password'];
            if (control) {
                this.resetControl(control);
            }
        }
    }
    /**
     * @private
     * @param {?} control
     * @return {?}
     */
    resetControl(control) {
        control.markAsPristine();
        control.markAsUntouched();
        control.updateValueAndValidity();
    }
    /**
     * @private
     * @param {?} allErrors
     * @param {?} control
     * @param {?} errors
     * @param {?} patternError
     * @return {?}
     */
    setControlErrors(allErrors, control, errors, patternError) {
        if (control) {
            this[allErrors] = control.hasError('pattern') ? [...errors, ...[patternError]] : [...errors];
            if (errors && errors.length && (control.valid || control.pristine)) {
                control.markAsTouched();
                control.markAsDirty();
                control.setErrors(this.customPasswordError);
            }
        }
    }
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    setUrlRedirect(url) {
        isExternalLink(url) ? window.open(url, '_blank') : this.router.navigate([url]);
    }
    /**
     * @private
     * @param {?} differ
     * @param {?} array
     * @return {?}
     */
    validateArrayChanges(differ, array) {
        array.forEach((/**
         * @param {?} element
         * @return {?}
         */
        element => {
            /** @type {?} */
            const changes = differ.diff(element.array);
            if (changes) {
                element.callback();
                this.changeDetector.detectChanges();
            }
        }));
    }
    /**
     * @protected
     * @param {?} contactEmail
     * @return {?}
     */
    concatenateLoginHintWithContactEmail(contactEmail) {
        /** @type {?} */
        const defaultLoginHintLiteral = poPageLoginLiteralsDefault[this.language].loginHint;
        /** @type {?} */
        const prepositionLiteral = poPageLoginLiteralIn[this.language];
        return this.concatenateLiteral(contactEmail, 'loginHint', defaultLoginHintLiteral, prepositionLiteral);
    }
    /**
     * @protected
     * @param {?} productName
     * @return {?}
     */
    concatenateTitleWithProductName(productName) {
        /** @type {?} */
        const defaultTitleLiteral = poPageLoginLiteralsDefault[this.language].title;
        /** @type {?} */
        const prepositionLiteral = poPageLoginLiteralTo[this.language];
        return this.concatenateLiteral(productName, 'title', defaultTitleLiteral, prepositionLiteral);
    }
    /**
     * @protected
     * @param {?} errors
     * @return {?}
     */
    setLoginErrors(errors) {
        /** @type {?} */
        const control = this.loginForm.form.controls['login'];
        this.setControlErrors('allLoginErrors', control, errors, this.pageLoginLiterals.loginErrorPattern);
    }
    /**
     * @protected
     * @param {?} errors
     * @return {?}
     */
    setPasswordErrors(errors) {
        /** @type {?} */
        const control = this.loginForm.form.controls['password'];
        this.setControlErrors('allPasswordErrors', control, errors, this.pageLoginLiterals.passwordErrorPattern);
    }
}
PoPageLoginComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-page-login',
                template: "<po-page-background #pageLogin\n  p-show-select-language\n  [p-background]=\"background\"\n  [p-highlight-info]=\"pageLoginLiterals.highlightInfo\"\n  [p-logo]=\"logo\"\n  [p-secondary-logo]=\"secondaryLogo\"\n  (p-selected-language)=\"onSelectedLanguage($event)\">\n\n  <div class=\"po-page-login-header po-mb-2 po-mb-sm-1 po-pt-sm-1\">\n    <div class=\"po-page-login-header-title po-mb-sm-1\">{{ pageLoginLiterals.title }}</div>\n    <po-tag\n      *ngIf=\"environment\"\n      p-type=\"warning\"\n      [p-value]=\"environment\">\n    </po-tag>\n  </div>\n\n  <form #loginForm=\"ngForm\" class=\"po-page-login-form\">\n    <div class=\"po-row\">\n      <div class=\"po-lg-12\">\n        <div class=\"po-page-login-hint po-page-login-info-container\">\n          <po-login\n            class=\"po-page-login-info-field\"\n            name=\"login\"\n            [(ngModel)]=\"login\"\n            p-auto-focus\n            p-required\n            [p-label]=\"pageLoginLiterals.loginLabel\"\n            [p-pattern]=\"loginPattern\"\n            [p-placeholder]=\"pageLoginLiterals.loginPlaceholder\"\n            (click)=\"closePopover()\"\n            (keyup.enter)=\"loginForm.valid && onLoginSubmit()\"\n            (p-change-model)=\"changeLoginModel()\">\n          </po-login>\n\n          <div class=\"po-page-login-info-icon-container\">\n            <span\n              class=\"po-icon po-field-icon po-icon-info\"\n              p-tooltip-position=\"right\"\n              [p-tooltip]=\"pageLoginLiterals.loginHint\">\n            </span>\n          </div>\n        </div>\n\n        <div class=\"po-field-container-bottom po-field-container-error-container\">\n          <div *ngFor=\"let error of allLoginErrors\u001C\" class=\"po-field-container-bottom-text-error po-field-container-error-item\">\n            <span class=\"po-icon po-icon-exclamation\"></span>\n            <span class=\"po-field-container-error-text\">{{error}}</span>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"po-lg-12\">\n        <div class=\"po-page-login-password-container\">\n          <po-password\n            class=\"po-page-login-field-size po-page-login-password-item\"\n            name=\"password\"\n            [(ngModel)]=\"password\"\n            p-required\n            [p-label]=\"pageLoginLiterals.passwordLabel\"\n            [p-pattern]=\"passwordPattern\"\n            [p-placeholder]=\"pageLoginLiterals.passwordPlaceholder\"\n            (click)=\"closePopover()\"\n            (keyup.enter)=\"loginForm.valid && onLoginSubmit()\"\n            (p-change-model)=\"changePasswordModel()\">\n          </po-password>\n          <div class=\"po-page-login-password-item po-page-login-password-popover-container\">\n            <po-page-login-popover *ngIf=\"showExceededAttemptsWarning && exceededAttemptsWarning\"\n              [p-literals]=\"pageLoginLiterals\"\n              [p-recovery]=\"recovery\"\n              [p-remaining-attempts]=\"exceededAttemptsWarning\"\n              (p-forgot-password) = \"openUrl($event)\">\n            </po-page-login-popover>\n          </div>\n        </div>\n        <div class=\"po-field-container-bottom po-field-container-error-container\">\n          <div *ngFor=\"let error of allPasswordErrors\" class=\"po-field-container-bottom-text-error po-field-container-error-item\">\n            <span class=\"po-icon po-icon-exclamation\"></span>\n            <span class=\"po-field-container-error-text\">{{error}}</span>\n          </div>\n        </div>\n      </div>\n\n      <po-input *ngIf=\"customField && customFieldType === 'input'\"\n        class=\"po-page-login-field-size po-lg-12\"\n        name=\"customFieldInput\"\n        [(ngModel)]=\"customFieldObject.value\"\n        p-required\n        [p-error-pattern]=\"customFieldObject.errorPattern || pageLoginLiterals.customFieldErrorPattern\"\n        [p-pattern]=\"customFieldObject.pattern\"\n        [p-placeholder]=\"customFieldObject.placeholder || pageLoginLiterals.customFieldPlaceholder\"\n        (keyup.enter)=\"loginForm.valid && onLoginSubmit()\">\n      </po-input>\n\n      <po-combo *ngIf=\"customField && customFieldType === 'combo'\"\n        class=\"po-page-login-field-size po-lg-12\"\n        name=\"customFieldCombo\"\n        [(ngModel)]=\"customFieldObject.value\"\n        p-required\n        [p-field-value]=\"customFieldObject.fieldValue\"\n        [p-filter-service]=\"customFieldObject.url\"\n        [p-placeholder]=\"customFieldObject.placeholder || pageLoginLiterals.customFieldPlaceholder\">\n      </po-combo>\n\n      <po-select *ngIf=\"customField && customFieldType === 'select'\"\n        class=\"po-page-login-field-size po-lg-12\"\n        name=\"customFieldSelect\"\n        [(ngModel)]=\"customFieldObject.value\"\n        p-required\n        [p-placeholder]=\"customFieldObject.placeholder || pageLoginLiterals.customFieldPlaceholder\"\n        [p-options]=\"customFieldObject.options\">\n      </po-select>\n\n      <div class=\"po-lg-12\">\n        <div *ngIf=\"!hideRememberUser\" class=\"po-page-login-hint po-page-login-info-container\">\n          <po-switch\n            class=\"po-page-login-info-field po-lg-7 po-offset-lg-5 po-offset-xl-5\"\n            name=\"rememberUser\"\n            [(ngModel)]=\"rememberUser\"\n            p-label-position=\"1\"\n            [p-label-off]=\"pageLoginLiterals.rememberUser\"\n            [p-label-on]=\"pageLoginLiterals.rememberUser\"\n            (keyup.enter)=\"loginForm.valid && onLoginSubmit()\">\n          </po-switch>\n\n          <div class=\"po-page-login-info-icon-container po-page-login-info-icon-remember-user\">\n            <span\n              class=\"po-icon po-field-icon po-icon-info\"\n              p-tooltip-position=\"right\"\n              [p-tooltip]=\"pageLoginLiterals.rememberUserHint\">\n            </span>\n          </div>\n        </div>\n      </div>\n\n\n      <po-button\n        class=\"po-lg-12 po-page-login-button po-page-login-field-size\"\n        p-type=\"primary\"\n        [p-disabled]=\"loginForm.invalid\"\n        [p-label]=\"loading ? pageLoginLiterals.submittedLabel : pageLoginLiterals.submitLabel\"\n        [p-loading]=\"loading\"\n        (p-click)=\"onLoginSubmit()\">\n      </po-button>\n\n      <div *ngIf=\"recovery\"\n        class=\"po-page-login-recovery-link\">\n        <a class=\"po-font-text-large-bold\" (click)=\"openUrl(recovery)\">{{ pageLoginLiterals.forgotPassword }}</a>\n      </div>\n\n      <div *ngIf=\"registerUrl\"\n        class=\"po-page-login-register-link\">\n        <a class=\"po-font-text-large-bold\" (click)=\"openUrl(registerUrl)\">{{ pageLoginLiterals.registerUrl }}</a>\n      </div>\n    </div>\n  </form>\n</po-page-background>\n"
            }] }
];
/** @nocollapse */
PoPageLoginComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ActivatedRoute },
    { type: PoComponentInjectorService },
    { type: IterableDiffers },
    { type: PoPageLoginService },
    { type: Router }
];
PoPageLoginComponent.propDecorators = {
    loginForm: [{ type: ViewChild, args: ['loginForm', { read: NgForm, static: true },] }],
    pageLogin: [{ type: ViewChild, args: ['pageLogin', { read: ViewContainerRef, static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageLoginComponent.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    PoPageLoginComponent.prototype.differ;
    /**
     * @type {?}
     * @private
     */
    PoPageLoginComponent.prototype.customPasswordError;
    /** @type {?} */
    PoPageLoginComponent.prototype.loginForm;
    /** @type {?} */
    PoPageLoginComponent.prototype.pageLogin;
    /** @type {?} */
    PoPageLoginComponent.prototype.changeDetector;
    /**
     * @type {?}
     * @private
     */
    PoPageLoginComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    PoPageLoginComponent.prototype.poComponentInjector;
}
