// @portinari/portinari-templates/esm2015/lib/components/po-page-job-scheduler/po-page-job-scheduler-parameters/po-page-job-scheduler-parameters.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
export class PoPageJobSchedulerParametersComponent {
    constructor() {
        this.literals = (/** @type {?} */ ({}));
        this.parameters = [];
        this.valueChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.form) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.form.valueChanges.subscribe((/**
                 * @param {?} value
                 * @return {?}
                 */
                value => {
                    this.valueChange.emit(value);
                }));
            }));
        }
    }
}
PoPageJobSchedulerParametersComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-page-job-scheduler-parameters',
                template: "<ng-container *ngIf=\"parameters && parameters.length; then formFieldsTemplate; else parametersNotFoundTemplate\">\n</ng-container>\n\n<ng-template #parametersNotFoundTemplate>\n\n  <div class=\"po-text-center\">\n    <span class=\"po-icon po-icon-info\"></span>\n    <span class=\"po-font-text-large\">\n      {{ literals.parametersNotFound }}\n    </span>\n  </div>\n\n</ng-template>\n\n<ng-template #formFieldsTemplate>\n  <form #parametersForm=\"ngForm\">\n    <po-dynamic-form\n      p-group-form\n      [p-fields]=\"parameters\"\n      [p-value]=\"value\">\n    </po-dynamic-form>\n  </form>\n</ng-template>\n"
            }] }
];
PoPageJobSchedulerParametersComponent.propDecorators = {
    form: [{ type: ViewChild, args: ['parametersForm', { static: true },] }],
    literals: [{ type: Input, args: ['p-literals',] }],
    parameters: [{ type: Input, args: ['p-parameters',] }],
    value: [{ type: Input, args: ['p-value',] }],
    valueChange: [{ type: Output, args: ['p-valueChange',] }]
};
if (false) {
    /** @type {?} */
    PoPageJobSchedulerParametersComponent.prototype.form;
    /** @type {?} */
    PoPageJobSchedulerParametersComponent.prototype.literals;
    /** @type {?} */
    PoPageJobSchedulerParametersComponent.prototype.parameters;
    /** @type {?} */
    PoPageJobSchedulerParametersComponent.prototype.value;
    /** @type {?} */
    PoPageJobSchedulerParametersComponent.prototype.valueChange;
}
