// @portinari/portinari-ui/esm2015/lib/components/po-chart/po-chart-types/po-chart-dynamic-type.component.js
import * as tslib_1 from "tslib";
import { ElementRef, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
import { PoChartType } from '../enums/po-chart-type.enum';
const Padding = 24;
export class PoChartDynamicTypeComponent {
    constructor() {
        this.innerRadius = 0;
        this.onSerieClick = new Subject();
        this.onSerieHover = new Subject();
        this.series = [];
    }
    get isChartGaugeType() {
        return this.type === PoChartType.Gauge;
    }
    calculateSVGContainerDimensions(chartWrapperElement, chartHeaderElement, chartLegendElement) {
        const svgContainerHeightCalc = this.height - chartHeaderElement - chartLegendElement - (Padding * 2);
        this.svgHeight = svgContainerHeightCalc <= 0 ? 0 : svgContainerHeightCalc;
        this.centerX = chartWrapperElement / 2;
    }
    calculateTotalValue() {
        this.totalValue = this.type === PoChartType.Gauge ? 100 : this.series.reduce((previousValue, serie) => previousValue + serie.value, 0);
    }
}
tslib_1.__decorate([
    ViewChild('chartBody', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], PoChartDynamicTypeComponent.prototype, "chartBody", void 0);
tslib_1.__decorate([
    ViewChild('svgContainer', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], PoChartDynamicTypeComponent.prototype, "svgContainer", void 0);
