// @portinari/portinari-ui/esm5/lib/components/po-loading/po-loading-overlay/po-loading-overlay-base.component.js
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { convertToBoolean } from './../../../utils/util';
export var poLoadingOverlayLiteralsDefault = {
    en: {
        loading: 'Loading',
    },
    es: {
        loading: 'Cargando',
    },
    pt: {
        loading: 'Carregando',
    },
    ru: {
        loading: 'погрузка',
    }
};
/**
 *
 * @description
 *
 * Este componente mostra ao usuário uma imagem de _loading_ e bloqueia a página inteira ou o container escolhido,
 * enquanto aguarda a resposta de alguma requisição.
 */
var PoLoadingOverlayBaseComponent = /** @class */ (function () {
    function PoLoadingOverlayBaseComponent(languageService) {
        this.languageService = languageService;
        this._screenLock = false;
        if (!this.text) {
            this.text = this.getTextDefault();
        }
    }
    Object.defineProperty(PoLoadingOverlayBaseComponent.prototype, "screenLock", {
        get: function () {
            return this._screenLock;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define se o *overlay* será aplicado a um *container* ou a página inteira.
         *
         * Para utilizar o componente como um *container*, o elemento pai deverá receber uma posição relativa, por exemplo:
         *
         * ```
         * <div style="position: relative">
         *
         *  <po-chart [p-series]="[{ value: 10, category: 'Example' }]">
         *  </po-chart>
         *
         *  <po-loading-overlay>
         *  </po-loading-overlay>
         * </div>
         * ```
         *
         * @default `false`
         */
        set: function (screenLock) {
            this._screenLock = convertToBoolean(screenLock);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoLoadingOverlayBaseComponent.prototype, "text", {
        get: function () {
            return this._text;
        },
        /**
         * @optional
         *
         * @description
         *
         * Texto a ser exibido no componente.
         *
         * > O valor padrão será traduzido acordo com o idioma configurado no [**PoI18n**](/documentation/po-i18n) ou navegador.
         *
         * @default `Carregando`
         */
        set: function (value) {
            this._text = value || this.getTextDefault();
        },
        enumerable: true,
        configurable: true
    });
    PoLoadingOverlayBaseComponent.prototype.getTextDefault = function () {
        var language = this.languageService.getShortLanguage();
        return poLoadingOverlayLiteralsDefault[language].loading;
    };
    tslib_1.__decorate([
        Input('p-screen-lock'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoLoadingOverlayBaseComponent.prototype, "screenLock", null);
    tslib_1.__decorate([
        Input('p-text'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], PoLoadingOverlayBaseComponent.prototype, "text", null);
    return PoLoadingOverlayBaseComponent;
}());
export { PoLoadingOverlayBaseComponent };
