// @portinari/portinari-templates/esm5/lib/components/po-modal-password-recovery/po-modal-password-recovery.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
import { Router } from '@angular/router';
import { isExternalLink } from '../../utils/util';
import { PoI18nPipe, PoModalComponent } from '@portinari/portinari-ui';
import { PoModalPasswordRecoveryBaseComponent } from './po-modal-password-recovery-base.component';
import { PoModalPasswordRecoveryModalContent } from './enums/po-modal-password-recovery-modal-content.enum';
import { PoModalPasswordRecoveryService } from './po-modal-password-recovery.service';
import { PoModalPasswordRecoveryType } from './enums/po-modal-password-recovery-type.enum';
/**
 * \@docsExtends PoModalPasswordRecoveryBaseComponent
 *
 * \@example
 *
 * <example name="po-modal-password-recovery-basic" title="Portinari Modal Password Recovery Basic">
 *  <file name="sample-po-modal-password-recovery-basic/sample-po-modal-password-recovery-basic.component.html"> </file>
 *  <file name="sample-po-modal-password-recovery-basic/sample-po-modal-password-recovery-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-modal-password-recovery-labs" title="Portinari Modal Password Recovery Labs">
 *  <file name="sample-po-modal-password-recovery-labs/sample-po-modal-password-recovery-labs.component.html"> </file>
 *  <file name="sample-po-modal-password-recovery-labs/sample-po-modal-password-recovery-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-modal-password-recovery-request" title="Portinari Modal Password Recovery Request">
 *  <file name="sample-po-modal-password-recovery-request/sample-po-modal-password-recovery-request.component.html"> </file>
 *  <file name="sample-po-modal-password-recovery-request/sample-po-modal-password-recovery-request.component.ts"> </file>
 * </example>
 */
var PoModalPasswordRecoveryComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoModalPasswordRecoveryComponent, _super);
    function PoModalPasswordRecoveryComponent(router, poI18nPipe, poModalPasswordRecoveryService) {
        var _this = _super.call(this) || this;
        _this.router = router;
        _this.poI18nPipe = poI18nPipe;
        _this.poModalPasswordRecoveryService = poModalPasswordRecoveryService;
        _this.chosenTypeFormOption = PoModalPasswordRecoveryType.Email;
        _this.codeMask = '9 9 9 9 9 9';
        _this.emailModal = true;
        _this.emailModalPhrases = { firstPhrase: (/** @type {?} */ (null)), secondPhrase: (/** @type {?} */ (null)) };
        _this.endpoint = '.';
        _this.invalidCode = false;
        _this.invalidEmail = false;
        _this.modalType = PoModalPasswordRecoveryModalContent.Email;
        _this.submittedCodeValue = (/** @type {?} */ ({}));
        _this.submittedContactValue = (/** @type {?} */ ({}));
        _this.primaryAction = { label: undefined, action: (/**
             * @return {?}
             */
            function () { }) };
        _this.secondaryAction = { label: undefined, action: (/**
             * @return {?}
             */
            function () { }) };
        _this.typeFormOptions = [
            { label: 'e-mail', value: PoModalPasswordRecoveryType.Email },
            { label: 'SMS', value: PoModalPasswordRecoveryType.SMS }
        ];
        return _this;
    }
    /**
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.passwordRecoverySubscription) {
            this.passwordRecoverySubscription.unsubscribe();
        }
        if (this.smsCodeSubscription) {
            this.smsCodeSubscription.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.completed = /**
     * @return {?}
     */
    function () {
        this.cancelAction();
    };
    /**
     * @param {?} form
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.formModelChangesCheck = /**
     * @param {?} form
     * @return {?}
     */
    function (form) {
        /** @type {?} */
        var invalidForm = form.invalid;
        this.invalidEmail = invalidForm && form.dirty;
        this.primaryAction.disabled = invalidForm;
        if (this.modalType === PoModalPasswordRecoveryModalContent.SMSCode) {
            /** @type {?} */
            var codeError = this.codeError !== undefined && this.codeError !== '';
            this.showCustomCodeError = codeError && form.pristine;
        }
    };
    /**
     * @param {?} type
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.getInputType = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        var _this = this;
        this.type = type;
        this.pipeModalPhrases();
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.control = _this.emailForm.controls[type];
            _this.formModelChangesCheck(_this.emailForm);
            _this.resetFormFields(_this.control);
        }));
    };
    /**
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var control = this.checkFormType(this.type);
        this.control = this.emailForm.controls[control];
        this.setEmailModalPhrasesAndActions();
        this.formModelChangesCheck(this.emailForm);
        this.recoveryModalElement.open();
    };
    /**
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.openConfirmation = /**
     * @return {?}
     */
    function () {
        this.modalTitle = this.literals.emailSentTitle;
        this.modalType = PoModalPasswordRecoveryModalContent.Confirmation;
        this.setActions(this.cancelAction, this.literals.closeButton, this.submitAction, this.literals.resendEmailButton, false);
    };
    /**
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.openSmsCode = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.modalTitle = this.literals.typeCodeTitle;
        this.modalType = PoModalPasswordRecoveryModalContent.SMSCode;
        this.setActions(this.submitSmsCodeAction, this.literals.continueButton, this.cancelAction, this.literals.cancelButton, true);
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.control = _this.smsCodeForm.controls['sms'];
            _this.formModelChangesCheck(_this.smsCodeForm);
        }));
    };
    /**
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.resendSmsCode = /**
     * @return {?}
     */
    function () {
        this.incrementRetryAttempts();
        if (this.urlRecovery) {
            this.submitActionRequest(this.submittedContactValue, this.type);
        }
        else {
            this.submit.emit(this.submittedContactValue);
        }
    };
    /**
     * @private
     * @param {?} responseObj
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.assignSmsResponse = /**
     * @private
     * @param {?} responseObj
     * @return {?}
     */
    function (responseObj) {
        this.smsBodyResponse = Object.assign({}, { hash: responseObj.hash });
        if (responseObj.urlValidationCode) {
            this.smsBodyResponse = Object.assign(this.smsBodyResponse, { urlValidationCode: responseObj.urlValidationCode });
        }
    };
    /**
     * @private
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.cancelAction = /**
     * @private
     * @return {?}
     */
    function () {
        this.resetFormFields(this.control);
        this.submittedContactValue = {};
        this.chosenTypeFormOption = PoModalPasswordRecoveryType.Email;
        this.modalType = PoModalPasswordRecoveryModalContent.Email;
        this.type = this.modalPasswordRecoveryTypeAll ? PoModalPasswordRecoveryType.All : this.type;
        this.recoveryModalElement.close();
    };
    /**
     * @private
     * @param {?} type
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.checkFormType = /**
     * @private
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return type !== PoModalPasswordRecoveryType.All ? type : PoModalPasswordRecoveryType.Email;
    };
    /**
     * @private
     * @param {?} control
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.formReset = /**
     * @private
     * @param {?} control
     * @return {?}
     */
    function (control) {
        control.markAsPristine();
        control.markAsUntouched();
        control.updateValueAndValidity();
    };
    /**
     * @private
     * @param {?} type
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.getEmitValue = /**
     * @private
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return type === PoModalPasswordRecoveryType.SMS ? this.phone : this.email;
    };
    /**
     * @private
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.incrementRetryAttempts = /**
     * @private
     * @return {?}
     */
    function () {
        this.submittedContactValue.retry = (this.submittedContactValue.retry + 1 || 1);
    };
    /**
     * @private
     * @param {?} url
     * @param {?} queryParam
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.openExternalLink = /**
     * @private
     * @param {?} url
     * @param {?} queryParam
     * @return {?}
     */
    function (url, queryParam) {
        window.open(url + "?token=" + queryParam, '_self');
    };
    /**
     * @private
     * @param {?} url
     * @param {?} endpoint
     * @param {?} queryParam
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.openInternalLink = /**
     * @private
     * @param {?} url
     * @param {?} endpoint
     * @param {?} queryParam
     * @return {?}
     */
    function (url, endpoint, queryParam) {
        this.router.navigate([url + "/" + endpoint], { queryParams: { token: queryParam } });
    };
    /**
     * @private
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.pipeModalPhrases = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.type === PoModalPasswordRecoveryType.SMS) {
            this.emailModalPhrases.firstPhrase = this.setPipeArguments(this.literals.recoveryPasswordPhrase, this.literals.sms);
            this.emailModalPhrases.secondPhrase = this.setPipeArguments(this.literals.supportContact, this.literals.telephone);
        }
        else {
            this.emailModalPhrases.firstPhrase = this.setPipeArguments(this.literals.recoveryPasswordPhrase, this.literals.email);
            this.emailModalPhrases.secondPhrase = this.setPipeArguments(this.literals.supportContact, this.literals.email);
        }
    };
    /**
     * @private
     * @param {?} recoveryToken
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.redirectToChangePassword = /**
     * @private
     * @param {?} recoveryToken
     * @return {?}
     */
    function (recoveryToken) {
        /** @type {?} */
        var urlChangePassword = recoveryToken.urlChangePassword;
        if (urlChangePassword) {
            isExternalLink(urlChangePassword) ?
                this.openExternalLink(urlChangePassword, recoveryToken.token) :
                this.openInternalLink(this.urlRecovery, urlChangePassword, recoveryToken.token);
        }
        else {
            /** @type {?} */
            var changePasswordEndpoint = 'changePassword';
            this.openInternalLink(this.urlRecovery, changePasswordEndpoint, recoveryToken.token);
        }
    };
    /**
     * @private
     * @param {?} control
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.resetFormFields = /**
     * @private
     * @param {?} control
     * @return {?}
     */
    function (control) {
        this.formReset(control);
        this.email = undefined;
        this.phone = undefined;
        this.smsCode = undefined;
    };
    /**
     * @private
     * @param {?} primaryAction
     * @param {?} primarylabel
     * @param {?} secondaryAction
     * @param {?} secondaryLabel
     * @param {?} disabled
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.setActions = /**
     * @private
     * @param {?} primaryAction
     * @param {?} primarylabel
     * @param {?} secondaryAction
     * @param {?} secondaryLabel
     * @param {?} disabled
     * @return {?}
     */
    function (primaryAction, primarylabel, secondaryAction, secondaryLabel, disabled) {
        var _this = this;
        this.primaryAction.action = (/**
         * @return {?}
         */
        function () { return primaryAction.call(_this); });
        this.primaryAction.label = primarylabel;
        this.secondaryAction.action = (/**
         * @return {?}
         */
        function () { return secondaryAction.call(_this); });
        this.secondaryAction.label = secondaryLabel;
        this.primaryAction.disabled = disabled;
    };
    /**
     * @private
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.setEmailModalPhrasesAndActions = /**
     * @private
     * @return {?}
     */
    function () {
        this.modalTitle = this.literals.forgotPasswordTitle;
        this.pipeModalPhrases();
        this.modalPasswordRecoveryTypeAll = this.type === PoModalPasswordRecoveryType.All;
        this.setActions(this.submitAction, this.literals.sendButton, this.cancelAction, this.literals.cancelButton, true);
    };
    /**
     * @private
     * @param {?=} urlValidationCode
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.setRequestEndpoint = /**
     * @private
     * @param {?=} urlValidationCode
     * @return {?}
     */
    function (urlValidationCode) {
        /** @type {?} */
        var endpoint = urlValidationCode || 'validation';
        return this.urlRecovery + "/" + endpoint;
    };
    /**
     * @private
     * @param {?} literalAttr
     * @param {?} arg
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.setPipeArguments = /**
     * @private
     * @param {?} literalAttr
     * @param {?} arg
     * @return {?}
     */
    function (literalAttr, arg) {
        return this.poI18nPipe.transform(literalAttr, arg);
    };
    /**
     * @private
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.submitAction = /**
     * @private
     * @return {?}
     */
    function () {
        this.modalType === PoModalPasswordRecoveryModalContent.Confirmation ? this.incrementRetryAttempts() : this.formReset(this.control);
        this.submittedContactValue[this.checkFormType(this.type)] = this.getEmitValue(this.type);
        if (this.urlRecovery) {
            this.submitActionRequest(this.submittedContactValue, this.type);
        }
        else {
            this.submit.emit(this.submittedContactValue);
        }
    };
    /**
     * @private
     * @param {?} data
     * @param {?} modalType
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.submitActionRequest = /**
     * @private
     * @param {?} data
     * @param {?} modalType
     * @return {?}
     */
    function (data, modalType) {
        var _this = this;
        /** @type {?} */
        var params = modalType === PoModalPasswordRecoveryType.SMS ? { type: 'sms' } : undefined;
        this.passwordRecoverySubscription =
            this.poModalPasswordRecoveryService.post(this.urlRecovery, data, params)
                .subscribe((/**
             * @param {?} response
             * @return {?}
             */
            function (response) {
                if ((modalType === PoModalPasswordRecoveryType.Email || modalType === PoModalPasswordRecoveryType.All) && response.status === 204) {
                    _this.openConfirmation();
                }
                else if (modalType === PoModalPasswordRecoveryType.SMS && response.status === 200) {
                    _this.assignSmsResponse(response.body);
                    _this.openSmsCode();
                }
            }));
    };
    /**
     * @private
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.submitSmsCodeAction = /**
     * @private
     * @return {?}
     */
    function () {
        this.submittedCodeValue.code = this.smsCode;
        if (this.urlRecovery) {
            this.submittedCodeValue = Object.assign(this.submittedCodeValue, { hash: this.smsBodyResponse.hash });
            this.submitSmsCodeRequest(this.submittedCodeValue);
        }
        else {
            this.codeSubmit.emit(this.submittedCodeValue);
        }
        this.resetFormFields(this.control);
    };
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    PoModalPasswordRecoveryComponent.prototype.submitSmsCodeRequest = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this.smsCodeSubscription = this.poModalPasswordRecoveryService
            .post(this.setRequestEndpoint(this.smsBodyResponse.urlValidationCode), data)
            .subscribe((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            /** @type {?} */
            var successStatus = response.status === 200;
            if (successStatus) {
                _this.completed();
                _this.redirectToChangePassword(response.body);
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.codeError = error.error.message;
            _this.openSmsCode();
        }));
    };
    PoModalPasswordRecoveryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-modal-password-recovery',
                    template: "<po-modal #recoveryModal\r\n  p-hide-close\r\n  p-size=\"auto\"\r\n  [ngSwitch]=\"modalType\"\r\n  [p-primary-action]=\"primaryAction\"\r\n  [p-secondary-action]=\"secondaryAction\"\r\n  [p-title]=\"modalTitle\"> \r\n  <div class=\"po-modal-password-recovery-wrapper\">\r\n    <div *ngSwitchCase=\"'email'\" class=\"po-modal-password-recovery-content po-row\">\r\n      <div class=\"po-modal-password-recovery-text po-md-12 po-mb-1\">\r\n        {{ emailModalPhrases.firstPhrase }}\r\n      </div>\r\n      <div class=\"po-mb-2 po-md-12\">\r\n        <form #emailForm=\"ngForm\">\r\n\r\n          <div *ngIf=\"modalPasswordRecoveryTypeAll\">\r\n            <po-radio-group \r\n              name=\"type\"\r\n              [(ngModel)]=\"chosenTypeFormOption\"\r\n              [p-options]=\"typeFormOptions\"\r\n              (p-change)=\"getInputType($event)\">\r\n            </po-radio-group>\r\n          </div>\r\n\r\n          <div class=\"po-mt-1\">\r\n            <po-email *ngIf=\"type === 'email' || type === 'all'\"\r\n              name=\"email\"\r\n              [(ngModel)]=\"email\"\r\n              p-required\r\n              [p-label]=\"literals.insertEmail\"\r\n              (p-change-model)=\"formModelChangesCheck(emailForm)\">\r\n            </po-email>\r\n    \r\n            <po-input *ngIf=\"type === 'sms'\"\r\n              name=\"sms\"\r\n              [(ngModel)]=\"phone\"\r\n              p-icon=\"po-icon-telephone\"\r\n              p-required\r\n              [p-label]=\"literals.insertPhone\"\r\n              [p-mask]=\"phoneMask\"\r\n              [p-maxlength]=\"maxLength\"\r\n              [p-minlength]=\"minLength\"\r\n              (p-change-model)=\"formModelChangesCheck(emailForm)\">\r\n            </po-input>\r\n          </div>\r\n          \r\n          <div class=\"po-field-container-bottom po-field-container-error-container\">\r\n            <po-modal-password-recovery-error-message *ngIf=\"invalidEmail && control.dirty\" \r\n              [p-text]=\"type === 'sms' ? literals.phoneErrorMessagePhrase : literals.emailErrorMessagePhrase\">\r\n            </po-modal-password-recovery-error-message>\r\n          </div>\r\n        </form>\r\n      </div>\r\n      <div class=\"po-modal-password-recovery-text po-md-12\">{{ emailModalPhrases.secondPhrase }}<span *ngIf=\"contactEmail\">\r\n        {{ literals.prepositionIn }} \r\n          <a class=\"po-modal-password-recovery-link\" href=\"mailto:{{ contactEmail }}\" target=\"_self\">{{ contactEmail }}</a>\r\n        </span>{{ endpoint }}\r\n      </div>\r\n    </div>\r\n\r\n    <div #smsCodeModal *ngSwitchCase=\"'smsCode'\" class=\"po-modal-password-recovery-content po-row\">\r\n      <div class=\"po-modal-password-recovery-text po-md-12 po-mb-1\">{{ literals.sentSmsCodePhrase }}</div>\r\n      <div class=\"po-mb-2 po-md-12\">\r\n        <form #smsCodeForm=\"ngForm\">\r\n          <po-input\r\n            name=\"sms\"\r\n            [(ngModel)]=\"smsCode\"\r\n            p-maxlength=\"11\"\r\n            p-minlength=\"11\"\r\n            p-required\r\n            [p-label]=\"literals.insertCode\"\r\n            [p-mask]=\"codeMask\"\r\n            (p-change-model)=\"formModelChangesCheck(smsCodeForm)\">\r\n          </po-input>\r\n          <div class=\"po-field-container-bottom po-field-container-error-container\">\r\n            <po-modal-password-recovery-error-message *ngIf=\"invalidEmail\"\r\n              [p-text]=\"smsCodeErrorMessage || this.literals.smsCodeErrorMessagePhrase\">\r\n            </po-modal-password-recovery-error-message>\r\n            <po-modal-password-recovery-error-message *ngIf=\"showCustomCodeError\"\r\n              [p-text]=\"codeError\">\r\n            </po-modal-password-recovery-error-message>\r\n          </div>\r\n        </form>\r\n      </div>\r\n      <div class=\"po-modal-password-recovery-text po-md-12\">\r\n        {{ literals.sendAgainPhrase }} \r\n        <span class=\"po-modal-password-recovery-link\" (click)=\"resendSmsCode()\">{{ literals.sendAgain }}</span>\r\n      </div>\r\n    </div>\r\n    \r\n    <div #confirmationModal *ngSwitchCase=\"'confirmation'\" class=\"po-modal-password-recovery-content po-row\">\r\n      <img \r\n        class=\"po-modal-password-recovery-user-image po-mb-2\" \r\n        src=\"./assets/images/email-sent.svg\"> \r\n      <div class=\"po-modal-password-recovery-text\">\r\n        {{ literals.emailSentConfirmationPhrase }}\r\n      </div>\r\n    </div>\r\n  </div>\r\n</po-modal>\r\n"
                }] }
    ];
    /** @nocollapse */
    PoModalPasswordRecoveryComponent.ctorParameters = function () { return [
        { type: Router },
        { type: PoI18nPipe },
        { type: PoModalPasswordRecoveryService }
    ]; };
    PoModalPasswordRecoveryComponent.propDecorators = {
        emailForm: [{ type: ViewChild, args: ['emailForm', { static: false },] }],
        recoveryModalElement: [{ type: ViewChild, args: ['recoveryModal', { static: true },] }],
        smsCodeForm: [{ type: ViewChild, args: ['smsCodeForm', { static: false },] }]
    };
    return PoModalPasswordRecoveryComponent;
}(PoModalPasswordRecoveryBaseComponent));
export { PoModalPasswordRecoveryComponent };
if (false) {
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.chosenTypeFormOption;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.codeMask;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.control;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.emailModal;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.emailModalPhrases;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.endpoint;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.invalidCode;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.invalidEmail;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.modalTitle;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.modalType;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.showCustomCodeError;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.smsCodeErrorMessagePhrase;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.submittedCodeValue;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.submittedContactValue;
    /**
     * @type {?}
     * @private
     */
    PoModalPasswordRecoveryComponent.prototype.passwordRecoverySubscription;
    /**
     * @type {?}
     * @private
     */
    PoModalPasswordRecoveryComponent.prototype.smsBodyResponse;
    /**
     * @type {?}
     * @private
     */
    PoModalPasswordRecoveryComponent.prototype.smsCodeSubscription;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.emailForm;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.recoveryModalElement;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.smsCodeForm;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.primaryAction;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.secondaryAction;
    /** @type {?} */
    PoModalPasswordRecoveryComponent.prototype.typeFormOptions;
    /**
     * @type {?}
     * @private
     */
    PoModalPasswordRecoveryComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PoModalPasswordRecoveryComponent.prototype.poI18nPipe;
    /**
     * @type {?}
     * @private
     */
    PoModalPasswordRecoveryComponent.prototype.poModalPasswordRecoveryService;
}
