// @portinari/portinari-ui/esm2015/lib/components/po-field/po-multiselect/po-multiselect-item/po-multiselect-item.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que constrói cada item no dropdown, contendo o checkbox e o label.
 */
let PoMultiselectItemComponent = class PoMultiselectItemComponent {
    /**
     * @docsPrivate
     *
     * @description
     *
     * Componente que constrói cada item no dropdown, contendo o checkbox e o label.
     */
    constructor() {
        /** Esta propriedade indica se o campo está selecionado ou não. */
        this.selected = false;
        /** Evento que será disparado toda vez que o usuário marcar ou desmarcar um item. */
        this.change = new EventEmitter();
    }
    itemClicked() {
        this.selected = !this.selected;
        this.change.emit(this.selected);
    }
};
tslib_1.__decorate([
    Input('p-label'),
    tslib_1.__metadata("design:type", String)
], PoMultiselectItemComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Input('p-selected'),
    tslib_1.__metadata("design:type", Boolean)
], PoMultiselectItemComponent.prototype, "selected", void 0);
tslib_1.__decorate([
    Output('p-change'),
    tslib_1.__metadata("design:type", Object)
], PoMultiselectItemComponent.prototype, "change", void 0);
PoMultiselectItemComponent = tslib_1.__decorate([
    Component({
        selector: 'po-multiselect-item',
        template: "<li [class.po-multiselect-item-selected]=\"selected\" (click)=\"itemClicked()\">\n\n  <a class=\"po-multiselect-item\">\n    <input\n      [checked]=\"selected\"\n      class=\"po-multiselect-checkbox-input\"\n      [class.po-multiselect-checkbox-input-checked]=\"selected\"\n      type=\"checkbox\">\n\n    <label class=\"po-multiselect-checkbox-label po-clickable\">\n      {{ label }}\n    </label>\n  </a>\n</li>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], PoMultiselectItemComponent);
export { PoMultiselectItemComponent };
