// @portinari/portinari-ui/esm5/lib/components/po-field/po-field-container/po-field-container.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { convertToBoolean, getShortBrowserLanguage } from '../../../utils/util';
/**
 * @docsPrivate
 *
 * Componente de uso interno, responsável por atribuir uma label para o campo
 */
var PoFieldContainerComponent = /** @class */ (function () {
    function PoFieldContainerComponent() {
        this._optional = false;
    }
    Object.defineProperty(PoFieldContainerComponent.prototype, "optional", {
        get: function () {
            return this._optional;
        },
        /** Indica se o campo será opcional. */
        set: function (value) {
            this._optional = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    PoFieldContainerComponent.prototype.getOptionalText = function () {
        var browserLanguage = getShortBrowserLanguage();
        var optional = {
            pt: '(Opcional)',
            en: '(Optional)',
            es: '(Opcional)'
        };
        return optional[browserLanguage];
    };
    tslib_1.__decorate([
        Input('p-label'),
        tslib_1.__metadata("design:type", String)
    ], PoFieldContainerComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input('p-help'),
        tslib_1.__metadata("design:type", String)
    ], PoFieldContainerComponent.prototype, "help", void 0);
    tslib_1.__decorate([
        Input('p-optional'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PoFieldContainerComponent.prototype, "optional", null);
    PoFieldContainerComponent = tslib_1.__decorate([
        Component({
            selector: 'po-field-container',
            template: "<div class=\"po-field-container\">\n  <div class=\"po-field-container-title\">\n    <span class=\"po-field-title\">{{ label }}</span>\n    <span class=\"po-field-optional\" *ngIf=\"optional && (label || help)\">{{ getOptionalText() }}</span>\n  </div>\n\n  <div class=\"po-field-help\" *ngIf=\"help\">{{ help }}</div>\n\n  <ng-content></ng-content>\n</div>\n"
        })
    ], PoFieldContainerComponent);
    return PoFieldContainerComponent;
}());
export { PoFieldContainerComponent };
