// @portinari/portinari-ui/esm5/lib/components/po-page/po-page-detail/po-page-detail.component.js
import * as tslib_1 from "tslib";
import { Component, ViewContainerRef } from '@angular/core';
import { PoPageDetailBaseComponent } from './po-page-detail-base.component';
import { callAction, hasAction } from '../po-page-util/po-page-util';
/**
 * @docsExtends PoPageDetailBaseComponent
 *
 * @example
 *
 * <example name="po-page-detail-basic" title="Portinari Page Detail Basic">
 *  <file name="sample-po-page-detail-basic/sample-po-page-detail-basic.component.html"> </file>
 *  <file name="sample-po-page-detail-basic/sample-po-page-detail-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-detail-labs" title="Portinari Page Detail Labs">
 *  <file name="sample-po-page-detail-labs/sample-po-page-detail-labs.component.html"> </file>
 *  <file name="sample-po-page-detail-labs/sample-po-page-detail-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-detail-user" title="Portinari Page Detail - User">
 *  <file name="sample-po-page-detail-user/sample-po-page-detail-user.component.html"> </file>
 *  <file name="sample-po-page-detail-user/sample-po-page-detail-user.component.ts"> </file>
 * </example>
 */
var PoPageDetailComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoPageDetailComponent, _super);
    function PoPageDetailComponent(viewRef) {
        var _this = _super.call(this) || this;
        _this.callActionFn = callAction;
        _this.hasActionFn = hasAction;
        _this.parentContext = viewRef['_view']['component'];
        return _this;
    }
    PoPageDetailComponent.prototype.hasAnyAction = function () {
        return this.hasActionFn('back', this.parentContext) ||
            this.hasActionFn('edit', this.parentContext) ||
            this.hasActionFn('remove', this.parentContext);
    };
    PoPageDetailComponent.prototype.hasEditFn = function (property) {
        if (property === 'icon') {
            return this.hasActionFn('edit', this.parentContext) ? '' : 'po-icon-delete';
        }
        else if (property === 'type') {
            return this.hasActionFn('edit', this.parentContext) ? 'default' : 'primary';
        }
        else {
            return '';
        }
    };
    PoPageDetailComponent.prototype.hasEditOrRemoveFn = function (property) {
        if (property === 'icon') {
            return this.hasActionFn('edit', this.parentContext) || this.hasActionFn('remove', this.parentContext) ? '' : 'po-icon-arrow-left';
        }
        else if (property === 'type') {
            return this.hasActionFn('edit', this.parentContext) || this.hasActionFn('remove', this.parentContext) ? 'default' : 'primary';
        }
        else {
            return '';
        }
    };
    PoPageDetailComponent.prototype.hasPageHeader = function () {
        return !!(this.title || this.hasAnyAction() || (this.breadcrumb && this.breadcrumb.items.length));
    };
    PoPageDetailComponent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    PoPageDetailComponent = tslib_1.__decorate([
        Component({
            selector: 'po-page-detail',
            template: "<po-page>\n  <!-- HEADER -->\n  <po-page-header\n    *ngIf=\"hasPageHeader()\"\n    [p-breadcrumb]=\"breadcrumb\"\n    [p-title]=\"title\">\n\n    <!-- OPERATIONS -->\n    <div *ngIf=\"hasAnyAction()\" class=\"po-page-header-actions\">\n      <po-button *ngIf=\"hasActionFn('back', parentContext)\"\n        [p-icon]=\"hasEditOrRemoveFn('icon')\"\n        [p-label]=\"literals.back\"\n        [p-type]=\"hasEditOrRemoveFn('type')\"\n        (p-click)=\"callActionFn('back', parentContext)\">\n      </po-button>\n\n      <po-button *ngIf=\"hasActionFn('remove', parentContext)\"\n        [p-icon]=\"hasEditFn('icon')\"\n        [p-label]=\"literals.remove\"\n        [p-type]=\"hasEditFn('type')\"\n        (p-click)=\"callActionFn('remove', parentContext)\">\n      </po-button>\n\n      <po-button *ngIf=\"hasActionFn('edit', parentContext)\"\n        p-icon=\"po-icon-edit\"\n        p-type=\"primary\"\n        [p-label]=\"literals.edit\"\n        (p-click)=\"callActionFn('edit', parentContext)\">\n      </po-button>\n    </div>\n\n  </po-page-header>\n\n  <!-- CONTENT -->\n  <po-page-content>\n    <ng-content>\n    </ng-content>\n  </po-page-content>\n\n</po-page>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [ViewContainerRef])
    ], PoPageDetailComponent);
    return PoPageDetailComponent;
}(PoPageDetailBaseComponent));
export { PoPageDetailComponent };
